
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>GWAS data QC workflow &#8212; ADSP-FG Workflows</title>
    
  <link href="../../../_static/css/theme.css" rel="stylesheet">
  <link href="../../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/togglebutton.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Genomic relationship matrices" href="GRM.html" />
    <link rel="prev" title="Genotype data preprocessing" href="../genotype_preprocessing.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">ADSP-FG Workflows</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../../README.html">
   xQTL analysis pipeline
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Complete Pipelines
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../complete_analysis/eQTL_analysis.html">
   eQTL analysis complete workflow
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../molecular_phenotypes/apa_calling.html">
   APA Calling
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="../genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     GWAS data QC workflow
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="GRM.html">
     Genomic relationship matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="genotype_formatting.html">
     Genotype data reformat
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../phenotype/normalization.html">
     Molecular phenotype normalization
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../phenotype/phenotype_formatting.html">
     Phenotype data reformat
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../covariate_preprocessing.html">
   Covariate data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../covariate/PCA.html">
     Principal component analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../covariate/PEER_factor.html">
     Factor analysis using PEER
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../covariate/BiCV_factor.html">
     Factor analysis using Bi-Cross validation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../covariate/remove_covariates.html">
     Computing covariate residuals
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../association_scan/cisQTL_scan.html">
   cisQTL analysis workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL calling pipeline
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/APEX/APEX_LMM.html">
     APEX QTL calling pipeline
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../association_scan/transQTL_scan.html">
   transQTL analysis workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../multivariate/MASH.html">
   MASH analysis workflow
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../multivariate/METAL.html">
   METAL workflow
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Statistical Fine-mapping
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../fine_mapping/univariate_fine_mapping.html">
   Univariate fine-mapping workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE_RSS.html">
     Fine-mapping with SuSiE RSS model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../fine_mapping/multivariate_fine_mapping.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Utilities
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../misc/summary_stats_merger.html">
   Summary statistics formatting
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../misc/rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH/NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../../_sources/workflow/data_preprocessing/genotype/GWAS_QC.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fworkflow/data_preprocessing/genotype/GWAS_QC.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/workflow/data_preprocessing/genotype/GWAS_QC.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#overview">
   Overview
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#run-this-workflow">
   Run this workflow
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#minimal-working-example">
   Minimal working example
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#first-scenario-estimate-kinship">
     First scenario: estimate kinship
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#first-scenario-do-qc">
     First scenario: do qc
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#estimate-kinship-in-the-sample">
   Estimate kinship in the sample
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#genotype-and-sample-qc">
   Genotype and sample QC
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="gwas-data-qc-workflow">
<h1>GWAS data QC workflow<a class="headerlink" href="#gwas-data-qc-workflow" title="Permalink to this headline">¶</a></h1>
<p>This workflow implements some prelimary data QC steps for PLINK input files.</p>
<div class="section" id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>This notebook includes workflow for</p>
<ul class="simple">
<li><p>Computer kinship matrix in sample and estimate related individuals</p></li>
<li><p>Genotype and sample QC: by MAF, missing data and HWE</p></li>
<li><p>LD pruning</p></li>
</ul>
</div>
<div class="section" id="run-this-workflow">
<h2>Run this workflow<a class="headerlink" href="#run-this-workflow" title="Permalink to this headline">¶</a></h2>
<p>Depending on the context of your problem, the workflow can be executed in two ways:</p>
<ol class="simple">
<li><p>Run <code class="docutils literal notranslate"><span class="pre">merge_plink</span></code> if necessary, to merge all samples first; then run <code class="docutils literal notranslate"><span class="pre">king</span></code> to perform kinship estimate and finally <code class="docutils literal notranslate"><span class="pre">qc</span></code> to do addition QC</p></li>
<li><p>When you have a separate data-set for kinship estimate different from your genotype of interest, you can run <code class="docutils literal notranslate"><span class="pre">king</span></code>, followed by <code class="docutils literal notranslate"><span class="pre">qc</span></code>.</p></li>
</ol>
<p>In both cases, you should use the <code class="docutils literal notranslate"><span class="pre">*.related_remove</span></code> output from <code class="docutils literal notranslate"><span class="pre">king</span></code> as the <code class="docutils literal notranslate"><span class="pre">--remove_samples</span></code> parameter input for <code class="docutils literal notranslate"><span class="pre">qc</span></code> step.</p>
</div>
<div class="section" id="minimal-working-example">
<h2>Minimal working example<a class="headerlink" href="#minimal-working-example" title="Permalink to this headline">¶</a></h2>
<p>FIXME: first specify which of the 2 scenarios this example is for, then show how to run it.</p>
<div class="section" id="first-scenario-estimate-kinship">
<h3>First scenario: estimate kinship<a class="headerlink" href="#first-scenario-estimate-kinship" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">bioworkflows</span><span class="o">/</span><span class="n">GWAS</span><span class="o">/</span><span class="n">GWAS_QC</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">king</span>\
    <span class="o">--</span><span class="n">cwd</span> <span class="o">~/</span><span class="n">output</span> \
    <span class="o">--</span><span class="n">genoFile</span> <span class="o">~/</span><span class="n">MWE_AD</span><span class="o">/</span><span class="n">rename_chr22</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">name</span> <span class="n">first</span> \ 
    <span class="o">--</span><span class="n">kinship</span> <span class="mf">0.05</span>
</pre></div>
</div>
</div>
<div class="section" id="first-scenario-do-qc">
<h3>First scenario: do qc<a class="headerlink" href="#first-scenario-do-qc" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">bioworkflows</span><span class="o">/</span><span class="n">GWAS</span><span class="o">/</span><span class="n">GWAS_QC</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">qc</span>\
    <span class="o">--</span><span class="n">cwd</span> <span class="o">~/</span><span class="n">output</span> \
    <span class="o">--</span><span class="n">genoFile</span> <span class="o">~/</span><span class="n">MWE_AD</span><span class="o">/</span><span class="n">rename_chr22</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">remove_samples</span> <span class="o">~/</span><span class="n">output</span><span class="o">/</span><span class="n">rename_chr22</span><span class="o">.</span><span class="n">first</span><span class="o">.</span><span class="n">related_remove</span> \
    <span class="o">--</span><span class="n">maf_filter</span> <span class="mf">0.5</span> \
    <span class="o">--</span><span class="n">geno_filter</span> <span class="mf">0.2</span> \
    <span class="o">--</span><span class="n">mind_filter</span> <span class="mf">0.1</span> \
    <span class="o">--</span><span class="n">hwe_filter</span> <span class="mf">0.0</span> \
    <span class="o">--</span><span class="n">name</span> <span class="n">first</span> \
    <span class="o">--</span><span class="n">window</span> <span class="mi">50</span> \
    <span class="o">--</span><span class="n">shift</span> <span class="mi">10</span> \
    <span class="o">--</span><span class="n">r2</span> <span class="mf">0.5</span> 
</pre></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">GWAS_QC</span><span class="o">.</span><span class="n">ipynb</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>usage: sos run GWAS_QC.ipynb [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &quot;sos run -h&quot; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  king
  qc
  merge_plink

Global Workflow Options:
  --cwd VAL (as path, required)
                        the output directory for generated files
  --name VAL (as str, required)
                        A string to identify your analysis run
  --job-size 1 (as int)
                        For cluster jobs, number commands to run per job
  --walltime 5h
                        Wall clock time expected
  --mem 16G
                        Memory expected
  --numThreads 20 (as int)
                        Number of threads
  --container-lmm &#39;statisticalgenetics/lmm:1.8&#39;
                        Software container option

Sections
  king:                 Inference of relationships in the sample to remove
                        closely related individuals
    Workflow Options:
      --kinship 0.0625 (as float)
                        Filter based on kinship coefficient higher than a number
                        (e.g first degree 0.25, second degree 0.125, third
                        degree 0.0625)
      --genoFile VAL (as path, required)
                        Plink binary file
  qc_1:                 Filter SNPs and select individuals
    Workflow Options:
      --remove-samples . (as path)
                        The path to the file that contains the list of samples
                        to remove (format FID, IID)
      --keep-samples . (as path)
                        The path to the file that contains the list of samples
                        to keep (format FID, IID)
      --keep-variants . (as path)
                        The path to the file that contains the list of variants
                        to keep
      --maf-filter 0.01 (as float)
                        MAF filter to use
      --geno-filter 0.01 (as float)
                        Maximum missingess per-variant
      --mind-filter 0.02 (as float)
                        Maximum missingness per-sample
      --hwe-filter 5e-08 (as float)
                        HWE filter
      --genoFile  paths

                        Plink binary files
  qc_2:                 LD prunning and remove related individuals (both ind of
                        a pair)
    Workflow Options:
      --window 50 (as int)
                        Window size
      --shift 10 (as int)
                        Shift window every 10 snps
      --r2 0.1 (as float)
  qc_3:                 Merge all the .bed files into one bed file
    Workflow Options:
      --merged-prefix &#39;&#39;
  merge_plink:
    Workflow Options:
      --merged-prefix VAL (as str, required)
      --genoFile  paths
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1"># the output directory for generated files</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># A string to identify your analysis run</span>
<span class="kn">parameter:</span> <span class="kp">name</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">b</span><span class="si">}</span><span class="s2">&quot;</span>
<span class="c1"># For cluster jobs, number commands to run per job</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">1</span>
<span class="c1"># Wall clock time expected</span>
<span class="kn">parameter:</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s2">&quot;5h&quot;</span>
<span class="c1"># Memory expected</span>
<span class="kn">parameter:</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s2">&quot;16G&quot;</span>
<span class="c1"># Number of threads</span>
<span class="kn">parameter:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="mi">20</span>
<span class="c1"># Software container option</span>
<span class="kn">parameter:</span> <span class="n">container_lmm</span> <span class="o">=</span> <span class="s1">&#39;statisticalgenetics/lmm:2.4&#39;</span>
<span class="c1"># use this function to edit memory string for PLINK input</span>
<span class="kn">from</span> <span class="nn">sos.utils</span> <span class="kn">import</span> <span class="n">expand_size</span>
<span class="n">cwd</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">&quot;</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="estimate-kinship-in-the-sample">
<h2>Estimate kinship in the sample<a class="headerlink" href="#estimate-kinship-in-the-sample" title="Permalink to this headline">¶</a></h2>
<p>The output is a list of related individuals, as well as the kinship matrix</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># Inference of relationships in the sample to remove closely related individuals</span>
<span class="p">[</span><span class="n">king_1</span><span class="p">]</span>
<span class="c1"># Plink binary file</span>
<span class="kn">parameter:</span> <span class="n">genoFile</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">parameter:</span> <span class="n">kinship</span> <span class="o">=</span> <span class="mf">0.0625</span>
<span class="kn">input:</span> <span class="n">genoFile</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">.kin0&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container_lmm</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
    <span class="n">plink2</span> \
      <span class="o">--</span><span class="n">bfile</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">n</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">make</span><span class="o">-</span><span class="n">king</span><span class="o">-</span><span class="n">table</span> \
      <span class="o">--</span><span class="n">king</span><span class="o">-</span><span class="n">table</span><span class="o">-</span><span class="nb">filter</span> <span class="err">$</span><span class="p">{</span><span class="n">kinship</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">threads</span> <span class="err">$</span><span class="p">{</span><span class="n">numThreads</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">memory</span> <span class="err">$</span><span class="p">{</span><span class="nb">int</span><span class="p">(</span><span class="n">expand_size</span><span class="p">(</span><span class="kp">mem</span><span class="p">)</span> <span class="o">*</span> <span class="mf">0.9</span><span class="p">)</span><span class="o">/</span><span class="mf">1e6</span><span class="p">}</span>
      
<span class="p">[</span><span class="n">king_2</span><span class="p">]</span>
<span class="c1"># Filter based on kinship coefficient higher than a number (e.g first degree 0.25, second degree 0.125, third degree 0.0625)</span>
<span class="kn">parameter:</span> <span class="n">kinship</span> <span class="o">=</span> <span class="mf">0.0625</span>
<span class="c1"># If set to true, the unrelated individuals in a family will be kept without being reported. </span>
<span class="c1"># Otherwise (use `--no-maximize-unrelated`) the entire family will be removed</span>
<span class="kn">parameter:</span> <span class="n">maximize_unrelated</span> <span class="o">=</span> <span class="s2">&quot;FALSE&quot;</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.related_id&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">R:</span>  <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
    <span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">igraph</span><span class="p">)</span>
    <span class="c1"># Remove related individuals while keeping maximum number of individuals</span>
    <span class="c1"># this function is simplified from: </span>
    <span class="c1"># https://rdrr.io/cran/plinkQC/src/R/utils.R</span>
    <span class="c1">#&#39; @param relatedness [data.frame] containing pair-wise relatedness estimates</span>
    <span class="c1">#&#39; (in column [relatednessRelatedness]) for individual 1 (in column</span>
    <span class="c1">#&#39; [relatednessID1] and individual 2 (in column [relatednessID1]). Columns</span>
    <span class="c1">#&#39; relatednessID1, relatednessID2 and relatednessRelatedness have to present,</span>
    <span class="c1">#&#39; while additional columns such as family IDs can be present. Default column</span>
    <span class="c1">#&#39; names correspond to column names in output of plink --genome</span>
    <span class="c1">#&#39; (\url{https://www.cog-genomics.org/plink/1.9/ibd}). All original</span>
    <span class="c1">#&#39; columns for pair-wise highIBDTh fails will be returned in fail_IBD.</span>
    <span class="c1">#&#39; @param relatednessTh [double] Threshold for filtering related individuals.</span>
    <span class="c1">#&#39; Individuals, whose pair-wise relatedness estimates are greater than this</span>
    <span class="c1">#&#39; threshold are considered related.</span>
    <span class="n">relatednessFilter</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">relatedness</span><span class="p">,</span> 
                                  <span class="n">relatednessTh</span><span class="p">,</span>
                                  <span class="n">relatednessID1</span><span class="o">=</span><span class="s2">&quot;ID1&quot;</span><span class="p">,</span> 
                                  <span class="n">relatednessID2</span><span class="o">=</span><span class="s2">&quot;ID2&quot;</span><span class="p">,</span>
                                  <span class="n">relatednessRelatedness</span><span class="o">=</span><span class="s2">&quot;KINSHIP&quot;</span><span class="p">)</span> <span class="p">{</span>
        <span class="c1"># format data</span>
        <span class="k">if</span> <span class="p">(</span><span class="err">!</span><span class="p">(</span><span class="n">relatednessID1</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">names</span><span class="p">(</span><span class="n">relatedness</span><span class="p">)))</span> <span class="p">{</span>
            <span class="n">stop</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">&quot;Column&quot;</span><span class="p">,</span> <span class="n">relatednessID1</span><span class="p">,</span> <span class="s2">&quot;for relatedness not found!&quot;</span><span class="p">))</span>
        <span class="p">}</span>
        <span class="k">if</span> <span class="p">(</span><span class="err">!</span><span class="p">(</span><span class="n">relatednessID2</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">names</span><span class="p">(</span><span class="n">relatedness</span><span class="p">)))</span> <span class="p">{</span>
            <span class="n">stop</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">&quot;Column&quot;</span><span class="p">,</span> <span class="n">relatednessID1</span><span class="p">,</span> <span class="s2">&quot;for relatedness not found!&quot;</span><span class="p">))</span>
        <span class="p">}</span>
        <span class="k">if</span> <span class="p">(</span><span class="err">!</span><span class="p">(</span><span class="n">relatednessRelatedness</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">names</span><span class="p">(</span><span class="n">relatedness</span><span class="p">)))</span> <span class="p">{</span>
            <span class="n">stop</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="s2">&quot;Column&quot;</span><span class="p">,</span> <span class="n">relatednessRelatedness</span><span class="p">,</span>
                       <span class="s2">&quot;for relatedness not found!&quot;</span><span class="p">))</span>
        <span class="p">}</span>

        <span class="n">ID1_index</span> <span class="o">&lt;-</span> <span class="n">which</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">relatedness</span><span class="p">)</span> <span class="o">==</span> <span class="n">relatednessID1</span><span class="p">)</span>
        <span class="n">ID2_index</span> <span class="o">&lt;-</span> <span class="n">which</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">relatedness</span><span class="p">)</span> <span class="o">==</span> <span class="n">relatednessID2</span><span class="p">)</span>

        <span class="n">relatedness</span><span class="p">[,</span><span class="n">ID1_index</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">character</span><span class="p">(</span><span class="n">relatedness</span><span class="p">[,</span><span class="n">ID1_index</span><span class="p">])</span>
        <span class="n">relatedness</span><span class="p">[,</span><span class="n">ID2_index</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">character</span><span class="p">(</span><span class="n">relatedness</span><span class="p">[,</span><span class="n">ID2_index</span><span class="p">])</span>

        <span class="n">relatedness_names</span> <span class="o">&lt;-</span> <span class="n">names</span><span class="p">(</span><span class="n">relatedness</span><span class="p">)</span>
        <span class="n">names</span><span class="p">(</span><span class="n">relatedness</span><span class="p">)[</span><span class="n">ID1_index</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="s2">&quot;ID1&quot;</span>
        <span class="n">names</span><span class="p">(</span><span class="n">relatedness</span><span class="p">)[</span><span class="n">ID2_index</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="s2">&quot;ID2&quot;</span>
        <span class="n">names</span><span class="p">(</span><span class="n">relatedness</span><span class="p">)[</span><span class="n">names</span><span class="p">(</span><span class="n">relatedness</span><span class="p">)</span> <span class="o">==</span> <span class="n">relatednessRelatedness</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="s2">&quot;M&quot;</span>

        <span class="c1"># Remove symmetric IID rows</span>
        <span class="n">relatedness_original</span> <span class="o">&lt;-</span> <span class="n">relatedness</span>
        <span class="n">relatedness</span> <span class="o">&lt;-</span> <span class="n">dplyr</span><span class="p">::</span><span class="n">select</span><span class="p">(</span><span class="n">relatedness</span><span class="p">,</span> <span class="n">ID1</span><span class="p">,</span> <span class="n">ID2</span><span class="p">,</span> <span class="n">M</span><span class="p">)</span>

        <span class="n">sortedIDs</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">t</span><span class="p">(</span><span class="n">apply</span><span class="p">(</span><span class="n">relatedness</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="n">pair</span><span class="p">)</span> <span class="p">{</span>
            <span class="n">c</span><span class="p">(</span><span class="n">sort</span><span class="p">(</span><span class="n">c</span><span class="p">(</span><span class="n">pair</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">pair</span><span class="p">[</span><span class="mi">2</span><span class="p">])))</span>
            <span class="p">})),</span> <span class="n">stringsAsFactors</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
        <span class="n">keepIndex</span> <span class="o">&lt;-</span> <span class="n">which</span><span class="p">(</span><span class="err">!</span><span class="n">duplicated</span><span class="p">(</span><span class="n">sortedIDs</span><span class="p">))</span>

        <span class="n">relatedness_original</span> <span class="o">&lt;-</span> <span class="n">relatedness_original</span><span class="p">[</span><span class="n">keepIndex</span><span class="p">,]</span>
        <span class="n">relatedness</span> <span class="o">&lt;-</span> <span class="n">relatedness</span><span class="p">[</span><span class="n">keepIndex</span><span class="p">,]</span>

        <span class="c1"># individuals with at least one pair-wise comparison &gt; relatednessTh</span>
        <span class="c1"># return NULL to failIDs if no one fails the relatedness check</span>
        <span class="n">highRelated</span> <span class="o">&lt;-</span> <span class="n">dplyr</span><span class="p">::</span><span class="nb">filter</span><span class="p">(</span><span class="n">relatedness</span><span class="p">,</span> <span class="n">M</span> <span class="o">&gt;</span> <span class="n">relatednessTh</span><span class="p">)</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">nrow</span><span class="p">(</span><span class="n">highRelated</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="p">{</span>
            <span class="k">return</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">relatednessFails</span><span class="o">=</span><span class="n">NULL</span><span class="p">,</span> <span class="n">failIDs</span><span class="o">=</span><span class="n">NULL</span><span class="p">))</span>
        <span class="p">}</span>

        <span class="c1"># all samples with related individuals</span>
        <span class="n">allRelated</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span><span class="n">highRelated</span><span class="err">$</span><span class="n">ID1</span><span class="p">,</span> <span class="n">highRelated</span><span class="err">$</span><span class="n">ID2</span><span class="p">)</span>
        <span class="n">uniqueIIDs</span> <span class="o">&lt;-</span> <span class="n">unique</span><span class="p">(</span><span class="n">allRelated</span><span class="p">)</span>

        <span class="c1"># Further selection of samples with relatives in cohort</span>
        <span class="n">multipleRelative</span> <span class="o">&lt;-</span> <span class="n">unique</span><span class="p">(</span><span class="n">allRelated</span><span class="p">[</span><span class="n">duplicated</span><span class="p">(</span><span class="n">allRelated</span><span class="p">)])</span>
        <span class="n">singleRelative</span> <span class="o">&lt;-</span> <span class="n">uniqueIIDs</span><span class="p">[</span><span class="err">!</span><span class="n">uniqueIIDs</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">multipleRelative</span><span class="p">]</span>

        <span class="n">highRelatedMultiple</span> <span class="o">&lt;-</span> <span class="n">highRelated</span><span class="p">[</span><span class="n">highRelated</span><span class="err">$</span><span class="n">ID1</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">multipleRelative</span> <span class="o">|</span>
                                            <span class="n">highRelated</span><span class="err">$</span><span class="n">ID2</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">multipleRelative</span><span class="p">,]</span>
        <span class="n">highRelatedSingle</span> <span class="o">&lt;-</span> <span class="n">highRelated</span><span class="p">[</span><span class="n">highRelated</span><span class="err">$</span><span class="n">ID1</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">singleRelative</span> <span class="o">&amp;</span>
                                           <span class="n">highRelated</span><span class="err">$</span><span class="n">ID2</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">singleRelative</span><span class="p">,]</span>

        <span class="c1"># Only one related samples per individual</span>
        <span class="k">if</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">singleRelative</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)</span> <span class="p">{</span>
          <span class="c1"># randomly choose one to exclude</span>
          <span class="n">failIDs_single</span> <span class="o">&lt;-</span> <span class="n">highRelatedSingle</span><span class="p">[,</span><span class="mi">1</span><span class="p">]</span>
            
        <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
          <span class="n">failIDs_single</span> <span class="o">&lt;-</span> <span class="n">NULL</span>
        <span class="p">}</span>
  
        <span class="c1"># An individual has multiple relatives</span>
        <span class="k">if</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">multipleRelative</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">)</span> <span class="p">{</span>
            <span class="n">relatedPerID</span> <span class="o">&lt;-</span> <span class="n">lapply</span><span class="p">(</span><span class="n">multipleRelative</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="p">{</span>
                <span class="n">tmp</span> <span class="o">&lt;-</span> <span class="n">highRelatedMultiple</span><span class="p">[</span><span class="n">rowSums</span><span class="p">(</span>
                    <span class="n">cbind</span><span class="p">(</span><span class="n">highRelatedMultiple</span><span class="err">$</span><span class="n">ID1</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">x</span><span class="p">,</span>
                          <span class="n">highRelatedMultiple</span><span class="err">$</span><span class="n">ID2</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">x</span><span class="p">))</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span>
                <span class="n">rel</span> <span class="o">&lt;-</span> <span class="n">unique</span><span class="p">(</span><span class="n">unlist</span><span class="p">(</span><span class="n">tmp</span><span class="p">))</span>
                <span class="k">return</span><span class="p">(</span><span class="n">rel</span><span class="p">)</span>
            <span class="p">})</span>
            <span class="n">names</span><span class="p">(</span><span class="n">relatedPerID</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">multipleRelative</span>

            <span class="n">keepIDs_multiple</span> <span class="o">&lt;-</span> <span class="n">lapply</span><span class="p">(</span><span class="n">relatedPerID</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="p">{</span>
                <span class="n">pairwise</span> <span class="o">&lt;-</span> <span class="n">t</span><span class="p">(</span><span class="n">combn</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
                <span class="n">index</span> <span class="o">&lt;-</span> <span class="p">(</span><span class="n">highRelatedMultiple</span><span class="err">$</span><span class="n">ID1</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">pairwise</span><span class="p">[,</span><span class="mi">1</span><span class="p">]</span> <span class="o">&amp;</span>
                              <span class="n">highRelatedMultiple</span><span class="err">$</span><span class="n">ID2</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">pairwise</span><span class="p">[,</span><span class="mi">2</span><span class="p">])</span> <span class="o">|</span>
                    <span class="p">(</span><span class="n">highRelatedMultiple</span><span class="err">$</span><span class="n">ID1</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">pairwise</span><span class="p">[,</span><span class="mi">2</span><span class="p">]</span> <span class="o">&amp;</span>
                         <span class="n">highRelatedMultiple</span><span class="err">$</span><span class="n">ID2</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">pairwise</span><span class="p">[,</span><span class="mi">1</span><span class="p">])</span>
                <span class="n">combination</span> <span class="o">&lt;-</span> <span class="n">highRelatedMultiple</span><span class="p">[</span><span class="n">index</span><span class="p">,]</span>
                <span class="n">combination_graph</span> <span class="o">&lt;-</span> <span class="n">igraph</span><span class="p">::</span><span class="n">graph_from_data_frame</span><span class="p">(</span><span class="n">combination</span><span class="p">,</span>
                                                                   <span class="n">directed</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
                <span class="n">all_iv_set</span> <span class="o">&lt;-</span> <span class="n">igraph</span><span class="p">::</span><span class="n">ivs</span><span class="p">(</span><span class="n">combination_graph</span><span class="p">)</span>
                <span class="n">length_iv_set</span> <span class="o">&lt;-</span> <span class="n">sapply</span><span class="p">(</span><span class="n">all_iv_set</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="n">length</span><span class="p">(</span><span class="n">x</span><span class="p">))</span>

                <span class="k">if</span> <span class="p">(</span><span class="nb">all</span><span class="p">(</span><span class="n">length_iv_set</span> <span class="o">==</span> <span class="mi">1</span><span class="p">))</span> <span class="p">{</span>
                    <span class="c1"># check how often they occurr elsewhere</span>
                    <span class="n">occurrence</span> <span class="o">&lt;-</span> <span class="n">sapply</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="nb">id</span><span class="p">)</span> <span class="p">{</span>
                        <span class="nb">sum</span><span class="p">(</span><span class="n">sapply</span><span class="p">(</span><span class="n">relatedPerID</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="n">idlist</span><span class="p">)</span> <span class="nb">id</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">idlist</span><span class="p">))</span>
                    <span class="p">})</span>
                    <span class="c1"># if occurrence the same everywhere, pick the first, else keep</span>
                    <span class="c1"># the one with minimum occurrence elsewhere</span>
                    <span class="k">if</span> <span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="n">occurrence</span><span class="p">))</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span> <span class="p">{</span>
                        <span class="n">nonRelated</span> <span class="o">&lt;-</span> <span class="n">sort</span><span class="p">(</span><span class="n">x</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
                    <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
                        <span class="n">nonRelated</span> <span class="o">&lt;-</span> <span class="n">names</span><span class="p">(</span><span class="n">occurrence</span><span class="p">)[</span><span class="n">which</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">occurrence</span><span class="p">)]</span>
                    <span class="p">}</span>
                <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
                    <span class="n">nonRelated</span> <span class="o">&lt;-</span> <span class="n">all_iv_set</span><span class="p">[</span><span class="n">which</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">length_iv_set</span><span class="p">)]</span>
                <span class="p">}</span>
                <span class="k">return</span><span class="p">(</span><span class="n">nonRelated</span><span class="p">)</span>
            <span class="p">})</span>
            <span class="n">keepIDs_multiple</span> <span class="o">&lt;-</span> <span class="n">unique</span><span class="p">(</span><span class="n">unlist</span><span class="p">(</span><span class="n">keepIDs_multiple</span><span class="p">))</span>
            <span class="n">failIDs_multiple</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span><span class="n">multipleRelative</span><span class="p">[</span><span class="err">!</span><span class="n">multipleRelative</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span>
                                                       <span class="n">keepIDs_multiple</span><span class="p">])</span>
        <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
            <span class="n">failIDs_multiple</span> <span class="o">&lt;-</span> <span class="n">NULL</span>
        <span class="p">}</span>
        <span class="n">allFailIIDs</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span><span class="n">failIDs_single</span><span class="p">,</span> <span class="n">failIDs_multiple</span><span class="p">)</span>
        <span class="n">relatednessFails</span> <span class="o">&lt;-</span> <span class="n">lapply</span><span class="p">(</span><span class="n">allFailIIDs</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="nb">id</span><span class="p">)</span> <span class="p">{</span>
            <span class="n">fail_inorder</span> <span class="o">&lt;-</span> <span class="n">relatedness_original</span><span class="err">$</span><span class="n">ID1</span> <span class="o">==</span> <span class="nb">id</span> <span class="o">&amp;</span>
                <span class="n">relatedness_original</span><span class="err">$</span><span class="n">M</span> <span class="o">&gt;</span> <span class="n">relatednessTh</span>
            <span class="n">fail_inreverse</span> <span class="o">&lt;-</span> <span class="n">relatedness_original</span><span class="err">$</span><span class="n">ID2</span> <span class="o">==</span> <span class="nb">id</span> <span class="o">&amp;</span>
                <span class="n">relatedness_original</span><span class="err">$</span><span class="n">M</span> <span class="o">&gt;</span> <span class="n">relatednessTh</span>
            <span class="k">if</span> <span class="p">(</span><span class="nb">any</span><span class="p">(</span><span class="n">fail_inreverse</span><span class="p">))</span> <span class="p">{</span>
                <span class="n">inreverse</span> <span class="o">&lt;-</span> <span class="n">relatedness_original</span><span class="p">[</span><span class="n">fail_inreverse</span><span class="p">,</span> <span class="p">]</span>
                <span class="n">id1</span> <span class="o">&lt;-</span> <span class="n">ID1_index</span>
                <span class="n">id2</span> <span class="o">&lt;-</span> <span class="n">ID2_index</span>
                <span class="n">inreverse</span><span class="p">[,</span><span class="n">c</span><span class="p">(</span><span class="n">id1</span><span class="p">,</span> <span class="n">id2</span><span class="p">)]</span> <span class="o">&lt;-</span> <span class="n">inreverse</span><span class="p">[,</span><span class="n">c</span><span class="p">(</span><span class="n">id2</span><span class="p">,</span> <span class="n">id1</span><span class="p">)]</span>
                <span class="n">names</span><span class="p">(</span><span class="n">inreverse</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">relatedness_names</span>
            <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
                <span class="n">inreverse</span> <span class="o">&lt;-</span> <span class="n">NULL</span>
            <span class="p">}</span>
            <span class="n">inorder</span> <span class="o">&lt;-</span> <span class="n">relatedness_original</span><span class="p">[</span><span class="n">fail_inorder</span><span class="p">,</span> <span class="p">]</span>
            <span class="n">names</span><span class="p">(</span><span class="n">inorder</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">relatedness_names</span>
            <span class="k">return</span><span class="p">(</span><span class="n">rbind</span><span class="p">(</span><span class="n">inorder</span><span class="p">,</span> <span class="n">inreverse</span><span class="p">))</span>
        <span class="p">})</span>
        <span class="n">relatednessFails</span> <span class="o">&lt;-</span> <span class="n">do</span><span class="o">.</span><span class="n">call</span><span class="p">(</span><span class="n">rbind</span><span class="p">,</span> <span class="n">relatednessFails</span><span class="p">)</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">nrow</span><span class="p">(</span><span class="n">relatednessFails</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="p">{</span>
            <span class="n">relatednessFails</span> <span class="o">&lt;-</span> <span class="n">NULL</span>
            <span class="n">failIDs</span> <span class="o">&lt;-</span> <span class="n">NULL</span>
        <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
            <span class="n">names</span><span class="p">(</span><span class="n">relatednessFails</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">relatedness_names</span>
            <span class="n">rownames</span><span class="p">(</span><span class="n">relatednessFails</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="mi">1</span><span class="p">:</span><span class="n">nrow</span><span class="p">(</span><span class="n">relatednessFails</span><span class="p">)</span>
            <span class="n">uniqueFails</span> <span class="o">&lt;-</span> <span class="n">relatednessFails</span><span class="p">[</span><span class="err">!</span><span class="n">duplicated</span><span class="p">(</span><span class="n">relatednessFails</span><span class="p">[,</span><span class="n">ID1_index</span><span class="p">]),]</span>
            <span class="n">failIDs</span> <span class="o">&lt;-</span> <span class="n">uniqueFails</span><span class="p">[,</span><span class="n">ID1_index</span><span class="p">]</span>
        <span class="p">}</span>
        <span class="k">return</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">relatednessFails</span><span class="o">=</span><span class="n">relatednessFails</span><span class="p">,</span> <span class="n">failIDs</span><span class="o">=</span><span class="n">failIDs</span><span class="p">))</span>
    <span class="p">}</span>
    
  
    <span class="c1"># main code</span>
    <span class="n">kin0</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">},</span> <span class="n">header</span><span class="o">=</span><span class="n">F</span><span class="p">,</span> <span class="n">stringsAsFactor</span><span class="o">=</span><span class="n">F</span><span class="p">)</span>
    <span class="n">colnames</span><span class="p">(</span><span class="n">kin0</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;FID1&quot;</span><span class="p">,</span><span class="s2">&quot;ID1&quot;</span><span class="p">,</span><span class="s2">&quot;FID2&quot;</span><span class="p">,</span><span class="s2">&quot;ID2&quot;</span><span class="p">,</span><span class="s2">&quot;NSNP&quot;</span><span class="p">,</span><span class="s2">&quot;HETHET&quot;</span><span class="p">,</span><span class="s2">&quot;IBS0&quot;</span><span class="p">,</span><span class="s2">&quot;KINSHIP&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">maximize_unrelated</span><span class="p">})</span> <span class="p">{</span>
        <span class="n">rel</span> <span class="o">&lt;-</span> <span class="n">relatednessFilter</span><span class="p">(</span><span class="n">kin0</span><span class="p">,</span> <span class="err">$</span><span class="p">{</span><span class="n">kinship</span><span class="p">},</span> <span class="s2">&quot;ID1&quot;</span><span class="p">,</span> <span class="s2">&quot;ID2&quot;</span><span class="p">,</span> <span class="s2">&quot;KINSHIP&quot;</span><span class="p">)</span><span class="err">$</span><span class="n">failIDs</span>
        <span class="n">tmp1</span> <span class="o">&lt;-</span> <span class="n">kin0</span><span class="p">[,</span><span class="mi">1</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span>
        <span class="n">tmp2</span> <span class="o">&lt;-</span> <span class="n">kin0</span><span class="p">[,</span><span class="mi">3</span><span class="p">:</span><span class="mi">4</span><span class="p">]</span>
        <span class="n">colnames</span><span class="p">(</span><span class="n">tmp1</span><span class="p">)</span> <span class="o">=</span> <span class="n">colnames</span><span class="p">(</span><span class="n">tmp2</span><span class="p">)</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;FID&quot;</span><span class="p">,</span> <span class="s2">&quot;ID&quot;</span><span class="p">)</span>
        <span class="c1"># Get the family ID for these rels so there are two columns FID and IID in the output</span>
        <span class="n">lookup</span> <span class="o">&lt;-</span> <span class="n">dplyr</span><span class="p">::</span><span class="n">distinct</span><span class="p">(</span><span class="n">rbind</span><span class="p">(</span><span class="n">tmp1</span><span class="p">,</span><span class="n">tmp2</span><span class="p">))</span>
        <span class="n">dat</span> <span class="o">&lt;-</span> <span class="n">lookup</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">lookup</span><span class="p">[,</span><span class="mi">2</span><span class="p">]</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">rel</span><span class="p">),]</span>
    <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
        <span class="n">rel</span> <span class="o">&lt;-</span> <span class="n">kin0</span> <span class="o">%&gt;%</span> <span class="nb">filter</span><span class="p">(</span><span class="n">KINSHIP</span> <span class="o">&gt;=</span> <span class="err">$</span><span class="p">{</span><span class="n">kinship</span><span class="p">})</span>
        <span class="n">IID</span> <span class="o">&lt;-</span> <span class="n">sort</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="n">unlist</span><span class="p">(</span><span class="n">rel</span><span class="p">[,</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;ID1&quot;</span><span class="p">,</span> <span class="s2">&quot;ID2&quot;</span><span class="p">)])))</span>
        <span class="n">dat</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">IID</span><span class="p">)</span>
        <span class="n">dat</span> <span class="o">&lt;-</span> <span class="n">dat</span> <span class="o">%&gt;%</span>
            <span class="n">mutate</span><span class="p">(</span><span class="n">FID</span> <span class="o">=</span> <span class="n">IID</span><span class="p">)</span> <span class="o">%&gt;%</span>
            <span class="n">select</span><span class="p">(</span><span class="n">FID</span><span class="p">,</span> <span class="n">IID</span><span class="p">)</span>
    <span class="p">}</span>
    <span class="n">cat</span><span class="p">(</span><span class="s2">&quot;There are&quot;</span><span class="p">,</span> <span class="n">nrow</span><span class="p">(</span><span class="n">dat</span><span class="p">),</span><span class="s2">&quot;related individuals using a kinship threshold of $</span><span class="si">{kinship}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">write</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">dat</span><span class="p">,</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">r</span><span class="p">},</span> <span class="n">quote</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span> <span class="n">row</span><span class="o">.</span><span class="n">names</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span> <span class="n">col</span><span class="o">.</span><span class="n">names</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="genotype-and-sample-qc">
<h2>Genotype and sample QC<a class="headerlink" href="#genotype-and-sample-qc" title="Permalink to this headline">¶</a></h2>
<p>QC the genetic data based on MAF, sample and variant missigness and Hardy-Weinberg Equilibrium (HWE).</p>
<p>In this step you may also provide a list of samples to keep, for example in the case when you would like to subset a sample based on their ancestries to perform independent analyses on each of these groups.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># Filter SNPs and select individuals </span>
<span class="p">[</span><span class="n">qc_1</span><span class="p">,</span><span class="n">basic_filter</span><span class="p">]</span>
<span class="c1"># The path to the file that contains the list of samples to remove (format FID, IID)</span>
<span class="kn">parameter:</span> <span class="n">remove_samples</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)</span>
<span class="c1"># The path to the file that contains the list of samples to keep (format FID, IID)</span>
<span class="kn">parameter:</span> <span class="n">keep_samples</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)</span>
<span class="c1"># The path to the file that contains the list of variants to keep</span>
<span class="kn">parameter:</span> <span class="n">keep_variants</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)</span>
<span class="c1"># minimum MAF filter to use. Notice that PLINK default is 0.01</span>
<span class="kn">parameter:</span> <span class="n">maf_filter</span> <span class="o">=</span> <span class="mf">0.01</span>
<span class="c1"># maximum MAF filter to use</span>
<span class="kn">parameter:</span> <span class="n">maf_max_filter</span> <span class="o">=</span> <span class="mf">0.0</span>
<span class="c1"># Maximum missingess per-variant</span>
<span class="kn">parameter:</span> <span class="n">geno_filter</span> <span class="o">=</span> <span class="mf">0.01</span>
<span class="c1"># Maximum missingness per-sample</span>
<span class="kn">parameter:</span> <span class="n">mind_filter</span> <span class="o">=</span> <span class="mf">0.02</span>
<span class="c1"># HWE filter </span>
<span class="kn">parameter:</span> <span class="n">hwe_filter</span> <span class="o">=</span> <span class="mf">5e-08</span>
<span class="c1"># Plink binary files</span>
<span class="kn">parameter:</span> <span class="n">genoFile</span> <span class="o">=</span> <span class="n">paths</span>

<span class="n">fail_if</span><span class="p">(</span><span class="ow">not</span> <span class="p">(</span><span class="n">keep_samples</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="ow">or</span> <span class="n">keep_samples</span> <span class="o">==</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)),</span> <span class="n">msg</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;Cannot find ``</span><span class="si">{</span><span class="n">keep_samples</span><span class="si">}</span><span class="s1">``&#39;</span><span class="p">)</span>
<span class="n">fail_if</span><span class="p">(</span><span class="ow">not</span> <span class="p">(</span><span class="n">keep_variants</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="ow">or</span> <span class="n">keep_variants</span> <span class="o">==</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)),</span> <span class="n">msg</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;Cannot find ``</span><span class="si">{</span><span class="n">keep_variants</span><span class="si">}</span><span class="s1">``&#39;</span><span class="p">)</span>
<span class="n">fail_if</span><span class="p">(</span><span class="ow">not</span> <span class="p">(</span><span class="n">remove_samples</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="ow">or</span> <span class="n">remove_samples</span> <span class="o">==</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)),</span> <span class="n">msg</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;Cannot find ``</span><span class="si">{</span><span class="n">remove_samples</span><span class="si">}</span><span class="s1">``&#39;</span><span class="p">)</span>

<span class="kn">input:</span> <span class="n">genoFile</span><span class="p">,</span> <span class="n">group_by</span><span class="o">=</span><span class="mi">1</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">.filtered</span><span class="si">{</span><span class="s2">&quot;.extracted&quot;</span> <span class="k">if</span> <span class="n">keep_variants</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="si">}</span><span class="s1">.bed&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">bash:</span>  <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
    <span class="n">plink</span> \
      <span class="o">--</span><span class="n">bfile</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">n</span><span class="p">}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--maf </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">maf_filter</span><span class="p">)</span> <span class="k">if</span> <span class="n">maf_filter</span> <span class="o">&gt;=</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--max-maf </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">maf_max_filter</span><span class="p">)</span> <span class="k">if</span> <span class="n">maf_max_filter</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--geno </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">geno_filter</span><span class="p">)</span> <span class="k">if</span> <span class="n">geno_filter</span> <span class="o">&gt;=</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--hwe </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">hwe_filter</span><span class="p">)</span> <span class="k">if</span> <span class="n">hwe_filter</span> <span class="o">&gt;=</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--mind </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">mind_filter</span><span class="p">)</span> <span class="k">if</span> <span class="n">mind_filter</span> <span class="o">&gt;=</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;&#39;</span><span class="p">}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--keep </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">keep_samples</span><span class="p">)</span> <span class="k">if</span> <span class="n">keep_samples</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="p">}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--remove </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">remove_samples</span><span class="p">)</span> <span class="k">if</span> <span class="n">remove_samples</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="p">}</span> \
      <span class="err">$</span><span class="p">{(</span><span class="s1">&#39;--extract </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">keep_variants</span><span class="p">)</span> <span class="k">if</span> <span class="n">keep_variants</span><span class="o">.</span><span class="n">is_file</span><span class="p">()</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">make</span><span class="o">-</span><span class="n">bed</span> \
      <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">threads</span> <span class="err">$</span><span class="p">{</span><span class="n">numThreads</span><span class="p">}</span> \
      <span class="o">--</span><span class="n">memory</span> <span class="err">$</span><span class="p">{</span><span class="nb">int</span><span class="p">(</span><span class="n">expand_size</span><span class="p">(</span><span class="kp">mem</span><span class="p">)</span> <span class="o">*</span> <span class="mf">0.9</span><span class="p">)</span><span class="o">/</span><span class="mf">1e6</span><span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># LD prunning and remove related individuals (both ind of a pair)</span>
<span class="p">[</span><span class="n">qc_2</span><span class="p">,</span><span class="n">LD_pruning</span><span class="p">]</span>
<span class="c1"># Window size</span>
<span class="kn">parameter:</span> <span class="n">window</span> <span class="o">=</span> <span class="mi">50</span>
<span class="c1"># Shift window every 10 snps</span>
<span class="kn">parameter:</span> <span class="n">shift</span> <span class="o">=</span> <span class="mi">10</span>
<span class="kn">parameter:</span> <span class="n">r2</span> <span class="o">=</span> <span class="mf">0.1</span>
<span class="kn">output:</span> <span class="n">bed</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.prune.bed&#39;</span><span class="p">,</span> <span class="n">prune</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.prune.in&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="n">container_lmm</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
    <span class="n">plink2</span> \
    <span class="o">--</span><span class="n">bfile</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">n</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">indep</span><span class="o">-</span><span class="n">pairwise</span> <span class="err">$</span><span class="p">{</span><span class="n">window</span><span class="p">}</span> <span class="err">$</span><span class="p">{</span><span class="n">shift</span><span class="p">}</span> <span class="err">$</span><span class="p">{</span><span class="n">r2</span><span class="p">}</span>  \
    <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="s2">&quot;prune&quot;</span><span class="p">]:</span><span class="n">nn</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">threads</span> <span class="err">$</span><span class="p">{</span><span class="n">numThreads</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">memory</span> <span class="err">$</span><span class="p">{</span><span class="nb">int</span><span class="p">(</span><span class="n">expand_size</span><span class="p">(</span><span class="kp">mem</span><span class="p">)</span> <span class="o">*</span> <span class="mf">0.9</span><span class="p">)</span><span class="o">/</span><span class="mf">1e6</span><span class="p">}</span>
   
    <span class="n">plink2</span> \
    <span class="o">--</span><span class="n">bfile</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">n</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">extract</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="s1">&#39;prune&#39;</span><span class="p">]}</span> \
    <span class="o">--</span><span class="n">make</span><span class="o">-</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">[</span><span class="s1">&#39;bed&#39;</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">threads</span> <span class="err">$</span><span class="p">{</span><span class="n">numThreads</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">memory</span> <span class="err">$</span><span class="p">{</span><span class="nb">int</span><span class="p">(</span><span class="n">expand_size</span><span class="p">(</span><span class="kp">mem</span><span class="p">)</span> <span class="o">*</span> <span class="mf">0.9</span><span class="p">)</span><span class="o">/</span><span class="mf">1e6</span><span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># Merge all the .bed files into one bed file </span>
<span class="p">[</span><span class="n">qc_3</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">merged_name</span> <span class="o">=</span> <span class="nb">str</span>
<span class="kn">parameter:</span> <span class="n">merged_prefix</span> <span class="o">=</span> <span class="n">merged_name</span>
<span class="kn">input:</span> <span class="n">named_output</span><span class="p">(</span><span class="s2">&quot;bed&quot;</span><span class="p">)</span> <span class="p">,</span> <span class="n">group_by</span> <span class="o">=</span> <span class="s1">&#39;all&#39;</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">merged_prefix</span><span class="si">}</span><span class="s2">.bed&quot;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">bash:</span>  <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
    <span class="n">echo</span> <span class="o">-</span><span class="n">e</span> <span class="err">$</span><span class="p">{</span><span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)[:</span><span class="o">-</span><span class="mi">4</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">:]])}</span> <span class="o">|</span> <span class="n">sed</span> <span class="s1">&#39;s/ /</span><span class="se">\n</span><span class="s1">/g&#39;</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span><span class="o">.</span><span class="n">merge_list</span>
    <span class="n">plink</span> \
    <span class="o">--</span><span class="n">bfile</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">merge</span><span class="o">-</span><span class="nb">list</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span><span class="o">.</span><span class="n">merge_list</span> \
    <span class="o">--</span><span class="n">make</span><span class="o">-</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">threads</span> <span class="err">$</span><span class="p">{</span><span class="n">numThreads</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">memory</span> <span class="err">$</span><span class="p">{</span><span class="nb">int</span><span class="p">(</span><span class="n">expand_size</span><span class="p">(</span><span class="kp">mem</span><span class="p">)</span> <span class="o">*</span> <span class="mf">0.9</span><span class="p">)</span><span class="o">/</span><span class="mf">1e06</span><span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">merge_plink</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">merged_prefix</span> <span class="o">=</span> <span class="nb">str</span>
<span class="kn">parameter:</span> <span class="n">genoFile</span> <span class="o">=</span> <span class="n">paths</span>
<span class="n">skip_if</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">genoFile</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span>
<span class="kn">input:</span> <span class="n">genoFile</span><span class="p">,</span> <span class="n">group_by</span> <span class="o">=</span> <span class="s1">&#39;all&#39;</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">merged_prefix</span><span class="si">}</span><span class="s2">.bed&quot;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">bash:</span>  <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
    <span class="n">echo</span> <span class="o">-</span><span class="n">e</span> <span class="err">$</span><span class="p">{</span><span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)[:</span><span class="o">-</span><span class="mi">4</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">:]])}</span> <span class="o">|</span> <span class="n">sed</span> <span class="s1">&#39;s/ /</span><span class="se">\n</span><span class="s1">/g&#39;</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span><span class="o">.</span><span class="n">merge_list</span>
    <span class="n">plink</span> \
    <span class="o">--</span><span class="n">bfile</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">n</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">merge</span><span class="o">-</span><span class="nb">list</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span><span class="o">.</span><span class="n">merge_list</span> \
    <span class="o">--</span><span class="n">make</span><span class="o">-</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">threads</span> <span class="err">$</span><span class="p">{</span><span class="n">numThreads</span><span class="p">}</span> \
    <span class="o">--</span><span class="n">memory</span> <span class="err">$</span><span class="p">{</span><span class="nb">int</span><span class="p">(</span><span class="n">expand_size</span><span class="p">(</span><span class="kp">mem</span><span class="p">)</span> <span class="o">*</span> <span class="mf">0.9</span><span class="p">)</span><span class="o">/</span><span class="mf">1e06</span><span class="p">}</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            kernelName: "sos",
            path: "./workflow/data_preprocessing/genotype"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="../genotype_preprocessing.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Genotype data preprocessing</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="GRM.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Genomic relationship matrices</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics Consortium<br/>
        
            &copy; Copyright 2021, ADSP-FG xQTL methods and data integration subgroup.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>