
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Genotype VCF file quality control &#8212; ADSP-FG Project</title>
    
  <link href="../../../_static/css/theme.css" rel="stylesheet">
  <link href="../../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/togglebutton.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Genotype PLINK file quality control" href="GWAS_QC.html" />
    <link rel="prev" title="Genotype data preprocessing" href="../genotype_preprocessing.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">ADSP-FG Project</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../../README.html">
   xQTL analysis pipeline
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Complete Pipelines
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../complete_analysis/eQTL_analysis.html">
   Bulk RNA-seq eQTL analysis
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference data
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../reference_data.html">
   Reference data standardization
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/bulk_expression.html">
   RNA-seq calling and QC
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/RNA_calling.html">
     RNA-seq Calling
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_QC.html">
     Sample level RNA-seq quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_normalization.html">
     Bulk RNA-seq counts normalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../molecular_phenotypes/scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../molecular_phenotypes/calling/apa_calling.html">
   APA Calling
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="../genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     Genotype VCF file quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="GWAS_QC.html">
     Genotype PLINK file quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="PCA.html">
     Principal component analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="GRM.html">
     Genomic relationship matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="genotype_formatting.html">
     Genotype data formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../phenotype/gene_annotation.html">
     Gene coordinate annotation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../phenotype/phenotype_formatting.html">
     Phenotype data formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../covariate_preprocessing.html">
   Covariate data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../covariate/PEER_factor.html">
     Factor analysis using PEER
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../covariate/BiCV_factor.html">
     Factor analysis using Bi-Cross validation
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../association_scan/cisQTL_scan.html">
   cisQTL analysis workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL calling pipeline
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/APEX/APEX.html">
     APEX QTL calling pipeline
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../association_scan/transQTL_scan.html">
   transQTL analysis workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../multivariate/METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/METAL/METAL.html">
     Meta-analysis with METAL
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../multivariate/MASH_pipeline.html">
   Multivariate association with MASH
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/fastqtl_to_mash.html">
     eQTL summary statistics formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/mixture_prior.html">
     A multivariate EBNM approach for mixture multivariate distribution estimate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/mashr.html">
     MASH analysis pipeline with data-driven prior matrices
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Statistical Fine-mapping
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../fine_mapping/univariate_fine_mapping.html">
   Univariate fine-mapping workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE_RSS.html">
     Fine-mapping with SuSiE RSS model
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE.html">
     Fine-mapping with SuSiE model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../fine_mapping/multivariate_fine_mapping.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multi-omics integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../integrative_analysis/colocalization.html">
   QTL and GWAS Colocalization
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/COLOC/fastenloc.html">
     Colocalization with FastENLOC
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/COLOC/mvenloc.html">
     Multivariate SuSiE and ENLOC model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../integrative_analysis/functional_annotation.html">
   Functional annotation integration
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/enrichment/gwas_enrichment.html">
     Enrichment analysis: TORUS and GREGOR
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/enrichment/ldsc.html">
     Stratified LD Score Regression
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/SuSiE_Ann/polyfun.html">
     Fine-mapping with PolyFun
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Utilities
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../misc/summary_stats_merger.html">
   Summary statistics merger
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../misc/rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH/NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../../_sources/code/data_preprocessing/genotype/VCF_QC.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/data_preprocessing/genotype/VCF_QC.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/data_preprocessing/genotype/VCF_QC.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#overview">
   Overview
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#multi-allelic-sites">
     Multi-allelic sites
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#default-vcf-qc-filters">
   Default VCF QC filters
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#a-note-on-ts-tv-summary-from-vcf-genotype-data">
   A note on TS/TV summary from VCF genotype data
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#input">
   Input
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#output">
   Output
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#minimal-working-example">
   Minimal working example
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#command-interface">
   Command Interface
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#global-parameters">
   Global parameters
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#annotation-of-known-and-novel-variants">
   Annotation of known and novel variants
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#genotype-qc">
   Genotype QC
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="genotype-vcf-file-quality-control">
<h1>Genotype VCF file quality control<a class="headerlink" href="#genotype-vcf-file-quality-control" title="Permalink to this headline">¶</a></h1>
<p>This implements some recommendations from UK Biobank on <a class="reference external" href="https://www.medrxiv.org/content/10.1101/2020.11.02.20222232v1.full-text">exome sequence data quality control</a>.</p>
<div class="section" id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>The goal of this module is to perform QC on VCF files, including</p>
<ol class="simple">
<li><p>Handling the formatting of multi-allelic sites,</p></li>
<li><p>Genotype and variant level filtering based on genotype calling qualities.</p></li>
<li><p>Known/novel variants annotation</p></li>
<li><p>Summary statistics before and after QC, in particular the ts/tv ratio, to assess the effectiveness of QC.</p></li>
</ol>
<p>3 and 4 above are for explorative analysis on the overall quality assessment of genotype data in the VCF files. We annotate known and novel variants because ts/tv are expected to be different between known and novel variants, and is important QC metric to assess the effectiveness of our QC.</p>
<div class="section" id="multi-allelic-sites">
<h3>Multi-allelic sites<a class="headerlink" href="#multi-allelic-sites" title="Permalink to this headline">¶</a></h3>
<p>Mult-allelic sites can be problematic in many ways for downstreams analysis, even of they are handled in terms of formatting after QC. We provide an optional workflow module to keep only bi-allelic sites from data, although by default we will include these sites in the VCF file we generate.</p>
</div>
</div>
<div class="section" id="default-vcf-qc-filters">
<h2>Default VCF QC filters<a class="headerlink" href="#default-vcf-qc-filters" title="Permalink to this headline">¶</a></h2>
<ol class="simple">
<li><p>Genotype depth filters: SNPs DP&gt;10 and Indels DP&gt;10 for indels.</p></li>
<li><p>At least one sample per site passed the allele balance threshold &gt;= 0.15 for SNPs and &gt;=0.20 for indels (heterozygous variants).</p>
<ul class="simple">
<li><p>Allele balance is calculated for heterozygotes as the number of bases supporting the least-represented allele over the total number of base observations.</p></li>
</ul>
</li>
<li><p>Genotype quality GQ&gt;20.</p></li>
</ol>
<p>Filtering are done with <code class="docutils literal notranslate"><span class="pre">bcftools</span></code>. Here is a <a class="reference external" href="https://gist.github.com/elowy01/93922762e131d7abd3c7e8e166a74a0b">useful cheatsheet from github user &#64;elowy01</a>.</p>
</div>
<div class="section" id="a-note-on-ts-tv-summary-from-vcf-genotype-data">
<h2>A note on TS/TV summary from VCF genotype data<a class="headerlink" href="#a-note-on-ts-tv-summary-from-vcf-genotype-data" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">bcftools</span> <span class="pre">stats</span></code> command provides useful summary statistics including TS/TV ratio, which is routinely used as a quality measure of variant calls. With dbSNP based annotation of novel and known variants, <code class="docutils literal notranslate"><span class="pre">bcftools</span></code> can compute TS/TV for novel and known variants at variant level, and at sample level. It should be noted that variant level TS/TV does not take sample genotype into consideration – it simply counts the TS and TV event for observed SNPs in the data. Other tools, such as <code class="docutils literal notranslate"><span class="pre">snpsift</span></code>, implements variant level TS/TV by counting TS and TV events in sample genotypes and compute the ratio after summing up TS and TV across all samples. See <a class="reference external" href="https://github.com/samtools/bcftools/issues/1526">here</a> some discussions on this issue. We provide these TS/TV calculations before and after QC but users should be aware of the difference when interpreting the results.</p>
</div>
<div class="section" id="input">
<h2>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h2>
<ol class="simple">
<li><p>The target <code class="docutils literal notranslate"><span class="pre">VCF</span></code> file</p>
<ul class="simple">
<li><p>If its chromosome name does not have the <code class="docutils literal notranslate"><span class="pre">chr</span></code> prefix and you need it to match with reference <code class="docutils literal notranslate"><span class="pre">fasta</span></code> file, please run <code class="docutils literal notranslate"><span class="pre">rename_chrs</span></code> workflow to add <code class="docutils literal notranslate"><span class="pre">chr</span></code>.</p></li>
</ul>
</li>
<li><p>dbSNP database in <code class="docutils literal notranslate"><span class="pre">VCF</span></code> format</p></li>
<li><p>A reference sequence <code class="docutils literal notranslate"><span class="pre">fasta</span></code> file</p></li>
</ol>
</div>
<div class="section" id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h2>
<ol class="simple">
<li><p>QC-ed genotype data in VCF and in PLINK format</p></li>
<li><p>A set of sumstats to help evaluate quality of genotype before and after QC</p>
<ul class="simple">
<li><p>Particularly useful is the TS/TV ratio</p></li>
</ul>
</li>
</ol>
</div>
<div class="section" id="minimal-working-example">
<h2>Minimal working example<a class="headerlink" href="#minimal-working-example" title="Permalink to this headline">¶</a></h2>
<p>The MWE is generated via</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">bcftools</span> <span class="n">query</span> <span class="o">-</span><span class="n">l</span> <span class="n">get</span><span class="o">-</span><span class="n">dosage</span><span class="o">.</span><span class="n">ALL</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> <span class="o">|</span> <span class="n">head</span> <span class="o">-</span><span class="mi">40</span> <span class="o">&gt;</span> <span class="n">MWE_sample_list</span>
<span class="n">bcftools</span> <span class="n">view</span> <span class="o">-</span><span class="n">S</span> <span class="n">MWE_sample_list</span>  <span class="n">get</span><span class="o">-</span><span class="n">dosage</span><span class="o">.</span><span class="n">ALL</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> <span class="o">&gt;</span> <span class="n">sample_filtered</span><span class="o">.</span><span class="n">vcf</span> <span class="o">&amp;</span>
<span class="n">bgzip</span> <span class="o">-</span><span class="n">c</span> <span class="n">sample_filtered</span><span class="o">.</span><span class="n">vcf</span> <span class="o">&gt;</span>  <span class="n">sample_filtered</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span>
<span class="n">tabix</span> <span class="o">-</span><span class="n">p</span> <span class="n">vcf</span> <span class="n">sample_filtered</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span>
<span class="n">bcftools</span> <span class="n">view</span> <span class="o">--</span><span class="n">regions</span> <span class="n">chr1</span> <span class="n">sample_filtered</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> <span class="o">&gt;</span> <span class="n">chr1_sample_filtered</span><span class="o">.</span><span class="n">vcf</span> <span class="o">&amp;</span>
<span class="n">cat</span> <span class="n">chr1_sample_filtered</span><span class="o">.</span><span class="n">vcf</span> <span class="o">|</span> <span class="n">head</span> <span class="o">-</span><span class="mi">20000</span> <span class="o">&gt;</span> <span class="n">MWE_genotype</span><span class="o">.</span><span class="n">vcf</span>
</pre></div>
</div>
<p>and was stored here: <a class="reference external" href="https://drive.google.com/file/d/1sxxPdPIyKma0mAl8TKwhgyRHlOh0Oyrc/view?usp=sharing">https://drive.google.com/file/d/1sxxPdPIyKma0mAl8TKwhgyRHlOh0Oyrc/view?usp=sharing</a></p>
<p>The MWE was used as follows:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">VCF_QC</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">rename_chrs</span> \
    <span class="o">--</span><span class="n">genoFile</span> <span class="n">reference_data</span><span class="o">/</span><span class="mi">00</span><span class="o">-</span><span class="n">All</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">reference_data</span> <span class="o">--</span><span class="n">container</span> <span class="o">./</span><span class="n">bioinfo</span><span class="o">.</span><span class="n">sif</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">VCF_QC</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">dbsnp_annotate</span> \
    <span class="o">--</span><span class="n">genoFile</span> <span class="n">reference_data</span><span class="o">/</span><span class="mi">00</span><span class="o">-</span><span class="n">All</span><span class="o">.</span><span class="n">add_chr</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">reference_data</span> <span class="o">--</span><span class="n">container</span> <span class="o">./</span><span class="n">bioinfo</span><span class="o">.</span><span class="n">sif</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">VCF_QC</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">qc</span>    \
<span class="o">--</span><span class="n">genoFile</span> <span class="n">data</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">MWE_genotype</span><span class="o">.</span><span class="n">vcf</span>     \
<span class="o">--</span><span class="n">dbsnp</span><span class="o">-</span><span class="n">variants</span> <span class="n">data</span><span class="o">/</span><span class="n">reference_data</span><span class="o">/</span><span class="mi">00</span><span class="o">-</span><span class="n">All</span><span class="o">.</span><span class="n">add_chr</span><span class="o">.</span><span class="n">variants</span><span class="o">.</span><span class="n">gz</span>  \
<span class="o">--</span><span class="n">reference</span><span class="o">-</span><span class="n">genome</span> <span class="n">data</span><span class="o">/</span><span class="n">reference_data</span><span class="o">/</span><span class="n">GRCh38_full_analysis_set_plus_decoy_hla</span><span class="o">.</span><span class="n">noALT_noHLA_noDecoy_ERCC</span><span class="o">.</span><span class="n">fasta</span>   \
<span class="o">--</span><span class="n">cwd</span> <span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">genotype_1</span> <span class="o">--</span><span class="n">container</span> <span class="o">./</span><span class="n">bioinfo</span><span class="o">.</span><span class="n">sif</span> <span class="o">-</span><span class="n">J</span> <span class="mi">1</span> <span class="o">-</span><span class="n">c</span> <span class="n">csg</span><span class="o">.</span><span class="n">yml</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span>  <span class="o">&amp;</span>
</pre></div>
</div>
<p>To produce the following results:</p>
<ul class="simple">
<li><p>Total TS/TV for 19639 known variants before QC: 2.599</p></li>
<li><p>Total TS/TV for 19573 known variants after QC: 2.600</p></li>
<li><p>There is no novel variants included in the MWE.</p></li>
</ul>
<p>The Total TS/TV is extracted from the last step of QC. For known variant before QC:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>grep Ts/Tv MWE_genotype.leftnorm.known_variant.snipsift_tstv <span class="p">|</span> rev <span class="p">|</span> cut -d<span class="s1">&#39;,&#39;</span> -f1 <span class="p">|</span> rev
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2.599
</pre></div>
</div>
</div>
</div>
<p>For known variant after QC:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>grep Ts/Tv MWE_genotype.leftnorm.filtered.*_variant.snipsift_tstv <span class="p">|</span> rev <span class="p">|</span> cut -d<span class="s1">&#39;,&#39;</span> -f1 <span class="p">|</span> rev
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2.600
</pre></div>
</div>
</div>
</div>
<p>For novel variant before/after QC, TS/TV is not avaible since no novel_variants presented in the MWE</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>grep Ts/Tv MWE_genotype.leftnorm.novel_variant.snipsift_tstv <span class="p">|</span> rev <span class="p">|</span> cut -d<span class="s1">&#39;,&#39;</span> -f1 <span class="p">|</span> rev
grep Ts/Tv MWE_genotype.leftnorm.filtered.novel_variant.snipsift_tstv <span class="p">|</span> rev <span class="p">|</span> cut -d<span class="s1">&#39;,&#39;</span> -f1 <span class="p">|</span> rev
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="command-interface">
<h2>Command Interface<a class="headerlink" href="#command-interface" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sos run VCF_QC.ipynb -h
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>usage: sos run VCF_QC.ipynb [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &quot;sos run -h&quot; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  rename_chrs
  dbsnp_annotate
  qc

Global Workflow Options:
  --genoFile VAL (as path, required)
                        input
  --cwd VAL (as path, required)
                        Workdir
  --numThreads 1 (as int)
                        Number of threads
  --job-size 1 (as int)
                        For cluster jobs, number commands to run per job
  --walltime 5h
                        Walltime
  --mem 60G
  --container &#39;&#39;
                        Software container option

Sections
  rename_chrs:
  dbsnp_annotate:
  qc_1:                 Handel multi-allelic sites, left normalization of indels
                        and add variant ID
    Workflow Options:
      --dbsnp-variants VAL (as path, required)
                        Path to dbSNP variants generated previously
      --reference-genome VAL (as path, required)
                        Path to fasta file for HG reference genome, eg
                        GRCh38_full_analysis_set_plus_decoy_hla.fa
      --[no-]bi-allelic (default to False)
      --[no-]snp-only (default to False)
  qc_2:                 genotype QC
    Workflow Options:
      --geno-filter 0.1 (as float)
                        Maximum missingess per-variant
      --DP-snp 10 (as int)
                        Sample level QC - read depth (DP) to filter out SNPs
                        below this value
      --GQ 20 (as int)
                        Sample level QC - genotype quality (GQ) of specific
                        sample. This measure tells you how confident we are that
                        the genotype we assigned to a particular sample is
                        correct
      --DP-indel 10 (as int)
                        Sample level QC - read depth (DP) to filter out indels
                        below this value
      --AB-snp 0.15 (as float)
                        Allele balance for snps
      --AB-indel 0.2 (as float)
                        Allele balance for indels
      --hwe-filter 1e-06 (as float)
                        HWE filter
  qc_3:
  qc_4:
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="global-parameters">
<h2>Global parameters<a class="headerlink" href="#global-parameters" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="o">[</span>global<span class="o">]</span>
<span class="c1"># input</span>
parameter: <span class="nv">genoFile</span> <span class="o">=</span> path
<span class="c1"># Workdir</span>
parameter: <span class="nv">cwd</span> <span class="o">=</span> path
<span class="c1"># Number of threads</span>
parameter: <span class="nv">numThreads</span> <span class="o">=</span> <span class="m">1</span>
<span class="c1"># For cluster jobs, number commands to run per job</span>
parameter: <span class="nv">job_size</span> <span class="o">=</span> <span class="m">1</span>
<span class="c1"># Walltime </span>
parameter: <span class="nv">walltime</span> <span class="o">=</span> <span class="s1">&#39;5h&#39;</span>
parameter: <span class="nv">mem</span> <span class="o">=</span> <span class="s1">&#39;60G&#39;</span>
<span class="c1"># Software container option</span>
parameter: <span class="nv">container</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="c1"># use this function to edit memory string for PLINK input</span>
from sos.utils import expand_size
<span class="nv">cwd</span> <span class="o">=</span> path<span class="o">(</span>f<span class="s2">&quot;{cwd:a}&quot;</span><span class="o">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="annotation-of-known-and-novel-variants">
<h2>Annotation of known and novel variants<a class="headerlink" href="#annotation-of-known-and-novel-variants" title="Permalink to this headline">¶</a></h2>
<p>The known variant reference can be downloaded from <a class="reference external" href="https://ftp.ncbi.nlm.nih.gov/snp/organisms//human_9606_b150_GRCh38p7/VCF/00-All.vcf.gz">https://ftp.ncbi.nlm.nih.gov/snp/organisms//human_9606_b150_GRCh38p7/VCF/00-All.vcf.gz</a>.</p>
<p>The procedure/rationale is <a class="reference external" href="https://hbctraining.github.io/In-depth-NGS-Data-Analysis-Course/sessionVI/lessons/03_annotation-snpeff.html">explained in this post</a>.</p>
<p>It takes ~1hr for <code class="docutils literal notranslate"><span class="pre">rename_chrs</span></code> to complete. For every hour it can process ~13Gb of data, please set the –walltime parameter according to the size of your input files.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="o">[</span>rename_chrs<span class="o">]</span>
parameter: <span class="nv">walltime</span> <span class="o">=</span> <span class="s1">&#39;24h&#39;</span>
<span class="c1"># This file can be downloaded from https://ftp.ncbi.nlm.nih.gov/snp/organisms//human_9606_b150_GRCh38p7/VCF/00-All.vcf.gz.</span>
input: genoFile
output: f<span class="s2">&quot;{cwd}/{_input:bnn}.add_chr.vcf.gz&quot;</span>
task: <span class="nv">trunk_workers</span> <span class="o">=</span> <span class="m">1</span>, <span class="nv">trunk_size</span> <span class="o">=</span> job_size, <span class="nv">walltime</span> <span class="o">=</span> walltime, <span class="nv">mem</span> <span class="o">=</span> mem, <span class="nv">cores</span> <span class="o">=</span> numThreads, <span class="nv">tags</span> <span class="o">=</span> f<span class="s1">&#39;{step_name}_{_output:bn}&#39;</span>
bash: <span class="nv">container</span> <span class="o">=</span> container, <span class="nv">expand</span><span class="o">=</span> <span class="s2">&quot;</span><span class="si">${</span><span class="p"> </span><span class="si">}</span><span class="s2">&quot;</span>, <span class="nv">stderr</span> <span class="o">=</span> f<span class="s1">&#39;{_output:nn}.stderr&#39;</span>, <span class="nv">stdout</span> <span class="o">=</span> f<span class="s1">&#39;{_output:nn}.stdout&#39;</span>
    <span class="k">for</span> i <span class="k">in</span> <span class="o">{</span><span class="m">1</span>..22<span class="o">}</span> X Y MT<span class="p">;</span> <span class="k">do</span> <span class="nb">echo</span> <span class="s2">&quot;</span><span class="nv">$i</span><span class="s2"> chr</span><span class="nv">$i</span><span class="s2">&quot;</span><span class="p">;</span> <span class="k">done</span> &gt; <span class="si">${</span><span class="nv">_output</span><span class="p">:</span><span class="nv">nn</span><span class="si">}</span>.chr_name_conv.txt
    bcftools annotate --rename-chrs <span class="si">${</span><span class="nv">_output</span><span class="p">:</span><span class="nv">nn</span><span class="si">}</span>.chr_name_conv.txt <span class="si">${</span><span class="nv">_input</span><span class="si">}</span> -Oz -o <span class="si">${</span><span class="nv">_output</span><span class="si">}</span>
    tabix -p vcf <span class="si">${</span><span class="nv">_output</span><span class="si">}</span>
    rm -f <span class="si">${</span><span class="nv">_output</span><span class="p">:</span><span class="nv">nn</span><span class="si">}</span>.chr_name_conv.txt
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="o">[</span>dbsnp_annotate<span class="o">]</span>
output: f<span class="s2">&quot;{_input:nn}.variants.gz&quot;</span>
task: <span class="nv">trunk_workers</span> <span class="o">=</span> <span class="m">1</span>, <span class="nv">trunk_size</span><span class="o">=</span><span class="m">5</span>, <span class="nv">walltime</span> <span class="o">=</span> walltime, <span class="nv">mem</span> <span class="o">=</span> mem, <span class="nv">cores</span> <span class="o">=</span> numThreads, <span class="nv">tags</span> <span class="o">=</span> f<span class="s1">&#39;{step_name}_{_output:bn}&#39;</span>
bash: <span class="nv">container</span> <span class="o">=</span> container, <span class="nv">expand</span><span class="o">=</span> <span class="s2">&quot;</span><span class="si">${</span><span class="p"> </span><span class="si">}</span><span class="s2">&quot;</span>, <span class="nv">stderr</span> <span class="o">=</span> f<span class="s1">&#39;{_output:n}.stderr&#39;</span>, <span class="nv">stdout</span> <span class="o">=</span> f<span class="s1">&#39;{_output:n}.stdout&#39;</span>
    bcftools query  -f<span class="s1">&#39;%CHROM\t%POS\t%ID\t%REF\t%ALT\n&#39;</span> <span class="si">${</span><span class="nv">_input</span><span class="si">}</span>  <span class="p">|</span> <span class="se">\</span>
        awk <span class="s1">&#39;BEGIN{OFS=&quot;\t&quot;;} {if (length ($4) &gt; length ($5)) {print $1,$2,$2+ (length ($4) - 1),$3} else {print $1,$2, $2 + (length ($4) -1 ),$3}}&#39;</span> <span class="p">|</span> <span class="se">\</span>
        bgzip -c &gt; <span class="si">${</span><span class="nv">_output</span><span class="si">}</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="genotype-qc">
<h2>Genotype QC<a class="headerlink" href="#genotype-qc" title="Permalink to this headline">¶</a></h2>
<p>This step handles multi-allelic sites and annotate variants to known and novel. We add an RS ID to variants in dbSNP. Variants without rsID are considered novel variants.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># Handel multi-allelic sites, left normalization of indels and add variant ID</span>
<span class="o">[</span>qc_1 <span class="o">(</span>variant preprocessing<span class="o">)]</span>
parameter: <span class="nv">walltime</span> <span class="o">=</span> <span class="s1">&#39;24h&#39;</span>
<span class="c1"># Path to dbSNP variants generated previously</span>
parameter: <span class="nv">dbsnp_variants</span> <span class="o">=</span> path
<span class="c1"># Path to fasta file for HG reference genome, eg GRCh38_full_analysis_set_plus_decoy_hla.fa</span>
parameter: <span class="nv">reference_genome</span> <span class="o">=</span> path
parameter: <span class="nv">bi_allelic</span> <span class="o">=</span> False
parameter: <span class="nv">snp_only</span> <span class="o">=</span> False
input: genoFile, <span class="nv">group_by</span> <span class="o">=</span> <span class="m">1</span>
output: f<span class="s1">&#39;{cwd}/{_input:bnn}.{&quot;leftnorm&quot; if not bi_allelic else &quot;biallelic&quot;}{&quot;.snp&quot; if snp_only else &quot;&quot;}.vcf.gz&#39;</span>
task: <span class="nv">trunk_workers</span> <span class="o">=</span> <span class="m">1</span>, <span class="nv">trunk_size</span><span class="o">=</span>job_size, <span class="nv">walltime</span> <span class="o">=</span> walltime, <span class="nv">mem</span> <span class="o">=</span> mem, <span class="nv">cores</span> <span class="o">=</span> numThreads, <span class="nv">tags</span> <span class="o">=</span> f<span class="s1">&#39;{step_name}_{_output:bn}&#39;</span>
bash: <span class="nv">container</span> <span class="o">=</span> container, <span class="nv">expand</span><span class="o">=</span> <span class="s2">&quot;</span><span class="si">${</span><span class="p"> </span><span class="si">}</span><span class="s2">&quot;</span>, <span class="nv">stderr</span> <span class="o">=</span> f<span class="s1">&#39;{_output:n}.stderr&#39;</span>, <span class="nv">stdout</span> <span class="o">=</span> f<span class="s1">&#39;{_output:n}.stdout&#39;</span>
        <span class="si">${</span><span class="s1">&#39;bcftools norm -m-any&#39;</span><span class="p"> if not bi_allelic else </span><span class="s1">&#39;bcftools view -m2 -M2&#39;</span><span class="si">}</span> <span class="si">${</span><span class="s1">&#39;-v snps&#39;</span><span class="p"> if snp_only else </span><span class="s2">&quot;&quot;</span><span class="si">}</span> <span class="si">${</span><span class="nv">_input</span><span class="si">}</span> <span class="p">|</span><span class="se">\</span>
        bcftools norm --check-ref w -f <span class="si">${</span><span class="nv">reference_genome</span><span class="si">}</span>  -Oz<span class="p">|</span><span class="se">\</span>
        bcftools +fill-tags -- -t all,F_MISSING,<span class="s1">&#39;VD=sum(DP)&#39;</span> <span class="p">|</span> <span class="se">\</span>
        bcftools annotate -x ID -I +<span class="s1">&#39;%CHROM:%POS:%REF:%ALT&#39;</span> <span class="p">|</span> <span class="se">\</span>
        bcftools annotate -a <span class="si">${</span><span class="nv">dbsnp_variants</span><span class="si">}</span>  -h &lt;<span class="o">(</span><span class="nb">echo</span> <span class="s1">&#39;##INFO=&lt;ID=RSID,Number=1,Type=String,Description=&quot;dbSNP rsID&quot;&gt;&#39;</span><span class="o">)</span> -c CHROM,FROM,TO,ID -Oz &gt; <span class="si">${</span><span class="nv">_output</span><span class="si">}</span>
</pre></div>
</div>
</div>
</div>
<p>This step filter variants based on FILTER PASS, DP and QC, fraction of missing genotypes (all samples), and on HWE, for snps and indels. It will also remove monomorphic sites – using <code class="docutils literal notranslate"><span class="pre">bcftools</span> <span class="pre">view</span> <span class="pre">-c1</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># genotype QC</span>
<span class="o">[</span>qc_2 <span class="o">(</span>variant level QC<span class="o">)]</span>
parameter: <span class="nv">walltime</span> <span class="o">=</span> <span class="s1">&#39;24h&#39;</span>
<span class="c1"># Maximum missingess per-variant</span>
parameter: <span class="nv">geno_filter</span> <span class="o">=</span> <span class="m">0</span>.1
<span class="c1"># Sample level QC - read depth (DP) to filter out SNPs below this value</span>
parameter: <span class="nv">DP_snp</span> <span class="o">=</span> <span class="m">10</span>
<span class="c1"># Sample level QC - genotype quality (GQ) of specific sample. This measure tells you how confident we are that the genotype we assigned to a particular sample is correct</span>
parameter: <span class="nv">GQ</span> <span class="o">=</span> <span class="m">20</span>
<span class="c1"># Sample level QC - read depth (DP) to filter out indels below this value</span>
parameter: <span class="nv">DP_indel</span> <span class="o">=</span> <span class="m">10</span>
<span class="c1"># Allele balance for snps</span>
parameter: <span class="nv">AB_snp</span> <span class="o">=</span> <span class="m">0</span>.15
<span class="c1"># Allele balance for indels</span>
parameter: <span class="nv">AB_indel</span> <span class="o">=</span> <span class="m">0</span>.2
<span class="c1"># HWE filter </span>
parameter: <span class="nv">hwe_filter</span> <span class="o">=</span> 1e-06
output: f<span class="s2">&quot;{_input:nn}.filtered.vcf.gz&quot;</span>
task: <span class="nv">trunk_workers</span> <span class="o">=</span> <span class="m">1</span>, <span class="nv">trunk_size</span><span class="o">=</span>job_size, <span class="nv">walltime</span> <span class="o">=</span> walltime, <span class="nv">mem</span> <span class="o">=</span> mem, <span class="nv">cores</span> <span class="o">=</span> numThreads, <span class="nv">tags</span> <span class="o">=</span> f<span class="s1">&#39;{step_name}_{_output:bn}&#39;</span>
bash: <span class="nv">container</span> <span class="o">=</span> container, <span class="nv">expand</span><span class="o">=</span> <span class="s2">&quot;</span><span class="si">${</span><span class="p"> </span><span class="si">}</span><span class="s2">&quot;</span>, <span class="nv">stderr</span> <span class="o">=</span> f<span class="s1">&#39;{_output:n}.stderr&#39;</span>, <span class="nv">stdout</span> <span class="o">=</span> f<span class="s1">&#39;{_output:n}.stdout&#39;</span>
    bcftools filter -S . -e <span class="s1">&#39;(TYPE=&quot;SNP&quot; &amp; (FMT/DP)&lt;${DP_snp} &amp; (FMT/GQ)&lt;${GQ})|(TYPE=&quot;INDEL&quot; &amp; (FMT/DP)&lt;${DP_indel} &amp; (FMT/GQ)&lt;${GQ})&#39;</span> <span class="si">${</span><span class="nv">_input</span><span class="si">}</span> <span class="p">|</span> <span class="se">\</span>
    bcftools view -c1  <span class="p">|</span> bcftools view -f PASS <span class="p">|</span> <span class="se">\</span>
    bcftools filter -i <span class="s1">&#39;GT=&quot;hom&quot; | TYPE=&quot;snp&quot; &amp; GT=&quot;het&quot; &amp; (FORMAT/AD[*:1])/(FORMAT/AD[*:0] + FORMAT/AD[*:1]) &gt;= ${AB_snp} | TYPE=&quot;indel&quot; &amp; GT=&quot;het&quot; &amp; (FORMAT/AD[*:1])/(FORMAT/AD[*:0] + FORMAT/AD[*:1]) &gt;= ${AB_indel}&#39;</span> <span class="p">|</span> <span class="se">\</span>
    bcftools filter -i <span class="s1">&#39;F_MISSING&lt;${geno_filter} &amp; HWE&gt;${hwe_filter}&#39;</span> -Oz -o <span class="si">${</span><span class="nv">_output</span><span class="si">}</span> 
</pre></div>
</div>
</div>
</div>
<p>Finally we export it to PLINK 1.0 format, <strong>without keeping allele orders</strong>. Notice that PLINK 1.0 format does not allow for dosages. PLINK 2.0 format support it, but it is generally not supported by downstreams data analysis.</p>
<p>In the following code block the option <code class="docutils literal notranslate"><span class="pre">--vcf-half-call</span> <span class="pre">m</span></code>  treat half-call as missing.</p>
<p>Also, intentionally, <code class="docutils literal notranslate"><span class="pre">--keep-allele-order</span></code> is not applied. The resulting PLINK will lose ref/alt allele information but will go by major/minor allele, as conventionally used in standard PLINK format.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="o">[</span>qc_3 <span class="o">(</span><span class="nb">export</span> to PLINK<span class="o">)]</span>
parameter: <span class="nv">walltime</span> <span class="o">=</span> <span class="s1">&#39;24h&#39;</span>
output: f<span class="s1">&#39;{_input:nn}.bed&#39;</span>
task: <span class="nv">trunk_workers</span> <span class="o">=</span> <span class="m">1</span>, <span class="nv">trunk_size</span> <span class="o">=</span> job_size, <span class="nv">walltime</span> <span class="o">=</span> walltime, <span class="nv">mem</span> <span class="o">=</span> mem, <span class="nv">cores</span> <span class="o">=</span> numThreads, <span class="nv">tags</span> <span class="o">=</span> f<span class="s1">&#39;{step_name}_{_output:bn}&#39;</span>
bash: <span class="nv">container</span> <span class="o">=</span> container, <span class="nv">expand</span><span class="o">=</span> <span class="s2">&quot;</span><span class="si">${</span><span class="p"> </span><span class="si">}</span><span class="s2">&quot;</span>, <span class="nv">stderr</span> <span class="o">=</span> f<span class="s1">&#39;{_output:n}.stderr&#39;</span>, <span class="nv">stdout</span> <span class="o">=</span> f<span class="s1">&#39;{_output:n}.stdout&#39;</span>
    plink --vcf <span class="si">${</span><span class="nv">_input</span><span class="si">}</span> <span class="se">\</span>
        --vcf-half-call m <span class="se">\</span>
        --vcf-require-gt <span class="se">\</span>
        --allow-extra-chr <span class="se">\</span>
        --make-bed --out <span class="si">${</span><span class="nv">_output</span><span class="p">:</span><span class="nv">n</span><span class="si">}</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="o">[</span>qc_4 <span class="o">(</span>genotype data summary statistics<span class="o">)]</span>
parameter: <span class="nv">walltime</span> <span class="o">=</span> <span class="s1">&#39;24h&#39;</span>
input: output_from<span class="o">(</span><span class="s1">&#39;qc_1&#39;</span><span class="o">)</span>, output_from<span class="o">(</span><span class="s1">&#39;qc_2&#39;</span><span class="o">)</span>, <span class="nv">group_by</span> <span class="o">=</span> <span class="m">1</span>
output: f<span class="s2">&quot;{cwd}/{_input:bnn}.novel_variant_sumstats&quot;</span>, 
        f<span class="s2">&quot;{cwd}/{_input:bnn}.known_variant_sumstats&quot;</span>, 
        f<span class="s2">&quot;{cwd}/{_input:bnn}.novel_variant.snipsift_tstv&quot;</span>,
        f<span class="s2">&quot;{cwd}/{_input:bnn}.known_variant.snipsift_tstv&quot;</span>
task: <span class="nv">trunk_workers</span> <span class="o">=</span> <span class="m">1</span>, <span class="nv">trunk_size</span> <span class="o">=</span> job_size, <span class="nv">walltime</span> <span class="o">=</span> walltime, <span class="nv">mem</span> <span class="o">=</span> mem, <span class="nv">cores</span> <span class="o">=</span> numThreads, <span class="nv">tags</span> <span class="o">=</span> f<span class="s1">&#39;{step_name}_{_output[0]:bn}&#39;</span>
bash: <span class="nv">container</span> <span class="o">=</span> container, <span class="nv">expand</span><span class="o">=</span> <span class="s2">&quot;</span><span class="si">${</span><span class="p"> </span><span class="si">}</span><span class="s2">&quot;</span>, <span class="nv">stderr</span> <span class="o">=</span> f<span class="s1">&#39;{_output[0]:n}.stderr&#39;</span>, <span class="nv">stdout</span> <span class="o">=</span> f<span class="s1">&#39;{_output[0]:n}.stdout&#39;</span>
    <span class="c1"># Compute summary statistics, including TS/TV</span>
    bcftools stats -i <span class="s1">&#39;ID=&quot;.&quot;&#39;</span> -v  <span class="si">${</span><span class="nv">_input</span><span class="si">}</span> &gt; <span class="si">${</span><span class="nv">_output</span><span class="p">[0]</span><span class="si">}</span>
    bcftools stats -i <span class="s1">&#39;ID!=&quot;.&quot;&#39;</span> -v  <span class="si">${</span><span class="nv">_input</span><span class="si">}</span> &gt; <span class="si">${</span><span class="nv">_output</span><span class="p">[1]</span><span class="si">}</span>
    bcftools filter -i <span class="s1">&#39;ID=&quot;.&quot;&#39;</span>  <span class="si">${</span><span class="nv">_input</span><span class="si">}</span>   <span class="p">|</span> java -jar /opt/snpEff/SnpSift.jar tstv - &gt; <span class="si">${</span><span class="nv">_output</span><span class="p">[2]</span><span class="si">}</span>
    bcftools filter -i <span class="s1">&#39;ID!=&quot;.&quot;&#39;</span> <span class="si">${</span><span class="nv">_input</span><span class="si">}</span>  <span class="p">|</span> java -jar /opt/snpEff/SnpSift.jar tstv - &gt; <span class="si">${</span><span class="nv">_output</span><span class="p">[3]</span><span class="si">}</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "bash"
        },
        kernelOptions: {
            kernelName: "bash",
            path: "./code/data_preprocessing/genotype"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'bash'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="../genotype_preprocessing.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Genotype data preprocessing</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="GWAS_QC.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Genotype PLINK file quality control</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics Consortium<br/>
        
            &copy; Copyright 2021, ADSP-FG xQTL methods and data integration subgroup.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>