
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Factor analysis using PEER &#8212; ADSP-FG Project</title>
    
  <link href="../../../_static/css/theme.css" rel="stylesheet">
  <link href="../../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/togglebutton.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Factor analysis using Bi-Cross validation" href="BiCV_factor.html" />
    <link rel="prev" title="Principal component analysis" href="PCA.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">ADSP-FG Project</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../../README.html">
   xQTL analysis pipeline
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Complete Pipelines
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../complete_analysis/eQTL_analysis.html">
   Bulk RNA-seq eQTL analysis
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../molecular_phenotypes/apa_calling.html">
   APA Calling
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../molecular_phenotypes/bulk_expression.html">
   Bulk RNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../molecular_phenotypes/scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../genotype/GWAS_QC.html">
     GWAS data QC workflow
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../genotype/GRM.html">
     Genomic relationship matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../genotype/genotype_formatting.html">
     Genotype data reformat
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../phenotype/normalization.html">
     Molecular phenotype normalization
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../phenotype/phenotype_formatting.html">
     Phenotype data reformat
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="../covariate_preprocessing.html">
   Covariate data preprocessing
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2">
    <a class="reference internal" href="PCA.html">
     Principal component analysis
    </a>
   </li>
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     Factor analysis using PEER
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="BiCV_factor.html">
     Factor analysis using Bi-Cross validation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="remove_covariates.html">
     Computing covariate residuals
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../association_scan/cisQTL_scan.html">
   cisQTL analysis workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL calling pipeline
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/APEX/APEX.html">
     APEX QTL calling pipeline
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../association_scan/transQTL_scan.html">
   transQTL analysis workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../multivariate/METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../multivariate/MASH_pipeline.html">
   Multivariate association with MASH
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Statistical Fine-mapping
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../fine_mapping/univariate_fine_mapping.html">
   Univariate fine-mapping workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE_RSS.html">
     Fine-mapping with SuSiE RSS model
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE.html">
     Fine-mapping with SuSiE model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../fine_mapping/multivariate_fine_mapping.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multi-omics integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../integrative_analysis/colocalization.html">
   QTL and GWAS Colocalization
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/fastENLOC/fastenloc.html">
     Colocalization with FastENLOC
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../integrative_analysis/functional_annotation.html">
   Functional annotation integration
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/enrichment/gwas_enrichment.html">
     Enrichment analysis: TORUS and GREGOR
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../integrative_analysis/SuSiE_Ann/polyfun.html">
     Fine-mapping with PolyFun
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Utilities
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../misc/summary_stats_merger.html">
   Summary statistics merger
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../misc/rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH/NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../../_sources/pipeline/data_preprocessing/covariate/PEER_factor.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fpipeline/data_preprocessing/covariate/PEER_factor.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/pipeline/data_preprocessing/covariate/PEER_factor.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#overview">
   Overview:
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#input-and-output">
   Input and output:
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#required-input">
     Required input
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#optional-inputs">
     Optional inputs
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#outputs">
     Outputs:
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#main-outputs">
       Main outputs:
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#other-outputs">
       Other outputs:
      </a>
     </li>
    </ul>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#useage">
   Useage:
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#setup-and-global-parameters">
   Setup and global parameters:
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#minimum-working-example">
   Minimum working example
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#reference">
   Reference:
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="factor-analysis-using-peer">
<h1>Factor analysis using PEER<a class="headerlink" href="#factor-analysis-using-peer" title="Permalink to this headline">¶</a></h1>
<div class="section" id="overview">
<h2>Overview:<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>This workflow is intend to perform <a class="reference external" href="https://github.com/PMBio/peer/wiki/Tutorial">Probabilistic Estimation of Expression Residuals (PEER) method</a> , which is a standard part in GTEX protocal</p>
</div>
<div class="section" id="input-and-output">
<h2>Input and output:<a class="headerlink" href="#input-and-output" title="Permalink to this headline">¶</a></h2>
<p>Format of inputs and outputs will be same as that in <a class="reference external" href="https://cumc.github.io/xqtl-pipeline/pipeline/data_preprocessing/covariate/PEER_factor.html">APEX workflow</a>:</p>
<div class="section" id="required-input">
<h3>Required input<a class="headerlink" href="#required-input" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>–expression_file: Molecular trait data must be stored in <code class="docutils literal notranslate"><span class="pre">bed.gz</span></code> file format with indexing. Note that you must supply <strong>compressed</strong> version of <code class="docutils literal notranslate"><span class="pre">bed.gz</span></code> file and indexed. (i.e. run <code class="docutils literal notranslate"><span class="pre">bgzip</span> <span class="pre">your_file.bed</span> <span class="pre">&amp;&amp;</span> <span class="pre">tabix</span> <span class="pre">your_file.bed.gz</span></code>). An example of <code class="docutils literal notranslate"><span class="pre">bed.gz</span></code> file will be:</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="kn">readr:</span><span class="p">:</span><span class="n">read_delim</span><span class="p">(</span><span class="s2">&quot;Example_data/example_data.bed.gz&quot;</span><span class="p">,</span><span class="n">show_col_types</span><span class="o">=</span> <span class="n">F</span><span class="p">)[</span><span class="mi">1</span><span class="p">:</span><span class="mi">3</span><span class="p">,</span><span class="mi">1</span><span class="p">:</span><span class="mi">8</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table class="dataframe">
<caption>A tibble: 3 × 8</caption>
<thead>
	<tr><th scope=col>#chr</th><th scope=col>start</th><th scope=col>end</th><th scope=col>gene_name</th><th scope=col>Sample1</th><th scope=col>Sample2</th><th scope=col>Sample3</th><th scope=col>Sample4</th></tr>
	<tr><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>
</thead>
<tbody>
	<tr><td>1</td><td>  1</td><td>100</td><td>Gene1</td><td> -1.627455</td><td> 1.684725</td><td> -0.0807924</td><td> 1.510522</td></tr>
	<tr><td>1</td><td>101</td><td>200</td><td>Gene2</td><td>-12.626715</td><td>17.746163</td><td>-14.2519447</td><td> 5.254536</td></tr>
	<tr><td>1</td><td>201</td><td>300</td><td>Gene3</td><td>  3.337390</td><td>-7.259074</td><td>  6.2643526</td><td>-1.375911</td></tr>
</tbody>
</table>
</div></div>
</div>
<ul class="simple">
<li><p>–name: a string specify the name of main output file <code class="docutils literal notranslate"><span class="pre">{name}.PEER.cov</span></code></p></li>
</ul>
</div>
<div class="section" id="optional-inputs">
<h3>Optional inputs<a class="headerlink" href="#optional-inputs" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>–covariate: The covariate file supplied. Note that the sample names should be same (case sensitive: i.e. <code class="docutils literal notranslate"><span class="pre">Sample</span> <span class="pre">1</span></code> is <strong>not</strong> equal to <code class="docutils literal notranslate"><span class="pre">sample1</span></code>). Matching of samples in both files will be conduct first and the output will only have samples in both file.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="kn">readr:</span><span class="p">:</span><span class="n">read_delim</span><span class="p">(</span><span class="s2">&quot;Example_data/example_cov.txt&quot;</span><span class="p">,</span><span class="n">show_col_types</span><span class="o">=</span> <span class="n">F</span><span class="p">)[,</span><span class="mi">1</span><span class="p">:</span><span class="mi">8</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><table class="dataframe">
<caption>A tibble: 2 × 8</caption>
<thead>
	<tr><th scope=col>#id</th><th scope=col>Sample1</th><th scope=col>Sample2</th><th scope=col>Sample3</th><th scope=col>Sample4</th><th scope=col>Sample5</th><th scope=col>Sample6</th><th scope=col>Sample7</th></tr>
	<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>
</thead>
<tbody>
	<tr><td>Cov1</td><td> 1.6243454</td><td>-2.301539</td><td> 1.462108</td><td>-1.0998913</td><td>-1.100619</td><td>-0.6837279</td><td>-0.6916608</td></tr>
	<tr><td>Cov2</td><td>-0.6117564</td><td> 1.744812</td><td>-2.060141</td><td>-0.1724282</td><td> 1.144724</td><td>-0.1228902</td><td>-0.3967535</td></tr>
</tbody>
</table>
</div></div>
</div>
<ul class="simple">
<li><p>–N: Number of PEER factor used. Default value will be 0. Under default, Recommended PEER factors will be generated based on the suggestion of <a class="reference external" href="https://gtexportal.org/home/documentationPage">GTEX</a> automatically. To obtain <code class="docutils literal notranslate"><span class="pre">N</span></code> used in this case, please see the <code class="docutils literal notranslate"><span class="pre">.stdout</span></code> file.</p></li>
<li><p>Other parameters sepcified by PEER</p></li>
</ul>
</div>
<div class="section" id="outputs">
<h3>Outputs:<a class="headerlink" href="#outputs" title="Permalink to this headline">¶</a></h3>
<div class="section" id="main-outputs">
<h4>Main outputs:<a class="headerlink" href="#main-outputs" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">{name}.PEER.cov</span></code>: result for peer factor covariates</p></li>
</ul>
</div>
<div class="section" id="other-outputs">
<h4>Other outputs:<a class="headerlink" href="#other-outputs" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><p>A diagnosis polt for check of convergence</p></li>
<li><p>peer presision /residuals /weights</p></li>
</ul>
</div>
</div>
</div>
<div class="section" id="useage">
<h2>Useage:<a class="headerlink" href="#useage" title="Permalink to this headline">¶</a></h2>
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>    <span class="n">sos</span> <span class="n">run</span> <span class="n">Call_PEER</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">PEER</span> \
        <span class="o">--</span><span class="n">expression_file</span> <span class="n">Peer_example_data</span><span class="o">/</span><span class="n">Peer_example_data</span><span class="o">.</span><span class="n">txt</span> \
        <span class="o">--</span><span class="n">covariate</span> <span class="p">(</span><span class="n">optional</span><span class="p">)</span> \
        <span class="o">--</span><span class="n">cwd</span> <span class="o">/</span><span class="n">Users</span><span class="o">/</span><span class="n">albert29</span><span class="o">/</span><span class="n">Documents</span><span class="o">/</span><span class="n">Lab</span><span class="o">/</span><span class="n">Pipelines</span><span class="o">/</span><span class="n">Call_PEER</span><span class="o">/</span><span class="n">Output</span><span class="o">/</span> \
        <span class="o">--</span><span class="n">N</span> <span class="mi">10</span>
        <span class="o">...</span>
</pre></div>
</div>
</div>
<div class="section" id="setup-and-global-parameters">
<h2>Setup and global parameters:<a class="headerlink" href="#setup-and-global-parameters" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1"># The output directory for generated files. MUST BE FULL PATH</span>
<span class="kn">parameter:</span> <span class="n">wd</span> <span class="o">=</span> <span class="n">path</span>
<span class="n">cwd</span> <span class="o">=</span> <span class="n">wd</span>
<span class="c1"># For cluster jobs, number commands to run per job</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">1</span>
<span class="c1"># Wall clock time expected</span>
<span class="kn">parameter:</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s2">&quot;5h&quot;</span>
<span class="c1"># Memory expected</span>
<span class="kn">parameter:</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s2">&quot;16G&quot;</span>
<span class="c1"># Number of threads</span>
<span class="kn">parameter:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="mi">8</span>
<span class="c1"># Software container option</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="nb">str</span>
<span class="kn">parameter:</span> <span class="kp">name</span> <span class="o">=</span> <span class="s2">&quot;peer_out&quot;</span>


<span class="c1"># N PEER factors, If do not specify or specified as 0, default values suggested by </span>
<span class="c1"># UCSC (based on different sample size) Will be used</span>
<span class="kn">parameter:</span> <span class="n">N</span> <span class="o">=</span> <span class="mi">4</span>
<span class="n">n_of_factor</span> <span class="o">=</span> <span class="n">N</span>

<span class="c1"># Default values from PEER:</span>
<span class="c1">## The number of max iteration</span>

<span class="kn">parameter:</span> <span class="n">max_iter</span> <span class="o">=</span> <span class="mi">1000</span>
<span class="n">iteration</span> <span class="o">=</span> <span class="n">max_iter</span>
<span class="c1">## Prior parameters</span>
<span class="kn">parameter:</span> <span class="n">Alpha_a</span> <span class="o">=</span> <span class="mf">0.001</span>
<span class="kn">parameter:</span> <span class="n">Alpha_b</span> <span class="o">=</span> <span class="mf">0.1</span>
<span class="kn">parameter:</span> <span class="n">Eps_a</span> <span class="o">=</span> <span class="mf">0.1</span>
<span class="kn">parameter:</span> <span class="n">Eps_b</span> <span class="o">=</span> <span class="mf">10.</span>
<span class="c1">## Tolarance parameters</span>
<span class="kn">parameter:</span> <span class="n">tol</span> <span class="o">=</span> <span class="mf">0.001</span>
<span class="kn">parameter:</span> <span class="n">var_tol</span> <span class="o">=</span> <span class="mf">1e-08</span>
<span class="c1"># The molecular phenotype matrix</span>
<span class="kn">parameter:</span> <span class="n">molecular_pheno</span> <span class="o">=</span> <span class="n">path</span>
<span class="n">expression_file</span> <span class="o">=</span> <span class="n">molecular_pheno</span>
<span class="c1"># The covariate file</span>
<span class="kn">parameter:</span> <span class="n">covariate</span> <span class="o">=</span> <span class="s2">&quot;none&quot;</span>
<span class="kn">parameter:</span> <span class="n">covariate_file</span> <span class="o">=</span> <span class="n">covariate</span>
<span class="kn">parameter:</span> <span class="n">genotype_list</span> <span class="o">=</span> <span class="s2">&quot;None&quot;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">PEER_factor</span><span class="o">.</span><span class="n">ipynb</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>usage: sos run PEER_factor.ipynb [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &quot;sos run -h&quot; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  PEER

Global Workflow Options:
  --wd VAL (as path, required)
                        The output directory for generated files. MUST BE FULL
                        PATH
  --job-size 1 (as int)
                        For cluster jobs, number commands to run per job
  --walltime 5h
                        Wall clock time expected
  --mem 16384
                        Memory expected
  --numThreads 8 (as int)
                        Number of threads
  --container-apex VAL (as str, required)
                        Software container option
  --name VAL (as str, required)
  --N 4 (as int)
                        N PEER factors, If do not specify or specified as 0,
                        default values suggested by UCSC (based on different
                        sample size) Will be used
  --max-iter 1000 (as int)
  --Alpha-a 0.001 (as float)
                        Prior parameters
  --Alpha-b 0.1 (as float)
  --Eps-a 0.1 (as float)
  --Eps-b 10.0 (as float)
  --tol 0.001 (as float)
                        Tolarance parameters
  --var-tol 1e-08 (as float)
  --molecular-pheno VAL (as path, required)
                        The molecular phenotype matrix
  --covariate-file none
                        The covariate file

Sections
  PEER:
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">PEER</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">expression_file</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">_diagnosis_plot.pdf&#39;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">.PEER.cov&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="kp">walltime</span><span class="p">,</span> <span class="kp">mem</span> <span class="o">=</span> <span class="kp">mem</span><span class="p">,</span> <span class="kp">cores</span> <span class="o">=</span> <span class="n">numThreads</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">R:</span> <span class="n">container</span><span class="o">=</span><span class="n">container</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;${ }&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
    <span class="n">library</span><span class="p">(</span><span class="n">peer</span><span class="p">)</span>
  
    <span class="c1">#### PLOTING FUNCTIONS: ####  </span>
      
    <span class="c1">## Adapted from the source code of package: https://github.com/PMBio/peer/blob/40bc4b2cd92459ce42f44dfe279717436395f3f6/R/peer/R/plot.R </span>
  
  
    <span class="n">PEER_plotModel_adapted</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">model</span><span class="p">){</span>
        <span class="n">par</span><span class="p">(</span><span class="n">mfrow</span><span class="o">=</span><span class="n">c</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">1</span><span class="p">))</span>
        <span class="n">bounds</span> <span class="o">=</span> <span class="n">PEER_getBounds</span><span class="p">(</span><span class="n">model</span><span class="p">)</span>
        <span class="nb">vars</span> <span class="o">=</span> <span class="n">PEER_getResidualVars</span><span class="p">(</span><span class="n">model</span><span class="p">)</span>
        <span class="n">plot</span><span class="p">(</span><span class="n">bounds</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;l&quot;</span><span class="p">,</span> <span class="n">col</span><span class="o">=</span><span class="s2">&quot;red&quot;</span><span class="p">,</span> <span class="n">lwd</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">xlab</span><span class="o">=</span><span class="s2">&quot;Iterations&quot;</span><span class="p">,</span> <span class="n">ylab</span><span class="o">=</span><span class="s2">&quot;Lower bound&quot;</span><span class="p">)</span>
        <span class="n">par</span><span class="p">(</span><span class="n">new</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
        <span class="n">plot</span><span class="p">(</span><span class="nb">vars</span><span class="p">,,</span><span class="nb">type</span><span class="o">=</span><span class="s2">&quot;l&quot;</span><span class="p">,</span><span class="n">col</span><span class="o">=</span><span class="s2">&quot;blue&quot;</span><span class="p">,</span><span class="n">xaxt</span><span class="o">=</span><span class="s2">&quot;n&quot;</span><span class="p">,</span><span class="n">yaxt</span><span class="o">=</span><span class="s2">&quot;n&quot;</span><span class="p">,</span><span class="n">xlab</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span><span class="n">ylab</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
        <span class="n">axis</span><span class="p">(</span><span class="mi">4</span><span class="p">)</span>
        <span class="n">mtext</span><span class="p">(</span><span class="s2">&quot;Residual variance&quot;</span><span class="p">,</span><span class="n">side</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span><span class="n">line</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
        <span class="n">legend</span><span class="p">(</span><span class="s2">&quot;right&quot;</span><span class="p">,</span><span class="n">col</span><span class="o">=</span><span class="n">c</span><span class="p">(</span><span class="s2">&quot;red&quot;</span><span class="p">,</span><span class="s2">&quot;blue&quot;</span><span class="p">),</span><span class="n">lty</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span><span class="n">legend</span><span class="o">=</span><span class="n">c</span><span class="p">(</span><span class="s2">&quot;Lower bound&quot;</span><span class="p">,</span><span class="s2">&quot;Residual variance&quot;</span><span class="p">))</span>
        <span class="n">alpha</span> <span class="o">=</span> <span class="n">PEER_getAlpha</span><span class="p">(</span><span class="n">model</span><span class="p">)</span>
        <span class="n">plot</span><span class="p">(</span><span class="n">alpha</span><span class="p">,</span><span class="n">xlab</span><span class="o">=</span><span class="s2">&quot;Factors&quot;</span><span class="p">,</span><span class="n">ylab</span><span class="o">=</span><span class="s2">&quot;Inverse variance of factor weights&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;b&quot;</span><span class="p">,</span> <span class="n">col</span><span class="o">=</span><span class="s2">&quot;blue&quot;</span><span class="p">,</span> <span class="n">lwd</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">xaxp</span><span class="o">=</span><span class="n">c</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="n">length</span><span class="p">(</span><span class="n">alpha</span><span class="p">),</span> <span class="n">length</span><span class="p">(</span><span class="n">alpha</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">))</span>
    <span class="p">}</span>
    
    <span class="n">WriteTable</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">filename</span><span class="p">,</span> <span class="n">index</span><span class="o">.</span><span class="kp">name</span><span class="p">)</span> <span class="p">{</span>
      <span class="n">datafile</span> <span class="o">&lt;-</span> <span class="n">file</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="nb">open</span> <span class="o">=</span> <span class="s2">&quot;wt&quot;</span><span class="p">)</span>
      <span class="n">on</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="n">close</span><span class="p">(</span><span class="n">datafile</span><span class="p">))</span>
      <span class="n">header</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span><span class="n">index</span><span class="o">.</span><span class="kp">name</span><span class="p">,</span> <span class="n">colnames</span><span class="p">(</span><span class="n">data</span><span class="p">))</span>
      <span class="n">writeLines</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="n">header</span><span class="p">,</span> <span class="n">collapse</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">),</span> <span class="n">con</span> <span class="o">=</span> <span class="n">datafile</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
      <span class="n">write</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">datafile</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">col</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="n">F</span><span class="p">,</span> <span class="n">quote</span> <span class="o">=</span> <span class="n">F</span><span class="p">)</span>
    <span class="p">}</span>
  
    <span class="c1">#### MAIN ####</span>
    
  
    <span class="c1"># Start analysis:</span>
    <span class="n">model</span> <span class="o">&lt;-</span> <span class="n">PEER</span><span class="p">()</span>
    <span class="n">cat</span><span class="p">(</span><span class="s2">&quot;PEER: loading input data... &quot;</span><span class="p">)</span>

    <span class="n">df</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">delim</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">},</span> <span class="n">check</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="n">F</span><span class="p">)</span>
    <span class="n">mtx</span> <span class="o">=</span> <span class="n">df</span><span class="p">[,</span><span class="mi">5</span><span class="p">:</span><span class="n">ncol</span><span class="p">(</span><span class="n">df</span><span class="p">)]</span>
    <span class="n">rownames</span><span class="p">(</span><span class="n">mtx</span><span class="p">)</span> <span class="o">=</span> <span class="n">df</span><span class="err">$</span><span class="n">gene_ID</span>
    
    <span class="c1"># Load covariate file:</span>
    <span class="n">pc</span> <span class="o">&lt;-</span> <span class="s2">&quot;$</span><span class="si">{covariate_file}</span><span class="s2">&quot;</span>
    <span class="k">if</span><span class="p">(</span><span class="n">pc</span> <span class="o">!=</span>  <span class="s2">&quot;none&quot;</span><span class="p">){</span>
        <span class="n">cat</span><span class="p">(</span><span class="s2">&quot;PEER: loading covariate file ...&quot;</span><span class="p">)</span>
        <span class="n">cov</span><span class="o">.</span><span class="n">mat</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">delim</span><span class="p">(</span><span class="n">pc</span><span class="p">,</span> <span class="n">row</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">check</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="n">F</span><span class="p">)</span>      
        <span class="n">common_name</span> <span class="o">&lt;-</span> <span class="n">intersect</span><span class="p">(</span><span class="n">names</span><span class="p">(</span><span class="n">cov</span><span class="o">.</span><span class="n">mat</span><span class="p">),</span><span class="n">names</span><span class="p">(</span><span class="n">mtx</span><span class="p">))</span>
        <span class="k">if</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">common_name</span><span class="p">)</span> <span class="o">==</span><span class="mi">0</span><span class="p">){</span>
              <span class="n">stop</span><span class="p">(</span><span class="s2">&quot;No common samples! &quot;</span><span class="p">)</span>
          <span class="p">}</span><span class="k">else</span><span class="p">{</span>
              <span class="n">mtx</span> <span class="o">&lt;-</span> <span class="n">mtx</span><span class="p">[,</span><span class="n">common_name</span><span class="p">]</span>
              <span class="n">cov</span><span class="o">.</span><span class="n">mat</span> <span class="o">&lt;-</span> <span class="n">cov</span><span class="o">.</span><span class="n">mat</span><span class="p">[,</span><span class="n">common_name</span><span class="p">]</span>
              <span class="n">M</span> <span class="o">&lt;-</span> <span class="n">t</span><span class="p">(</span><span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">mtx</span><span class="p">,</span> <span class="n">rownames</span> <span class="o">=</span> <span class="n">T</span><span class="p">))</span>
              <span class="n">cov</span><span class="o">.</span><span class="n">mat</span> <span class="o">&lt;-</span> <span class="n">t</span><span class="p">(</span><span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">cov</span><span class="o">.</span><span class="n">mat</span><span class="p">))</span>
              <span class="n">cat</span><span class="p">(</span><span class="s2">&quot;done!</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
              <span class="n">cat</span><span class="p">(</span><span class="s2">&quot;PEER: Input summary:&quot;</span><span class="p">,</span><span class="n">dim</span><span class="p">(</span><span class="n">M</span><span class="p">)[</span><span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;samples with&quot;</span><span class="p">,</span><span class="n">dim</span><span class="p">(</span><span class="n">M</span><span class="p">)[</span><span class="mi">2</span><span class="p">],</span> <span class="s2">&quot;genes and&quot;</span><span class="p">,</span> <span class="n">dim</span><span class="p">(</span><span class="n">cov</span><span class="o">.</span><span class="n">mat</span><span class="p">)[</span><span class="mi">2</span><span class="p">],</span><span class="s2">&quot;covariates </span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
              <span class="n">invisible</span><span class="p">(</span><span class="n">PEER_setCovariates</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">cov</span><span class="o">.</span><span class="n">mat</span><span class="p">))</span>          
        <span class="p">}</span>     
    <span class="p">}</span><span class="k">else</span><span class="p">{</span>
        <span class="n">cov</span><span class="o">.</span><span class="n">mat</span> <span class="o">&lt;-</span> <span class="n">NULL</span>
        <span class="n">M</span> <span class="o">&lt;-</span> <span class="n">t</span><span class="p">(</span><span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">mtx</span><span class="p">,</span> <span class="n">rownames</span> <span class="o">=</span> <span class="n">T</span><span class="p">))</span>
        <span class="n">cat</span><span class="p">(</span><span class="s2">&quot;PEER: Input summary:&quot;</span><span class="p">,</span><span class="n">dim</span><span class="p">(</span><span class="n">M</span><span class="p">)[</span><span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;samples with&quot;</span><span class="p">,</span><span class="n">dim</span><span class="p">(</span><span class="n">M</span><span class="p">)[</span><span class="mi">2</span><span class="p">],</span> <span class="s2">&quot;genes </span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="p">}</span>
       
    <span class="c1"># Suggest the number of factors to use if no input value</span>
      
    <span class="k">if</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">N</span><span class="p">}</span> <span class="o">==</span> <span class="mi">0</span><span class="p">){</span>
      <span class="c1"># Use suggestion</span>
      <span class="k">if</span><span class="p">(</span><span class="n">dim</span><span class="p">(</span><span class="n">M</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">150</span><span class="p">){</span>
          <span class="n">num_factor</span> <span class="o">=</span> <span class="mi">15</span><span class="p">}</span><span class="k">else</span><span class="p">{</span>
              <span class="k">if</span><span class="p">(</span><span class="n">dim</span><span class="p">(</span><span class="n">M</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">250</span><span class="p">){</span><span class="n">num_factor</span> <span class="o">=</span> <span class="mi">30</span><span class="p">}</span><span class="k">else</span><span class="p">{</span><span class="n">num_factor</span> <span class="o">=</span> <span class="mi">35</span><span class="p">}</span>
          <span class="p">}</span>
    <span class="p">}</span><span class="k">else</span><span class="p">{</span><span class="n">num_factor</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">N</span><span class="p">}}</span>
      
    <span class="c1"># run PEER</span>
    <span class="n">cat</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;Setting initialization parameters ...&quot;</span><span class="p">))</span>
    
    <span class="n">invisible</span><span class="p">(</span><span class="n">PEER_setNk</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">num_factor</span><span class="p">))</span>
    <span class="n">invisible</span><span class="p">(</span><span class="n">PEER_setPhenoMean</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">M</span><span class="p">))</span>
    <span class="n">invisible</span><span class="p">(</span><span class="n">PEER_setPriorAlpha</span><span class="p">(</span><span class="n">model</span><span class="p">,</span><span class="err">$</span><span class="p">{</span><span class="n">Alpha_a</span><span class="p">},</span><span class="err">$</span><span class="p">{</span><span class="n">Alpha_b</span><span class="p">}))</span>
    <span class="n">invisible</span><span class="p">(</span><span class="n">PEER_setPriorEps</span><span class="p">(</span><span class="n">model</span><span class="p">,</span><span class="err">$</span><span class="p">{</span><span class="n">Eps_a</span><span class="p">},</span> <span class="err">$</span><span class="p">{</span><span class="n">Eps_b</span><span class="p">}))</span>
    <span class="n">invisible</span><span class="p">(</span><span class="n">PEER_setTolerance</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="err">$</span><span class="p">{</span><span class="n">tol</span><span class="p">}))</span>
    <span class="n">invisible</span><span class="p">(</span><span class="n">PEER_setVarTolerance</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="err">$</span><span class="p">{</span><span class="n">var_tol</span><span class="p">}))</span>
    <span class="n">invisible</span><span class="p">(</span><span class="n">PEER_setNmax_iterations</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="err">$</span><span class="p">{</span><span class="n">max_iter</span><span class="p">}))</span>
    <span class="k">if</span><span class="p">(</span><span class="err">!</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">cov</span><span class="o">.</span><span class="n">mat</span><span class="p">)){</span>
        <span class="n">invisible</span><span class="p">(</span><span class="n">PEER_setCovariates</span><span class="p">(</span><span class="n">model</span><span class="p">,</span> <span class="n">cov</span><span class="o">.</span><span class="n">mat</span><span class="p">))</span>
    <span class="p">}</span>
    <span class="n">cat</span><span class="p">(</span><span class="s2">&quot;Done.</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
  
    
    <span class="n">cat</span><span class="p">(</span><span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;PEER: estimating hidden confounders (&quot;</span><span class="p">,</span> <span class="n">num_factor</span><span class="p">,</span> <span class="s2">&quot;)</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">))</span>
    <span class="n">time</span> <span class="o">&lt;-</span> <span class="n">system</span><span class="o">.</span><span class="n">time</span><span class="p">(</span><span class="n">PEER_update</span><span class="p">(</span><span class="n">model</span><span class="p">))</span>
    
    <span class="c1"># Add relevant row/column names</span>
    
    <span class="n">factor</span><span class="o">.</span><span class="n">mat</span> <span class="o">&lt;-</span> <span class="n">PEER_getX</span><span class="p">(</span><span class="n">model</span><span class="p">)</span>  <span class="c1"># samples x PEER factors</span>
    <span class="n">weight</span><span class="o">.</span><span class="n">mat</span> <span class="o">&lt;-</span> <span class="n">PEER_getW</span><span class="p">(</span><span class="n">model</span><span class="p">)</span>  <span class="c1"># omic features x PEER factors</span>
    <span class="n">precision</span><span class="o">.</span><span class="n">mat</span> <span class="o">&lt;-</span> <span class="n">PEER_getAlpha</span><span class="p">(</span><span class="n">model</span><span class="p">)</span>  <span class="c1"># PEER factors x 1</span>
    <span class="n">resid</span><span class="o">.</span><span class="n">mat</span> <span class="o">&lt;-</span> <span class="n">t</span><span class="p">(</span><span class="n">PEER_getResiduals</span><span class="p">(</span><span class="n">model</span><span class="p">))</span>  <span class="c1"># omic features x samples</span>
    
    <span class="n">peer</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">names</span> <span class="o">&lt;-</span> <span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;peer.factor_&quot;</span><span class="p">,</span> <span class="mi">1</span><span class="p">:</span><span class="n">ncol</span><span class="p">(</span><span class="n">factor</span><span class="o">.</span><span class="n">mat</span><span class="p">))</span>
    <span class="n">rownames</span><span class="p">(</span><span class="n">factor</span><span class="o">.</span><span class="n">mat</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">rownames</span><span class="p">(</span><span class="n">M</span><span class="p">)</span>
    <span class="n">colnames</span><span class="p">(</span><span class="n">factor</span><span class="o">.</span><span class="n">mat</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">peer</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">names</span>
    <span class="n">colnames</span><span class="p">(</span><span class="n">weight</span><span class="o">.</span><span class="n">mat</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">peer</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">names</span>
    <span class="n">rownames</span><span class="p">(</span><span class="n">weight</span><span class="o">.</span><span class="n">mat</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">colnames</span><span class="p">(</span><span class="n">M</span><span class="p">)</span>
    <span class="n">rownames</span><span class="p">(</span><span class="n">precision</span><span class="o">.</span><span class="n">mat</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">peer</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">names</span>
    <span class="n">colnames</span><span class="p">(</span><span class="n">precision</span><span class="o">.</span><span class="n">mat</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="s2">&quot;alpha&quot;</span>
    <span class="n">precision</span><span class="o">.</span><span class="n">mat</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">precision</span><span class="o">.</span><span class="n">mat</span><span class="p">)</span>
    <span class="n">precision</span><span class="o">.</span><span class="n">mat</span><span class="err">$</span><span class="n">relevance</span> <span class="o">&lt;-</span> <span class="mf">1.0</span> <span class="o">/</span> <span class="n">precision</span><span class="o">.</span><span class="n">mat</span><span class="err">$</span><span class="n">alpha</span>
    <span class="n">rownames</span><span class="p">(</span><span class="n">resid</span><span class="o">.</span><span class="n">mat</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">colnames</span><span class="p">(</span><span class="n">M</span><span class="p">)</span>
    <span class="n">colnames</span><span class="p">(</span><span class="n">resid</span><span class="o">.</span><span class="n">mat</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">rownames</span><span class="p">(</span><span class="n">M</span><span class="p">)</span>
    
    <span class="n">cat</span><span class="p">(</span><span class="s2">&quot;Exporting results ... &quot;</span><span class="p">)</span>
      
    <span class="c1"># Diagnosis plot:</span>
     <span class="n">pdf</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{_output[0]}</span><span class="s2">&quot;</span><span class="p">)</span>
     <span class="n">invisible</span><span class="p">(</span><span class="n">PEER_plotModel_adapted</span><span class="p">(</span><span class="n">model</span><span class="p">))</span>
     <span class="n">invisible</span><span class="p">(</span><span class="n">dev</span><span class="o">.</span><span class="n">off</span><span class="p">())</span>
    
    <span class="c1"># Write 4 numeric results</span>
    <span class="n">WriteTable</span><span class="p">(</span><span class="n">t</span><span class="p">(</span><span class="n">factor</span><span class="o">.</span><span class="n">mat</span><span class="p">),</span> <span class="n">file</span><span class="o">.</span><span class="n">path</span><span class="p">(</span><span class="s2">&quot;$</span><span class="si">{_output[1]:d}</span><span class="s2">&quot;</span><span class="p">,</span><span class="s2">&quot;$</span><span class="si">{_output[1]:b}</span><span class="s2">&quot;</span><span class="p">),</span> <span class="s2">&quot;#id&quot;</span><span class="p">)</span>  
    <span class="n">WriteTable</span><span class="p">(</span><span class="n">weight</span><span class="o">.</span><span class="n">mat</span><span class="p">,</span> <span class="n">file</span><span class="o">.</span><span class="n">path</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">cwd</span><span class="p">:</span><span class="n">r</span><span class="p">},</span> <span class="s2">&quot;peer_weights.txt&quot;</span><span class="p">),</span> <span class="s2">&quot;#id&quot;</span><span class="p">)</span>
    <span class="n">WriteTable</span><span class="p">(</span><span class="n">precision</span><span class="o">.</span><span class="n">mat</span><span class="p">,</span> <span class="n">file</span><span class="o">.</span><span class="n">path</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">cwd</span><span class="p">:</span><span class="n">r</span><span class="p">},</span> <span class="s2">&quot;peer_precisions.txt&quot;</span><span class="p">),</span> <span class="s2">&quot;#id&quot;</span><span class="p">)</span>
    <span class="n">WriteTable</span><span class="p">(</span><span class="n">resid</span><span class="o">.</span><span class="n">mat</span><span class="p">,</span> <span class="n">file</span><span class="o">.</span><span class="n">path</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">cwd</span><span class="p">:</span><span class="n">r</span><span class="p">},</span> <span class="s2">&quot;peer_residuals.txt&quot;</span><span class="p">),</span> <span class="s2">&quot;#id&quot;</span><span class="p">)</span>
    <span class="n">cat</span><span class="p">(</span><span class="s2">&quot;Done.</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="minimum-working-example">
<h2>Minimum working example<a class="headerlink" href="#minimum-working-example" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">PEER_factor</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">PEER</span> \
        <span class="o">--</span><span class="n">molecular_pheno</span> <span class="n">Example_data</span><span class="o">/</span><span class="n">example_data</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span> \
        <span class="o">--</span><span class="n">covariate_file</span> <span class="n">Example_data</span><span class="o">/</span><span class="n">example_cov</span><span class="o">.</span><span class="n">txt</span> \
        <span class="o">--</span><span class="n">wd</span> <span class="o">./</span><span class="n">peer_output</span> \
        <span class="o">--</span><span class="kp">name</span> <span class="n">peer_out</span> \
        <span class="o">--</span><span class="n">container_apex</span> <span class="n">gouwh</span><span class="o">/</span><span class="n">r</span><span class="o">-</span><span class="n">peer</span><span class="p">:</span><span class="mf">1.0</span><span class="o">.</span><span class="mi">0</span> \
        <span class="o">--</span><span class="n">N</span> <span class="mi">10</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>INFO: Running <span class=" -Color -Color-Green">PEER</span>: 
HINT: Pulling docker image gouwh/r-peer:1.0.0
HINT: Docker image gouwh/r-peer:1.0.0 is now up to date
INFO: <span class=" -Color -Color-Green">PEER</span> is <span class=" -Color -Color-Green">completed</span>.
INFO: <span class=" -Color -Color-Green">PEER</span> output:   <span class=" -Color -Color-Green">peer_output/example_data.bed_diagnosis_plot.jpeg peer_output/peer_out.PEER.cov</span>
INFO: Workflow PEER (ID=w4ee1e4bf9082c0af) is executed successfully with 1 completed step.
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">tree</span> <span class="o">./</span><span class="n">peer_output</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Bold -Color-Bold-Blue">./peer_output</span>
├── <span class=" -Color -Color-Bold -Color-Bold-Magenta">example_data.bed_diagnosis_plot.jpeg</span>
├── peer_out.PEER.cov
├── peer_out.PEER.cov.stderr
├── peer_out.PEER.cov.stdout
├── peer_precisions.txt
├── peer_residuals.txt
└── peer_weights.txt

0 directories, 7 files
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="n">preview</span> <span class="o">./</span><span class="n">peer_output</span><span class="o">/</span><span class="n">example_data</span><span class="o">.</span><span class="n">bed_diagnosis_plot</span><span class="o">.</span><span class="n">jpeg</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div class="sos_hint">> ./peer_output/example_data.bed_diagnosis_plot.jpeg (22.2 KiB):</div></div><img alt="../../../_images/PEER_factor_13_1.jpg" src="../../../_images/PEER_factor_13_1.jpg" />
<div class="output text_html"><div class="sos_hint">> ./peer_output/example_data.bed_diagnosis_plot.jpeg:</div></div><div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>bash: ./peer_output/example_data.bed_diagnosis_plot.jpeg: Permission denied
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">cat</span> <span class="o">./</span><span class="n">peer_output</span><span class="o">/</span><span class="n">peer_out</span><span class="o">.</span><span class="n">PEER</span><span class="o">.</span><span class="n">cov</span><span class="o">.</span><span class="n">stdout</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>PEER: loading input data... PEER: loading covariate file ...done!
PEER: Input summary: 200 samples with 100 genes and 2 covariates 
Setting initialization parameters ...Done.
PEER: estimating hidden confounders (10)
	iteration 0/1000
	iteration 1/1000
	iteration 2/1000
	iteration 3/1000
	iteration 4/1000
	iteration 5/1000
	iteration 6/1000
	iteration 7/1000
	iteration 8/1000
Converged (var(residuals)) after 8 iterations

Exporting results ... Done.
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="reference">
<h2>Reference:<a class="headerlink" href="#reference" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Codes are adapted from <a class="reference external" href="https://github.com/RTIInternational/biocloud_docker_tools/blob/master/peer/v1.3/run_peer.R">here</a></p></li>
<li><p>GTEX recommandation of PEER factors is <a class="reference external" href="https://gtexportal.org/home/documentationPage">here</a></p></li>
<li><p>Examples by PEER is at <a class="reference external" href="https://github.com/PMBio/peer/wiki/Tutorial">github</a></p></li>
</ul>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            kernelName: "sos",
            path: "./pipeline/data_preprocessing/covariate"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="PCA.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Principal component analysis</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="BiCV_factor.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Factor analysis using Bi-Cross validation</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics Consortium<br/>
        
            &copy; Copyright 2021, ADSP-FG xQTL methods and data integration subgroup.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>