
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17: http://docutils.sourceforge.net/" />

    <title>TensorQTL QTL calling pipeline &#8212; ADSP-FG Workflows</title>
    
  <link href="../../../_static/css/theme.css" rel="stylesheet">
  <link href="../../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/togglebutton.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="cis/trans QTL analysis workflows" href="../QTL_association.html" />
    <link rel="prev" title="Gene region extraction" href="../../Data_Processing/Phenotype/Region_extraction.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../../_static/xqtl_wf.svg" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">ADSP-FG Workflows</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../../README.html">
   xQTL analysis pipeline
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Complete Pipelines
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../master_control.html">
   xQTL master workflow template
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Genotypes and Covariates
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../Data_Processing/Factor_and_Covariate/Apex_factor_test_log.html">
   APEX factor analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../Data_Processing/Factor_and_Covariate/factor.html">
   PEER factor analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../Data_Processing/Data_Processing.html">
   Genotype and covariate data preprocessing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../Data_Processing/Genotype/GWAS_QC.html">
   GWAS data QC workflow
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../Data_Processing/Genotype/LOCO_GRM.html">
   Leave one chromosome out GRM
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../Data_Processing/Genotype/PCA.html">
   Principal component analysis
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../Phenotypes/apacalling.html">
   APA calling Pipeline
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../Data_Processing/Phenotype/Normailzation.html">
   Molecular phenotype normalization
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../Data_Processing/Phenotype/Residual_Y.html">
   Molecular phenotype residuals
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../Data_Processing/Phenotype/Region_extraction.html">
   Gene region extraction
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   TensorQTL QTL calling pipeline
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../QTL_association.html">
   cis/trans QTL analysis workflows
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../APEX/APEX.html">
   APEX QTL calling pipeline
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Statistical Fine-mapping
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../Fine_Mapping/SumStat/SuSiE_RSS.html">
   Fine-mapping with SuSiE RSS model
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../Meta_Analysis/Reformatting/Signal_Extraction.html">
   Top signal extraction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../Meta_Analysis/Reformatting/Partitioned.html">
   Consolidating summary statistics from multiple studies
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../Meta_Analysis/METAL/METAL.html">
   Meta-analysis with METAL
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../Meta_Analysis/MASH/mashr_flashr_workflow.html">
   MASH analysis pipeline with data-driven prior matrices
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../Meta_Analysis/MASH/mixture_prior.html">
   A multivariate EBNM approach for mixture multivariate distribution estimate
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../Meta_Analysis/Meta_Analysis.html">
   Meta analysis
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Utilities
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../Data_Processing/Genotype/plink2VCF.html">
   Genotype data reformatting
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../Meta_Analysis/Reformatting/RDS_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../Meta_Analysis/Reformatting/Merged.html">
   Summary statistics formatting: RDS files in R
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH/NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../../_sources/workflow/QTL_association/TensorQTL/TensorQTL.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fworkflow/QTL_association/TensorQTL/TensorQTL.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/workflow/QTL_association/TensorQTL/TensorQTL.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#">
   TensorQTL QTL calling pipeline
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#pre-requisites">
     Pre-requisites
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#input-and-output">
   Input and Output
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#input">
     Input
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#output">
     Output
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#command-interface">
   Command interface
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#working-example">
   Working example
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#global-parameter-settings">
   Global parameter settings
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#qtl-sumstat-generation">
     QTL Sumstat generation
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#cis-qtl-sumstat-generation-via-tensorqtl">
     Cis QTL Sumstat generation via tensorQTL
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#trans-qtl-sumstat-generation-via-tensorqtl">
     Trans QTL Sumstat generation via tensorQTL
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#process-of-output">
     Process of output
    </a>
   </li>
  </ul>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="tensorqtl-qtl-calling-pipeline">
<h1>TensorQTL QTL calling pipeline<a class="headerlink" href="#tensorqtl-qtl-calling-pipeline" title="Permalink to this headline">¶</a></h1>
<p>This notebook Document a prototype workflow for using the TensorQTL in place of the APEX cis/trans to conduct analysis.</p>
<p><img alt="Blank diagram-10.png" src="workflow/QTL_association/TensorQTL/attachment:8073388f-23bf-4c8e-ae4f-5efcec3df68e.png" /></p>
<section id="pre-requisites">
<h2>Pre-requisites<a class="headerlink" href="#pre-requisites" title="Permalink to this headline">¶</a></h2>
<p>This pipeline require the tensorQTL and R:Tidyverse. A singularity container is provided along with the mwe</p>
</section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="input-and-output">
<h1>Input and Output<a class="headerlink" href="#input-and-output" title="Permalink to this headline">¶</a></h1>
<section id="input">
<h2>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">--molecular-pheno</span></code>, The text file containing the table describing the molecular phenotype. It shall have regions(genes) as rows and samples as columnes</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--region_list</span></code> The text file with 4 columns specifying the #Chr, P0 (Start position), P1(End position) and names of regions to analyze. The name of the column is not important but the order of the columns. It is also important that the column name of the first column starts with a #. The region_list can can be generated by using another sos pipeline SOS_ROSMAP_gene_exp_processing.ipynb.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">genotype_list</span></code> as the pipeline required whole genome vcf file for each chromosome, those vcf are converted beforehand from plink trio using the following commands:</p></li>
</ul>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="err">$</span><span class="p">(</span><span class="n">seq</span> <span class="mi">1</span> <span class="mi">22</span><span class="p">);</span> <span class="n">do</span>

<span class="n">plink</span> <span class="o">--</span><span class="n">bfile</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">ctcn</span><span class="o">/</span><span class="n">datasets</span><span class="o">/</span><span class="n">rosmap</span><span class="o">/</span><span class="n">wgs</span><span class="o">/</span><span class="n">ampad</span><span class="o">/</span><span class="n">variants</span><span class="o">/</span><span class="n">snvCombined</span><span class="o">/</span><span class="n">plink_files</span><span class="o">/</span><span class="n">ROSMAP_chr</span><span class="err">$</span><span class="n">i</span> \
<span class="o">--</span><span class="n">recode</span> <span class="n">vcf</span><span class="o">-</span><span class="n">iid</span>       <span class="o">--</span><span class="n">out</span> <span class="n">ROSMAP_chr</span><span class="err">$</span><span class="n">i</span> <span class="p">;</span>
<span class="n">bgzip</span> <span class="n">ROSMAP_chr</span><span class="err">$</span><span class="n">i</span><span class="o">.</span><span class="n">vcf</span>  <span class="p">;</span>
<span class="n">tabix</span> <span class="o">-</span><span class="n">p</span> <span class="n">vcf</span> <span class="n">ROSMAP_chr</span><span class="err">$</span><span class="n">i</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> <span class="p">;</span>
<span class="n">tabix</span> <span class="o">-</span><span class="n">H</span> <span class="n">ROSMAP_chr</span><span class="err">$</span><span class="n">i</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span> <span class="p">;</span>
<span class="n">echo</span> <span class="err">$</span><span class="n">i</span><span class="p">;</span>
<span class="n">done</span>
</pre></div>
</div>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">grm_list</span></code> are a file containing list of grm matrixs that generated by the GRM module of this pipeline.</p></li>
</ul>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h2>
</section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="command-interface">
<h1>Command interface<a class="headerlink" href="#command-interface" title="Permalink to this headline">¶</a></h1>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="err">!</span><span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">GIT</span><span class="o">/</span><span class="n">neuro</span><span class="o">-</span><span class="n">apex</span><span class="o">/</span><span class="n">apex</span><span class="o">/</span><span class="n">Apex_prototyping</span><span class="o">.</span><span class="n">ipynb</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>usage: sos run /home/hs3163/GIT/neuro-apex/apex/Apex_prototyping.ipynb
               [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &quot;sos run -h&quot; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  Expression_Filtering
  apex_cis
  Factor_analysis
  LMM
  cis
  trans
  GRM

Global Workflow Options:
  --molecular-pheno-whole VAL (as path, required)
                        Path to the input molecular phenotype data.
  --ori-covariate &#39;&#39;
                        Covariate file, in similar format as the molecular_pheno
  --genotype-list VAL (as path, required)
                        List of Genotype file in plink trio format
  --region-list VAL (as path, required)
                        An index text file with 4 columns specifying the chr,
                        start, end and names of regions to analyze
  --wd . (as path)
                        Path to the work directory of the weight computation:
                        output weights and cache will be saved to this
                        directory.
  --job-size 2 (as int)
                        Specify the number of jobs per run.
  --container &#39;&#39;
                        Container option for software to run the analysis:
                        docker or singularity
  --Prefix ROSMAP
                        Prefix for the analysis output
  --place-holder-vcf /mnt/mfs/statgen/neuro-apex/mwe/data/ROSMAP_chr2.vcf.gz (as path)
                        Path to the vcf file that act as a place holder
  --grm-list . (as path)
                        GRM: a prefix for the folder and prefix for the GRM of
                        each chromosome
  --factor-covariate  f&#39;{wd:a}/cache/{Prefix}.cov.gz&#39;

                        Covariate
  --n-of-factor 5 (as int)
                        Parameter to be used (TB Number of latent common factors
  --iteration 0 (as int)
                        Number of factor analysis iterations (0 for PCA).
  --p 0.05 (as float)
                        Factor analysis prior
  --tau 0.05 (as float)
  --window 1000000 (as list)
                        Specify the scanning window for the up and downstream
                        radius to analyze around the region of interest, in
                        units of Kb

Sections
  Expression_Filtering_1, apex_cis_1:
  Expression_Filtering_2, apex_cis_2:
  Factor_analysis, apex_cis_3:
  LMM, apex_cis_4:
    Workflow Options:
      --threads 3 (as int)
  cis, apex_cis_5:
    Workflow Options:
      --threads 3 (as int)
  trans:
    Workflow Options:
      --threads 3 (as int)
  GRM_1:
  GRM_2:
  GRM_3:
  GRM_4:
</pre></div>
</div>
</div>
</div>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="working-example">
<h1>Working example<a class="headerlink" href="#working-example" title="Permalink to this headline">¶</a></h1>
<p>please <a class="reference external" href="https://drive.google.com/file/d/1838xUOQuWTszQ0WJGXNiJMszY05cw3RS/view?usp=sharing">download a minimal working example here</a>, unzip the mwe file, and in the folder containing said folder run the following commands:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1">## Test pipeline with test data</span>
<span class="c1">## Switch back to abosolute path, otherwise there will be file not found error in step 5</span>
<span class="n">nohup</span> <span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">GIT</span><span class="o">/</span><span class="n">neuro</span><span class="o">-</span><span class="n">apex</span><span class="o">/</span><span class="n">apex</span><span class="o">/</span><span class="n">Apex_prototyping</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">apex_cis</span> \
  <span class="o">--</span><span class="n">molecular_pheno_whole</span> <span class="o">./</span><span class="n">mwe</span><span class="o">/</span><span class="n">mole_pheno</span>   \
  <span class="o">--</span><span class="n">wd</span> <span class="o">./</span> \
  <span class="o">--</span><span class="n">genotype_list</span> <span class="o">./</span><span class="n">mwe</span><span class="o">/</span><span class="n">mwe</span><span class="o">.</span><span class="n">genotype_list</span><span class="o">.</span><span class="n">txt</span> \
  <span class="o">--</span><span class="n">region_list</span> <span class="o">./</span><span class="n">mwe</span><span class="o">/</span><span class="n">mwe_region</span> \
  <span class="o">--</span><span class="n">iteration</span> <span class="mi">2</span> \
  <span class="o">--</span><span class="n">container</span> <span class="s2">&quot;./mwe/apex.sif&quot;</span> \
  <span class="o">--</span><span class="n">job_size</span> <span class="mi">1</span> \
  <span class="o">--</span><span class="n">Prefix</span> <span class="s2">&quot;MWE&quot;</span> \
  <span class="o">--</span><span class="n">grm_list</span> <span class="s2">&quot;./mwe/grm/MWE.grm_list.txt&quot;</span> \
  <span class="o">-</span><span class="n">J</span> <span class="mi">50</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span> <span class="o">-</span><span class="n">c</span> <span class="o">~/</span><span class="n">GIT</span><span class="o">/</span><span class="n">neuro</span><span class="o">-</span><span class="n">twas</span><span class="o">/</span><span class="n">code</span><span class="o">/</span><span class="n">csg</span><span class="o">.</span><span class="n">yml</span> <span class="o">&amp;</span>



<span class="n">nohup</span> <span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">GIT</span><span class="o">/</span><span class="n">neuro</span><span class="o">-</span><span class="n">apex</span><span class="o">/</span><span class="n">workflow</span><span class="o">/</span><span class="n">Apex_TensorQTL</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">tensor_cis</span> \
  <span class="o">--</span><span class="n">molecular_pheno_whole</span> <span class="o">./</span><span class="n">mwe</span><span class="o">/</span><span class="n">mole_pheno</span>   \
  <span class="o">--</span><span class="n">wd</span> <span class="o">./</span> \
  <span class="o">--</span><span class="n">genotype_list</span> <span class="o">./</span><span class="n">mwe</span><span class="o">/</span><span class="n">mwe</span><span class="o">.</span><span class="n">genotype_list</span><span class="o">.</span><span class="n">txt</span> \
  <span class="o">--</span><span class="n">genotype_list_tensor</span> <span class="o">./</span><span class="n">mwe</span><span class="o">/</span><span class="n">mwe</span><span class="o">.</span><span class="n">tensor_genotype</span> \
  <span class="o">--</span><span class="n">region_list</span> <span class="o">./</span><span class="n">mwe</span><span class="o">/</span><span class="n">mwe_region</span> \
  <span class="o">--</span><span class="n">iteration</span> <span class="mi">2</span> \
  <span class="o">--</span><span class="n">container</span> <span class="s2">&quot;./mwe/apex.sif&quot;</span> \
  <span class="o">--</span><span class="n">job_size</span> <span class="mi">1</span> \
  <span class="o">--</span><span class="n">Prefix</span> <span class="s2">&quot;MWE&quot;</span> \
  <span class="o">-</span><span class="n">J</span> <span class="mi">50</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span> <span class="o">-</span><span class="n">c</span> <span class="o">~/</span><span class="n">GIT</span><span class="o">/</span><span class="n">neuro</span><span class="o">-</span><span class="n">twas</span><span class="o">/</span><span class="n">code</span><span class="o">/</span><span class="n">csg</span><span class="o">.</span><span class="n">yml</span> <span class="o">&amp;</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>ERROR: Failed to locate twas_fusion.ipynb.sos
</pre></div>
</div>
</div>
</div>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="global-parameter-settings">
<h1>Global parameter settings<a class="headerlink" href="#global-parameter-settings" title="Permalink to this headline">¶</a></h1>
<p>The section outlined the parameters that can be set in the command interface.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1"># Path to the input molecular phenotype file, per chrm, in bed.gz format.</span>
<span class="kn">parameter:</span> <span class="n">molecular_pheno</span> <span class="o">=</span> <span class="n">path</span>
<span class="n">molecular_pheno_str</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">molecular_pheno</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="c1"># Covariate file, in similar format as the molecular_pheno</span>
<span class="kn">parameter:</span> <span class="n">covariate</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># Genotype file in plink trio format, per chrm</span>
<span class="kn">parameter:</span> <span class="n">genotype_file</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># An optional subset of region list containing a column of ENSG gene_id to limit the analysis</span>
<span class="kn">parameter:</span> <span class="n">region_list</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;./&quot;</span><span class="p">)</span>
<span class="c1"># Path to the work directory of the analysis.</span>
<span class="kn">parameter:</span> <span class="n">wd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;./&#39;</span><span class="p">)</span>
<span class="c1"># Specify the number of jobs per run.</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">2</span>
<span class="c1"># Container option for software to run the analysis: docker or singularity</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
<span class="c1"># Prefix for the analysis output</span>
<span class="kn">parameter:</span> <span class="n">Prefix</span> <span class="o">=</span> <span class="s1">&#39;ROSMAP&#39;</span>


<span class="c1"># Specify the scanning window for the up and downstream radius to analyze around the region of interest, in units of bp</span>
<span class="kn">parameter:</span> <span class="n">window</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;1000000&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<section id="qtl-sumstat-generation">
<h2>QTL Sumstat generation<a class="headerlink" href="#qtl-sumstat-generation" title="Permalink to this headline">¶</a></h2>
<p>This step generate the cis-QTL summary statistics and vcov (covariate-adjusted LD) files for downstream analysis from summary statistics. The analysis is done per chromosome to reduce running time.</p>
</section>
<section id="cis-qtl-sumstat-generation-via-tensorqtl">
<h2>Cis QTL Sumstat generation via tensorQTL<a class="headerlink" href="#cis-qtl-sumstat-generation-via-tensorqtl" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">tensor_cis_1</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">molecular_pheno</span><span class="p">,</span> <span class="n">genotype_file</span><span class="p">,</span> <span class="n">covariate</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">wd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bnnn</span><span class="si">}</span><span class="s1">.cis_qtl_pairs.</span><span class="si">{</span><span class="n">molecular_pheno_str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">)</span><span class="si">}</span><span class="s1">.parquet&#39;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">wd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bnnn</span><span class="si">}</span><span class="s1">.norminal.cis_long_table.txt&#39;</span><span class="c1">#,</span>
        <span class="c1">#f&#39;{wd:a}/{_input[0]:bnnn}.emprical.cis_sumstats.txt&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;12h&#39;</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;40G&#39;</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span><span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
    <span class="n">touch</span>  <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span><span class="n">time_stamp</span>
<span class="kn">python:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
    <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
    <span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
    <span class="kn">import</span> <span class="nn">tensorqtl</span>
    <span class="kn">from</span> <span class="nn">tensorqtl</span> <span class="kn">import</span> <span class="n">genotypeio</span><span class="p">,</span> <span class="n">cis</span><span class="p">,</span> <span class="n">trans</span>
    <span class="c1">## Defineing parameter</span>
    <span class="n">plink_prefix_path</span> <span class="o">=</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span><span class="n">nr</span><span class="p">]</span>
    <span class="n">expression_bed</span> <span class="o">=</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">r</span><span class="p">]</span>
    <span class="n">covariates_file</span> <span class="o">=</span> <span class="s2">&quot;$[_input[2]]&quot;</span>
    <span class="n">Prefix</span> <span class="o">=</span> <span class="s2">&quot;$[_output[0]:nn]&quot;</span>
    <span class="c1">## Loading Data</span>
    <span class="n">phenotype_df</span><span class="p">,</span> <span class="n">phenotype_pos_df</span> <span class="o">=</span> <span class="n">tensorqtl</span><span class="o">.</span><span class="n">read_phenotype_bed</span><span class="p">(</span><span class="n">expression_bed</span><span class="p">)</span>
    <span class="c1">### Filter by the optional keep gene</span>
    <span class="c1">##if $[region_list].is_file():</span>
    <span class="c1">##    region = pd.read_csv(&quot;$[region_list]&quot;,&quot;\t&quot;)</span>
    <span class="c1">##    keep_gene = region[&quot;gene_ID&quot;].to_list()</span>
    <span class="c1">##    phenotype_df = phenotype_df.query(&#39;gene_ID  in keep_gene&#39;)</span>
    <span class="c1">##    phenotype_pos_df = phenotype_pos_df.query(&#39;gene_ID  in keep_gene&#39;)</span>


    <span class="n">covariates_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">covariates_file</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">T</span>
    <span class="n">pr</span> <span class="o">=</span> <span class="n">genotypeio</span><span class="o">.</span><span class="n">PlinkReader</span><span class="p">(</span><span class="n">plink_prefix_path</span><span class="p">)</span>
    <span class="n">genotype_df</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">load_genotypes</span><span class="p">()</span>
    <span class="n">variant_df</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">bim</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;snp&#39;</span><span class="p">)[[</span><span class="s1">&#39;chrom&#39;</span><span class="p">,</span> <span class="s1">&#39;pos&#39;</span><span class="p">]]</span>
    <span class="c1">## Retaining only common samples</span>
    <span class="n">phenotype_df</span> <span class="o">=</span> <span class="n">phenotype_df</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">intersect1d</span><span class="p">(</span><span class="n">phenotype_df</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> <span class="n">covariates_df</span><span class="o">.</span><span class="n">index</span><span class="p">)]</span>
    <span class="n">covariates_df</span> <span class="o">=</span> <span class="n">covariates_df</span><span class="o">.</span><span class="n">transpose</span><span class="p">()[</span><span class="n">np</span><span class="o">.</span><span class="n">intersect1d</span><span class="p">(</span><span class="n">phenotype_df</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> <span class="n">covariates_df</span><span class="o">.</span><span class="n">index</span><span class="p">)]</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span>
    <span class="c1">## cis-QTL mapping: nominal associations for all variant-phenotype pairs</span>
    <span class="n">cis</span><span class="o">.</span><span class="n">map_nominal</span><span class="p">(</span><span class="n">genotype_df</span><span class="p">,</span> <span class="n">variant_df</span><span class="p">,</span>
                <span class="n">phenotype_df</span><span class="p">,</span>
                <span class="n">phenotype_pos_df</span><span class="p">,</span>
                <span class="n">Prefix</span><span class="p">,</span> <span class="n">covariates_df</span><span class="o">=</span><span class="n">covariates_df</span><span class="p">)</span>
    <span class="c1">## cis-QTL mapping: empirical p-values for phenotypes</span>
    <span class="c1">#cis_df = cis.map_cis(genotype_df, variant_df, </span>
    <span class="c1">#                 phenotype_df,</span>
    <span class="c1">#                 phenotype_pos_df,</span>
    <span class="c1">#                 covariates_df=covariates_df, seed=999)</span>
    <span class="c1">#cis_df.index.name = &quot;gene_id&quot;</span>
    <span class="c1">#cis_df.to_csv(&quot;$[_output[2]]&quot;, sep=&#39;\t&#39;)</span>

    <span class="c1">## Load the parquet and save it as txt</span>
    <span class="n">pairs_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_parquet</span><span class="p">(</span><span class="s2">&quot;$[_output[0]]&quot;</span><span class="p">)</span>
    <span class="n">pairs_df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span>  <span class="o">=</span> <span class="s2">&quot;pval&quot;</span>
    <span class="n">pairs_df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span>  <span class="o">=</span> <span class="s2">&quot;beta&quot;</span>
    <span class="n">pairs_df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span>  <span class="o">=</span> <span class="s2">&quot;se&quot;</span>
    <span class="n">pairs_df</span> <span class="o">=</span> <span class="n">pairs_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
    <span class="n">alt</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">dataframe</span><span class="p">:</span> <span class="n">dataframe</span><span class="p">[</span><span class="s1">&#39;variant_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">variant_id</span><span class="p">:</span><span class="n">variant_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]))</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
    <span class="n">ref</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">dataframe</span><span class="p">:</span> <span class="n">dataframe</span><span class="p">[</span><span class="s1">&#39;variant_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">variant_id</span><span class="p">:</span><span class="n">variant_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">2</span><span class="p">]))</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
    <span class="n">pos</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">dataframe</span><span class="p">:</span> <span class="n">dataframe</span><span class="p">[</span><span class="s1">&#39;variant_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">variant_id</span><span class="p">:</span><span class="n">variant_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]))</span>
    <span class="n">pairs_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;$[_output[1]]&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span><span class="n">index</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="trans-qtl-sumstat-generation-via-tensorqtl">
<h2>Trans QTL Sumstat generation via tensorQTL<a class="headerlink" href="#trans-qtl-sumstat-generation-via-tensorqtl" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">tensor_trans_1</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">molecular_pheno</span><span class="p">,</span> <span class="n">genotype_file</span><span class="p">,</span> <span class="n">covariate</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">wd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bnnn</span><span class="si">}</span><span class="s1">.trans_sumstats.txt&#39;</span>

<span class="kn">parameter:</span> <span class="n">batch_size</span> <span class="o">=</span> <span class="mi">10000</span>
<span class="kn">parameter:</span> <span class="n">pval_threshold</span> <span class="o">=</span> <span class="mf">1e-5</span>
<span class="kn">parameter:</span> <span class="n">maf_threshold</span> <span class="o">=</span> <span class="mf">0.05</span>

<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;12h&#39;</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;40G&#39;</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span><span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
    <span class="n">touch</span>  <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span><span class="n">time_stamp</span>
<span class="kn">python:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span><span class="n">container</span> <span class="o">=</span><span class="n">container</span> 
    <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
    <span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
    <span class="kn">import</span> <span class="nn">tensorqtl</span>
    <span class="kn">from</span> <span class="nn">tensorqtl</span> <span class="kn">import</span> <span class="n">genotypeio</span><span class="p">,</span> <span class="n">cis</span><span class="p">,</span> <span class="n">trans</span>
    <span class="c1">## Defineing parameter</span>
    <span class="n">plink_prefix_path</span> <span class="o">=</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">2</span><span class="p">]:</span><span class="n">r</span><span class="p">]</span>
    <span class="n">expression_bed</span> <span class="o">=</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">r</span><span class="p">]</span>
    <span class="n">covariates_file</span> <span class="o">=</span> <span class="s2">&quot;$[_input[2]]&quot;</span>
    <span class="n">Prefix</span> <span class="o">=</span> <span class="s2">&quot;$[_output[0]:nnn]&quot;</span>
    <span class="c1">## Loading Data</span>
    <span class="n">phenotype_df</span><span class="p">,</span> <span class="n">phenotype_pos_df</span> <span class="o">=</span> <span class="n">tensorqtl</span><span class="o">.</span><span class="n">read_phenotype_bed</span><span class="p">(</span><span class="n">expression_bed</span><span class="p">)</span>


    <span class="c1">##### Filter by the optional keep gene</span>
    <span class="c1">##if $[region_list].is_file():</span>
    <span class="c1">##    region = pd.read_csv(&quot;$[region_list]&quot;,&quot;\t&quot;)</span>
    <span class="c1">##    keep_gene = region[&quot;gene_ID&quot;].to_list()</span>
    <span class="c1">##    phenotype_df = phenotype_df.query(&#39;gene_ID  in keep_gene&#39;)</span>
    <span class="c1">##    phenotype_pos_df = phenotype_pos_df.query(&#39;gene_ID  in keep_gene&#39;)</span>


    <span class="n">covariates_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">covariates_file</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">T</span>
    <span class="n">pr</span> <span class="o">=</span> <span class="n">genotypeio</span><span class="o">.</span><span class="n">PlinkReader</span><span class="p">(</span><span class="n">plink_prefix_path</span><span class="p">)</span>
    <span class="n">genotype_df</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">load_genotypes</span><span class="p">()</span>
    <span class="n">variant_df</span> <span class="o">=</span> <span class="n">pr</span><span class="o">.</span><span class="n">bim</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;snp&#39;</span><span class="p">)[[</span><span class="s1">&#39;chrom&#39;</span><span class="p">,</span> <span class="s1">&#39;pos&#39;</span><span class="p">]]</span>
    <span class="c1">## Retaining only common samples</span>
    <span class="n">phenotype_df</span> <span class="o">=</span> <span class="n">phenotype_df</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">intersect1d</span><span class="p">(</span><span class="n">phenotype_df</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> <span class="n">covariates_df</span><span class="o">.</span><span class="n">index</span><span class="p">)]</span>
    <span class="n">covariates_df</span><span class="o">.</span><span class="n">transpose</span><span class="p">()[</span><span class="n">np</span><span class="o">.</span><span class="n">intersect1d</span><span class="p">(</span><span class="n">phenotype_df</span><span class="o">.</span><span class="n">columns</span><span class="p">,</span> <span class="n">covariates_df</span><span class="o">.</span><span class="n">index</span><span class="p">)]</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span>
    <span class="c1">## Trans analysis</span>
    <span class="n">trans_df</span> <span class="o">=</span> <span class="n">trans</span><span class="o">.</span><span class="n">map_trans</span><span class="p">(</span><span class="n">genotype_df</span><span class="p">,</span> <span class="n">phenotype_df</span><span class="p">,</span> <span class="n">covariates_df</span><span class="p">,</span> <span class="n">batch_size</span><span class="o">=</span><span class="err">$</span><span class="p">[</span><span class="n">batch_size</span><span class="p">],</span>
                           <span class="n">return_sparse</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">pval_threshold</span><span class="o">=</span><span class="err">$</span><span class="p">[</span><span class="n">pval_threshold</span><span class="p">],</span> <span class="n">maf_threshold</span><span class="o">=</span><span class="err">$</span><span class="p">[</span><span class="n">maf_threshold</span><span class="p">])</span>

    <span class="c1">## Filter out cis signal</span>
    <span class="n">trans_df</span> <span class="o">=</span> <span class="n">trans</span><span class="o">.</span><span class="n">filter_cis</span><span class="p">(</span><span class="n">trans_df</span><span class="p">,</span> <span class="n">phenotype_pos_df</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">to_dict</span><span class="p">(),</span> <span class="n">variant_df</span><span class="p">,</span> <span class="n">window</span><span class="o">=</span><span class="err">$</span><span class="p">[</span><span class="n">window</span><span class="p">])</span>
    <span class="c1">## Output</span>
    <span class="n">trans_df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>  <span class="o">=</span> <span class="s2">&quot;gene_ID&quot;</span>
    <span class="n">trans_df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span>  <span class="o">=</span> <span class="s2">&quot;pval&quot;</span>
    <span class="n">trans_df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span>  <span class="o">=</span> <span class="s2">&quot;beta&quot;</span>
    <span class="n">trans_df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span>  <span class="o">=</span> <span class="s2">&quot;se&quot;</span>
    <span class="n">trans_df</span> <span class="o">=</span> <span class="n">pairs_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
    <span class="n">alt</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">dataframe</span><span class="p">:</span> <span class="n">dataframe</span><span class="p">[</span><span class="s1">&#39;variant_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">variant_id</span><span class="p">:</span><span class="n">variant_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]))</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
    <span class="n">ref</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">dataframe</span><span class="p">:</span> <span class="n">dataframe</span><span class="p">[</span><span class="s1">&#39;variant_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">variant_id</span><span class="p">:</span><span class="n">variant_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">2</span><span class="p">]))</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
    <span class="n">pos</span> <span class="o">=</span> <span class="k">lambda</span> <span class="n">dataframe</span><span class="p">:</span> <span class="n">dataframe</span><span class="p">[</span><span class="s1">&#39;variant_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">variant_id</span><span class="p">:</span><span class="n">variant_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]))</span>
    <span class="n">trans_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;$[_output[0]]&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="process-of-output">
<h2>Process of output<a class="headerlink" href="#process-of-output" title="Permalink to this headline">¶</a></h2>
<p>Temp version, will change after adapting to what metal need</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">tensor_</span><span class="o">*</span><span class="n">_2</span><span class="p">,</span><span class="n">output_processing</span><span class="p">]</span>
<span class="c1"># Temp version, will change after adapting to what metal need</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.reformmated.txt&#39;</span>
        <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.trimmed.rds&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;24h&#39;</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;70G&#39;</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">step_name</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span><span class="p">,</span><span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span><span class="n">volumes</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">wd</span><span class="si">:</span><span class="s1">ad</span><span class="si">}</span><span class="s1">:</span><span class="si">{</span><span class="n">wd</span><span class="si">:</span><span class="s1">ad</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">]</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;dplyr&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;data.table&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;tibble&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;readr&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;modelr&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;purrr&quot;</span><span class="p">)</span>
    <span class="n">data</span> <span class="o">&lt;-</span> <span class="n">read_delim</span><span class="p">(</span><span class="s2">&quot;$[_input[1]]&quot;</span><span class="p">,</span><span class="n">delim</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">%&gt;%</span><span class="n">select</span><span class="p">(</span><span class="o">-</span><span class="n">gene_ID</span><span class="p">,</span><span class="o">-</span><span class="n">tss_distance</span><span class="p">,</span> <span class="o">-</span><span class="n">af</span><span class="p">,</span><span class="o">-</span><span class="n">ma_count</span><span class="p">)</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">%&gt;%</span><span class="n">group_by</span><span class="p">(</span><span class="n">variant_id</span><span class="p">)</span><span class="o">%&gt;%</span><span class="n">top_n</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="nb">abs</span><span class="p">(</span><span class="n">pval_nominal</span><span class="p">))</span><span class="o">%&gt;%</span><span class="n">mutate</span><span class="p">(</span><span class="nb">chr</span> <span class="o">=</span> <span class="n">map_dbl</span><span class="p">(</span><span class="n">variant_id</span><span class="p">,</span><span class="o">~</span><span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">text</span> <span class="o">=</span> <span class="o">.</span><span class="n">x</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;:&quot;</span><span class="p">)[,</span><span class="mi">1</span><span class="p">]),</span>
                                                                         <span class="n">res</span> <span class="o">=</span> <span class="n">map_chr</span><span class="p">(</span><span class="n">variant_id</span><span class="p">,</span><span class="o">~</span><span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">text</span> <span class="o">=</span> <span class="o">.</span><span class="n">x</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;:&quot;</span><span class="p">,</span><span class="n">colClasses</span> <span class="o">=</span> <span class="s2">&quot;character&quot;</span><span class="p">)</span><span class="err">$</span><span class="n">V2</span><span class="p">),</span>
                                                                         <span class="n">pos</span> <span class="o">=</span> <span class="n">map_dbl</span><span class="p">(</span><span class="n">res</span><span class="p">,</span><span class="o">~</span><span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">text</span> <span class="o">=</span> <span class="o">.</span><span class="n">x</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span><span class="err">$</span><span class="n">V1</span><span class="p">,</span><span class="k">as</span><span class="o">.</span><span class="ow">is</span> <span class="o">=</span> <span class="n">T</span><span class="p">),</span>
                                                                         <span class="n">ref</span> <span class="o">=</span> <span class="n">map_chr</span><span class="p">(</span><span class="n">variant_id</span><span class="p">,</span><span class="o">~</span><span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">text</span> <span class="o">=</span> <span class="o">.</span><span class="n">x</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;_&quot;</span><span class="p">,</span><span class="n">colClasses</span> <span class="o">=</span> <span class="s2">&quot;character&quot;</span><span class="p">)</span><span class="err">$</span><span class="n">V2</span><span class="p">),</span>
                                                                         <span class="n">alt</span> <span class="o">=</span> <span class="n">map_chr</span><span class="p">(</span><span class="n">variant_id</span><span class="p">,</span><span class="o">~</span><span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">text</span> <span class="o">=</span> <span class="o">.</span><span class="n">x</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;_&quot;</span><span class="p">,</span><span class="n">colClasses</span> <span class="o">=</span> <span class="s2">&quot;character&quot;</span><span class="p">)</span><span class="err">$</span><span class="n">V3</span><span class="p">))</span><span class="o">%&gt;%</span><span class="n">select</span><span class="p">(</span><span class="o">-</span><span class="n">res</span><span class="p">)</span>
    
    <span class="n">data</span> <span class="o">%&gt;%</span><span class="n">write_delim</span><span class="p">(</span><span class="s2">&quot;$[_output]&quot;</span><span class="p">,</span><span class="n">delim</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            kernelName: "sos",
            path: "./workflow/QTL_association/TensorQTL"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="../../Data_Processing/Phenotype/Region_extraction.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Gene region extraction</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="../QTL_association.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">cis/trans QTL analysis workflows</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics Consortium<br/>
        
            &copy; Copyright 2021, ADSP-FG xQTL methods and data integration subgroup.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>