
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17: http://docutils.sourceforge.net/" />

    <title>Fine-mapping with PolyFun &#8212; ADSP-FG Project</title>
    
  <link href="../../../_static/css/theme.css" rel="stylesheet">
  <link href="../../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/togglebutton.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Multivariate fine-mapping workflows" href="../multivariate_fine_mapping.html" />
    <link rel="prev" title="Fine-mapping with SuSiE RSS model" href="../SuSiE/SuSiE_RSS.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">ADSP-FG Project</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../../README.html">
   xQTL analysis pipeline
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Complete Pipelines
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../complete_analysis/eQTL_analysis.html">
   Bulk RNA-seq eQTL analysis
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../molecular_phenotypes/apa_calling.html">
   APA Calling
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../molecular_phenotypes/bulk_expression.html">
   Bulk RNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../molecular_phenotypes/scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/GWAS_QC.html">
     GWAS data QC workflow
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/GRM.html">
     Genomic relationship matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/genotype_formatting.html">
     Genotype data reformat
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/phenotype/normalization.html">
     Molecular phenotype normalization
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_formatting.html">
     Phenotype data reformat
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/covariate_preprocessing.html">
   Covariate data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/covariate/PCA.html">
     Principal component analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/covariate/PEER_factor.html">
     Factor analysis using PEER
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/covariate/BiCV_factor.html">
     Factor analysis using Bi-Cross validation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/covariate/remove_covariates.html">
     Computing covariate residuals
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../association_scan/cisQTL_scan.html">
   cisQTL analysis workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL calling pipeline
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/APEX/APEX_LMM.html">
     APEX QTL calling pipeline
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../association_scan/transQTL_scan.html">
   transQTL analysis workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../multivariate/METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/METAL/METAL.html">
     Meta-analysis with METAL
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../multivariate/MASH_pipeline.html">
   Multivariate association with MASH
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/fastqtl_to_mash.html">
     eQTL summary statistics formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/mixture_prior.html">
     A multivariate EBNM approach for mixture multivariate distribution estimate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/mashr.html">
     MASH analysis pipeline with data-driven prior matrices
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Statistical Fine-mapping
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="../univariate_fine_mapping.html">
   Univariate fine-mapping workflows
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2">
    <a class="reference internal" href="../SuSiE/SuSiE_RSS.html">
     Fine-mapping with SuSiE RSS model
    </a>
   </li>
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     Fine-mapping with PolyFun
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../multivariate_fine_mapping.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Utilities
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../misc/summary_stats_merger.html">
   Summary statistics merger
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../misc/rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH/NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../../_sources/pipeline/fine_mapping/SuSiE_Ann/polyfun.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fpipeline/fine_mapping/SuSiE_Ann/polyfun.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/pipeline/fine_mapping/SuSiE_Ann/polyfun.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#aim">
   Aim
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#methods-overview">
   Methods Overview
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#input">
   Input
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#output">
   Output
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#workflow">
   Workflow
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#step-1-compute-prior-causal-probabilities">
     Step 1: Compute Prior Causal Probabilities
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#method-1-use-precomputed-prior-causal-probabilities">
       Method 1: Use precomputed prior causal probabilities
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#method-2-compute-via-l2-regularized-extension-of-s-ldsc-preferred">
       Method 2: Compute via L2-regularized extension of S-LDSC (preferred)
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#step-2-create-functional-annotations">
     Step 2: Create functional annotations
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#method-1-use-existing-function-annotation-files">
       Method 1: Use existing function annotation files
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#method-2-create-annotations">
       Method 2: Create annotations
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#step-3-compute-ld-scores-for-annotations">
     Step 3: Compute LD-scores for annotations
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#method-1-compute-with-reference-panel-of-sequenced-individuals">
       Method 1: Compute with reference panel of sequenced individuals
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#method-2-compute-with-pre-computed-uk-biobank-ld-matrices">
       Method 2: Compute with pre-computed UK Biobank LD matrices
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#method-3-compute-with-own-pre-computed-ld-matrices">
       Method 3: Compute with own pre-computed LD matrices
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#step-4-run-polyfun-with-l2-regularized-s-ldsc">
     Step 4: Run PolyFun with L2-regularized S-LDSC
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#minimal-working-example">
   Minimal Working Example
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#summary">
     Summary
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#summary-of-fine-mapping-result-without-functional-annotations">
     Summary of Fine-mapping Result Without Functional Annotations
    </a>
   </li>
  </ul>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="fine-mapping-with-polyfun">
<h1>Fine-mapping with PolyFun<a class="headerlink" href="#fine-mapping-with-polyfun" title="Permalink to this headline">¶</a></h1>
<section id="aim">
<h2>Aim<a class="headerlink" href="#aim" title="Permalink to this headline">¶</a></h2>
<p>The purpose of this notebook is to demonstrate a functionally-informed fine-mapping workflow using the PolyFun method.</p>
</section>
<section id="methods-overview">
<h2>Methods Overview<a class="headerlink" href="#methods-overview" title="Permalink to this headline">¶</a></h2>
</section>
<section id="input">
<h2>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h2>
<ol class="simple">
<li><p>GWAS summary statistics including the following variables:</p></li>
</ol>
<ul class="simple">
<li><p>variant_id - variant ID</p></li>
<li><p>P - p-value</p></li>
<li><p>CHR - chromosome number</p></li>
<li><p>BP - base pair position</p></li>
<li><p>A1 - The effect allele (i.e., the sign of the effect size is with respect to A1)</p></li>
<li><p>A2 - the second allele</p></li>
<li><p>MAF - minor allele frequency</p></li>
<li><p>BETA - effect size</p></li>
<li><p>SE - effect size standard error</p></li>
</ul>
<ol>
<li><p>SNP-identifier file or S-LDSC (stratified LD-score regression) LD-score and annotation file</p>
<p>SNP-identifier file should include the following columns:</p>
</li>
</ol>
<ul class="simple">
<li><p>CHR - chromosome</p></li>
<li><p>BP - base pair position (in hg19 coordinates)</p></li>
<li><p>A1 - The effect allele</p></li>
<li><p>A2 - the second allele</p></li>
</ul>
<ol class="simple">
<li><p>Ld-score weights file</p></li>
</ol>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h2>
<p>A <code class="docutils literal notranslate"><span class="pre">.gz</span></code> file containing input summary statistics columns and additionally the following columns:</p>
<ul class="simple">
<li><p>PIP - posterior causal probability</p></li>
<li><p>BETA_MEAN - posterior mean of causal effect size (in standardized genotype scale)</p></li>
<li><p>BETA_SD - posterior standard deviation of causal effect size (in standardized genotype scale)</p></li>
<li><p>CREDIBLE_SET - the index of the first (typically smallest) credible set that the SNP belongs to (0 means none).</p></li>
</ul>
</section>
<section id="workflow">
<h2>Workflow<a class="headerlink" href="#workflow" title="Permalink to this headline">¶</a></h2>
<section id="step-1-compute-prior-causal-probabilities">
<h3>Step 1: Compute Prior Causal Probabilities<a class="headerlink" href="#step-1-compute-prior-causal-probabilities" title="Permalink to this headline">¶</a></h3>
<section id="method-1-use-precomputed-prior-causal-probabilities">
<h4>Method 1: Use precomputed prior causal probabilities<a class="headerlink" href="#method-1-use-precomputed-prior-causal-probabilities" title="Permalink to this headline">¶</a></h4>
<p>Use precomputed prior causal probabilities of 19 million imputed UK Biobank SNPs with MAF&gt;0.1%, based on a meta-analysis of 15 UK Biobank traits.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">prior_causal_prob</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">sumstats</span> <span class="o">=</span> 
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="s1">&#39;/mnt/mfs/statgen/tl3030/SIF/polyfun_ninth.sif&#39;</span>
    <span class="n">mkdir</span> <span class="o">-</span><span class="n">p</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">tl3030</span><span class="o">/</span><span class="n">AD_output</span>
    <span class="n">python</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">tl3030</span><span class="o">/</span><span class="n">extract_snpvar</span><span class="o">.</span><span class="n">py</span> \
        <span class="o">--</span><span class="n">sumstats</span> <span class="n">AD_sumstats_Jansenetal_2019sept</span><span class="o">.</span><span class="n">txt</span><span class="o">.</span><span class="n">gz</span> \
        <span class="o">--</span><span class="n">out</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">tl3030</span><span class="o">/</span><span class="n">AD_output</span><span class="o">/</span><span class="n">AD_snps_with_var</span><span class="o">.</span><span class="n">gz</span> \
        <span class="o">--</span><span class="n">allow</span><span class="o">-</span><span class="n">missing</span>
    <span class="n">cat</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">tl3030</span><span class="o">/</span><span class="n">AD_output</span><span class="o">/</span><span class="n">AD_snps_with_var</span><span class="o">.</span><span class="n">gz</span> <span class="o">|</span> <span class="n">zcat</span> <span class="o">|</span> <span class="n">head</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="method-2-compute-via-l2-regularized-extension-of-s-ldsc-preferred">
<h4>Method 2: Compute via L2-regularized extension of S-LDSC (preferred)<a class="headerlink" href="#method-2-compute-via-l2-regularized-extension-of-s-ldsc-preferred" title="Permalink to this headline">¶</a></h4>
<p>Compute via an L2-regularized extension of stratified LD-score regression (S-LDSC). Procedure for both methods is shown in this workflow.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">munged_sumstats</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">sumstats</span> <span class="o">=</span> 
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="s1">&#39;/mnt/mfs/statgen/tl3030/SIF/polyfun_ninth.sif&#39;</span>
    <span class="n">mkdir</span> <span class="o">-</span><span class="n">p</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">tl3030</span><span class="o">/</span><span class="n">AD_SLDSC_output</span>
    <span class="n">python</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">tl3030</span><span class="o">/</span><span class="n">munge_polyfun_sumstats</span><span class="o">.</span><span class="n">py</span> \
      <span class="o">--</span><span class="n">sumstats</span> <span class="n">AD_sumstats_Jansenetal_2019sept</span><span class="o">.</span><span class="n">txt</span><span class="o">.</span><span class="n">gz</span> \
      <span class="o">--</span><span class="n">n</span> <span class="mi">450734</span> \
      <span class="o">--</span><span class="n">out</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">tl3030</span><span class="o">/</span><span class="n">AD_SLDSC_output</span><span class="o">/</span><span class="n">sumstats_munged</span><span class="o">.</span><span class="n">parquet</span> \
      <span class="o">--</span><span class="nb">min</span><span class="o">-</span><span class="n">info</span> <span class="mi">0</span> \
      <span class="o">--</span><span class="nb">min</span><span class="o">-</span><span class="n">maf</span> <span class="mi">0</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="step-2-create-functional-annotations">
<h3>Step 2: Create functional annotations<a class="headerlink" href="#step-2-create-functional-annotations" title="Permalink to this headline">¶</a></h3>
<section id="method-1-use-existing-function-annotation-files">
<h4>Method 1: Use existing function annotation files<a class="headerlink" href="#method-1-use-existing-function-annotation-files" title="Permalink to this headline">¶</a></h4>
<p>Use functional annotations for ~19 million UK Biobank imputed SNPs with MAF&gt;0.1%, based on the baseline-LF 2.2.UKB annotations</p>
<p>Download (30G): <a class="reference external" href="https://data.broadinstitute.org/alkesgroup/LDSCORE/baselineLF_v2.2.UKB.polyfun.tar.gz">https://data.broadinstitute.org/alkesgroup/LDSCORE/baselineLF_v2.2.UKB.polyfun.tar.gz</a></p>
</section>
<section id="method-2-create-annotations">
<h4>Method 2: Create annotations<a class="headerlink" href="#method-2-create-annotations" title="Permalink to this headline">¶</a></h4>
<p>To create your own annotations, for each chromosome, the following files are needed:</p>
<ol class="simple">
<li><p>A <code class="docutils literal notranslate"><span class="pre">.gz</span></code> or <code class="docutils literal notranslate"><span class="pre">.parquet</span></code> Annotations file containing the following columns:</p></li>
</ol>
<ul class="simple">
<li><p>CHR - chromosome number</p></li>
<li><p>BP base pair position</p></li>
<li><p>SNP - dbSNP reference number</p></li>
<li><p>A1 - The effect allele</p></li>
<li><p>A2 - the second allele</p></li>
<li><p>Arbitrary additional columns representing annotations</p></li>
</ul>
<ol class="simple">
<li><p>A <code class="docutils literal notranslate"><span class="pre">.l2.M</span></code> white-space delimited file containing a single line with the sums of the columns of each annotation</p></li>
<li><p>(Optional) A <code class="docutils literal notranslate"><span class="pre">l2.M_5_50</span></code> file that is the <code class="docutils literal notranslate"><span class="pre">.l2.M</span></code> but only containing common SNPS (MAF between 5% and 50%)</p></li>
</ol>
</section>
</section>
<section id="step-3-compute-ld-scores-for-annotations">
<h3>Step 3: Compute LD-scores for annotations<a class="headerlink" href="#step-3-compute-ld-scores-for-annotations" title="Permalink to this headline">¶</a></h3>
<section id="method-1-compute-with-reference-panel-of-sequenced-individuals">
<h4>Method 1: Compute with reference panel of sequenced individuals<a class="headerlink" href="#method-1-compute-with-reference-panel-of-sequenced-individuals" title="Permalink to this headline">¶</a></h4>
<p>Reference panel should have at least 3000 sequenced individuals from target population.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">ld_score</span><span class="p">]</span>
<span class="kn">bash:</span> <span class="n">container</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
    <span class="n">mkdir</span> <span class="o">-</span><span class="n">p</span>
    <span class="n">python</span> <span class="n">compute_ldscores</span><span class="o">.</span><span class="n">py</span> \
    <span class="o">--</span><span class="n">bfile</span> <span class="n">example_data</span><span class="o">/</span><span class="n">reference</span><span class="mf">.1</span> \
    <span class="o">--</span><span class="n">annot</span> <span class="n">example_data</span><span class="o">/</span><span class="n">annotations</span><span class="mf">.1</span><span class="o">.</span><span class="n">annot</span><span class="o">.</span><span class="n">parquet</span> \
    <span class="o">--</span><span class="n">out</span> <span class="n">output</span><span class="o">/</span><span class="n">ldscores_example</span><span class="o">.</span><span class="n">parquet</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="method-2-compute-with-pre-computed-uk-biobank-ld-matrices">
<h4>Method 2: Compute with pre-computed UK Biobank LD matrices<a class="headerlink" href="#method-2-compute-with-pre-computed-uk-biobank-ld-matrices" title="Permalink to this headline">¶</a></h4>
<p>Matrices download: <a class="reference external" href="https://data.broadinstitute.org/alkesgroup/UKBB_LD">https://data.broadinstitute.org/alkesgroup/UKBB_LD</a></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">ld_score_uk</span><span class="p">]</span>
<span class="kn">base:</span> <span class="n">container</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
    <span class="n">mkdir</span> <span class="o">-</span><span class="n">p</span> 
    <span class="n">python</span> <span class="n">compute_ldscores_from_ld</span><span class="o">.</span><span class="n">py</span> \
    <span class="o">--</span><span class="n">annot</span> <span class="n">example_data</span><span class="o">/</span><span class="n">annotations</span><span class="mf">.1</span><span class="o">.</span><span class="n">annot</span><span class="o">.</span><span class="n">parquet</span> \
    <span class="o">--</span><span class="n">ukb</span> \
    <span class="o">--</span><span class="n">out</span> <span class="n">output</span><span class="o">/</span><span class="n">ldscores_example2</span><span class="o">.</span><span class="n">parquet</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="method-3-compute-with-own-pre-computed-ld-matrices">
<h4>Method 3: Compute with own pre-computed LD matrices<a class="headerlink" href="#method-3-compute-with-own-pre-computed-ld-matrices" title="Permalink to this headline">¶</a></h4>
<p>Own pre-computed LD matrices should be in <code class="docutils literal notranslate"><span class="pre">.bcor</span></code> format.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">ld_score_own</span><span class="p">]</span>
<span class="kn">base:</span> <span class="n">container</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
    <span class="n">mkdir</span> <span class="o">-</span><span class="n">p</span> 
    <span class="n">python</span> <span class="n">compute_ldscores_from_ld</span><span class="o">.</span><span class="n">py</span> \
    <span class="o">--</span><span class="n">annot</span> <span class="n">example_data</span><span class="o">/</span><span class="n">annotations</span><span class="mf">.1</span><span class="o">.</span><span class="n">annot</span><span class="o">.</span><span class="n">parquet</span> \
    <span class="o">--</span><span class="n">out</span> <span class="n">output</span><span class="o">/</span><span class="n">ldscores_example3</span><span class="o">.</span><span class="n">parquet</span> \
    <span class="o">--</span><span class="n">n</span> <span class="mi">10000</span> \
    <span class="n">bcor_files</span><span class="o">/*.</span><span class="n">bcor</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="step-4-run-polyfun-with-l2-regularized-s-ldsc">
<h3>Step 4: Run PolyFun with L2-regularized S-LDSC<a class="headerlink" href="#step-4-run-polyfun-with-l2-regularized-s-ldsc" title="Permalink to this headline">¶</a></h3>
<p>If prior causal probabilities aren’t computed,then use <code class="docutils literal notranslate"><span class="pre">finemapper.py</span></code> instead of <code class="docutils literal notranslate"><span class="pre">polyfun.py</span></code> to perform non-functionally-informed fine-mapping.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">L2_regu_SLDSC</span><span class="p">]</span>
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="s1">&#39;/mnt/mfs/statgen/tl3030/SIF/polyfun_ninth.sif&#39;</span>
    <span class="n">python</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">tl3030</span><span class="o">/</span><span class="n">polyfun</span><span class="o">.</span><span class="n">py</span> \
    <span class="o">--</span><span class="n">compute</span><span class="o">-</span><span class="n">h2</span><span class="o">-</span><span class="n">L2</span> \
    <span class="o">--</span><span class="n">no</span><span class="o">-</span><span class="n">partitions</span> \
    <span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="n">prefix</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">tl3030</span><span class="o">/</span><span class="n">AD_SLDSC_output</span><span class="o">/</span><span class="n">testrun</span> \
    <span class="o">--</span><span class="n">sumstats</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">tl3030</span><span class="o">/</span><span class="n">AD_SLDSC_output</span><span class="o">/</span><span class="n">sumstats_munged</span><span class="o">.</span><span class="n">parquet</span> \
    <span class="o">--</span><span class="n">ref</span><span class="o">-</span><span class="n">ld</span><span class="o">-</span><span class="nb">chr</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">tl3030</span><span class="o">/</span><span class="n">baselineLF2</span><span class="mf">.2</span><span class="o">.</span><span class="n">UKB</span><span class="o">/</span><span class="n">baselineLF2</span><span class="mf">.2</span><span class="o">.</span><span class="n">UKB</span><span class="o">.</span> \
    <span class="o">--</span><span class="n">w</span><span class="o">-</span><span class="n">ld</span><span class="o">-</span><span class="nb">chr</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">tl3030</span><span class="o">/</span><span class="n">weights</span><span class="o">.</span><span class="n">UKB</span><span class="o">.</span><span class="n">l2</span><span class="o">.</span><span class="n">ldscore</span><span class="o">/</span><span class="n">weights</span><span class="o">.</span><span class="n">UKB</span><span class="o">.</span> \
    <span class="o">--</span><span class="n">allow</span><span class="o">-</span><span class="n">missing</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="minimal-working-example">
<h2>Minimal Working Example<a class="headerlink" href="#minimal-working-example" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">module</span> <span class="n">load</span> <span class="n">Singularity</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">cat</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">tl3030</span><span class="o">/</span><span class="n">GCST90012877_buildGRCh37</span><span class="o">.</span><span class="n">tsv</span><span class="o">.</span><span class="n">gz</span> <span class="o">|</span> <span class="n">zcat</span> <span class="o">|</span> <span class="n">head</span> <span class="o">-</span><span class="mi">2</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">modification</span><span class="p">]</span>

<span class="kn">python:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;${ }&quot;</span>
    <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
    <span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>


    <span class="c1"># Read in data</span>
    <span class="n">sumstat</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;/mnt/mfs/statgen/tl3030/GCST90012877_buildGRCh37.tsv&#39;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

    <span class="c1"># Rename columns so that munge_polyfun_sumstats.py could recognize it</span>
    <span class="n">sumstat</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;SNP_ID&#39;</span><span class="p">:</span> <span class="s1">&#39;SNP&#39;</span><span class="p">,</span> <span class="s1">&#39;chromosome&#39;</span><span class="p">:</span> <span class="s1">&#39;CHR&#39;</span><span class="p">,</span> <span class="s1">&#39;base_pair_location&#39;</span><span class="p">:</span> <span class="s1">&#39;BP&#39;</span><span class="p">,</span> 
                        <span class="s1">&#39;effect_allele&#39;</span><span class="p">:</span> <span class="s1">&#39;A1&#39;</span><span class="p">,</span> <span class="s1">&#39;other_allele&#39;</span><span class="p">:</span> <span class="s1">&#39;A2&#39;</span><span class="p">,</span> <span class="s1">&#39;effect_allele_frequency&#39;</span><span class="p">:</span> <span class="s1">&#39;MAF&#39;</span><span class="p">,</span> 
                        <span class="s1">&#39;p_value&#39;</span><span class="p">:</span> <span class="s1">&#39;P&#39;</span><span class="p">,</span> <span class="s1">&#39;standard_error&#39;</span><span class="p">:</span> <span class="s1">&#39;SE&#39;</span><span class="p">},</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">errors</span><span class="o">=</span><span class="s1">&#39;raise&#39;</span><span class="p">)</span>

    <span class="c1"># Replace NaN with string `chr_bp_ref_alt`</span>
    <span class="n">cols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;CHR&#39;</span><span class="p">,</span> <span class="s1">&#39;BP&#39;</span><span class="p">,</span> <span class="s1">&#39;A2&#39;</span><span class="p">,</span> <span class="s1">&#39;A1&#39;</span><span class="p">]</span>

    <span class="n">sumstat</span><span class="p">[</span><span class="s1">&#39;variant_id&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">sumstat</span><span class="p">[</span><span class="s1">&#39;variant_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="s1">&#39;refill&#39;</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">k</span><span class="p">,</span><span class="n">row</span> <span class="ow">in</span> <span class="n">sumstat</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;variant_id&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;refill&#39;</span><span class="p">:</span>
            <span class="n">replace</span> <span class="o">=</span> <span class="s1">&#39;_&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="n">cols</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">))</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">k</span><span class="p">)</span>
            <span class="n">sumstat</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">k</span><span class="p">,</span> <span class="s1">&#39;variant_id&#39;</span><span class="p">]</span><span class="o">=</span><span class="n">replace</span> 

    <span class="c1"># Write out data in a .txt file</span>
    <span class="n">sumstat</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;/mnt/mfs/statgen/tl3030/GCST90012877_buildGRCh37_colrenamed.txt&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;w&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">munged_sumstats</span><span class="p">]</span>
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="s1">&#39;/mnt/mfs/statgen/tl3030/SIF/polyfun_ninth.sif&#39;</span>
    <span class="n">mkdir</span> <span class="o">-</span><span class="n">p</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">tl3030</span><span class="o">/</span><span class="n">AD_2021_output</span>
    <span class="n">python</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">tl3030</span><span class="o">/</span><span class="n">munge_polyfun_sumstats</span><span class="o">.</span><span class="n">py</span> \
      <span class="o">--</span><span class="n">sumstats</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">tl3030</span><span class="o">/</span><span class="n">GCST90012877_buildGRCh37_colrenamed</span><span class="o">.</span><span class="n">txt</span><span class="o">.</span><span class="n">gz</span> \
      <span class="o">--</span><span class="n">n</span> <span class="mi">472868</span> \
      <span class="o">--</span><span class="n">out</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">tl3030</span><span class="o">/</span><span class="n">AD_2021_output</span><span class="o">/</span><span class="n">sumstats_munged</span><span class="o">.</span><span class="n">parquet</span> \
      <span class="o">--</span><span class="nb">min</span><span class="o">-</span><span class="n">info</span> <span class="mf">0.6</span> \
      <span class="o">--</span><span class="nb">min</span><span class="o">-</span><span class="n">maf</span> <span class="mf">0.001</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">tl3030</span><span class="o">/</span><span class="n">Untitled1</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">munged_sumstats</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="n">munged_file</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_parquet</span><span class="p">(</span><span class="s1">&#39;/mnt/mfs/statgen/tl3030/AD_2021_output/sumstats_munged.parquet&#39;</span><span class="p">,</span> <span class="n">engine</span><span class="o">=</span><span class="s1">&#39;auto&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">munged_file</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">L2_regu_SLDSC</span><span class="p">]</span>
<span class="kn">bash:</span> <span class="n">container</span><span class="o">=</span><span class="s1">&#39;/mnt/mfs/statgen/tl3030/SIF/polyfun_ninth.sif&#39;</span>
    <span class="n">python</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">tl3030</span><span class="o">/</span><span class="n">polyfun</span><span class="o">.</span><span class="n">py</span> \
    <span class="o">--</span><span class="n">compute</span><span class="o">-</span><span class="n">h2</span><span class="o">-</span><span class="n">L2</span> \
    <span class="o">--</span><span class="n">no</span><span class="o">-</span><span class="n">partitions</span> \
    <span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="n">prefix</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">tl3030</span><span class="o">/</span><span class="n">AD_2021_output</span><span class="o">/</span><span class="n">testrun</span> \
    <span class="o">--</span><span class="n">sumstats</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">tl3030</span><span class="o">/</span><span class="n">AD_2021_output</span><span class="o">/</span><span class="n">sumstats_munged</span><span class="o">.</span><span class="n">parquet</span> \
    <span class="o">--</span><span class="n">ref</span><span class="o">-</span><span class="n">ld</span><span class="o">-</span><span class="nb">chr</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">tl3030</span><span class="o">/</span><span class="n">baselineLF2</span><span class="mf">.2</span><span class="o">.</span><span class="n">UKB</span><span class="o">/</span><span class="n">baselineLF2</span><span class="mf">.2</span><span class="o">.</span><span class="n">UKB</span><span class="o">.</span> \
    <span class="o">--</span><span class="n">w</span><span class="o">-</span><span class="n">ld</span><span class="o">-</span><span class="nb">chr</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">tl3030</span><span class="o">/</span><span class="n">weights</span><span class="o">.</span><span class="n">UKB</span><span class="o">.</span><span class="n">l2</span><span class="o">.</span><span class="n">ldscore</span><span class="o">/</span><span class="n">weights</span><span class="o">.</span><span class="n">UKB</span><span class="o">.</span> \
    <span class="o">--</span><span class="n">allow</span><span class="o">-</span><span class="n">missing</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">tl3030</span><span class="o">/</span><span class="n">Untitled1</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">L2_regu_SLDSC</span>
</pre></div>
</div>
</div>
</div>
<section id="summary">
<h3>Summary<a class="headerlink" href="#summary" title="Permalink to this headline">¶</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os.path</span>
<span class="c1"># get the location of finemapping result files</span>
<span class="n">file_with_annot_location</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="s1">&#39;/mnt&#39;</span><span class="p">,</span> <span class="s1">&#39;mfs&#39;</span><span class="p">,</span> <span class="s1">&#39;statgen&#39;</span><span class="p">,</span><span class="s1">&#39;tl3030&#39;</span><span class="p">,</span><span class="s1">&#39;AD_2021_output&#39;</span><span class="p">,</span><span class="s1">&#39;with_annot&#39;</span><span class="p">,</span> <span class="s1">&#39;finemap.*.gz&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">file_with_annot_location</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">glob</span>
<span class="c1"># get a list of result file name</span>
<span class="n">filenames_with_annot</span> <span class="o">=</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="n">file_with_annot_location</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">filenames_with_annot</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">filenames_with_annot</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">snp_with_annot</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>

<span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">filenames_with_annot</span><span class="p">:</span>
    <span class="c1"># read the data</span>
    <span class="n">outfile</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="n">delimiter</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    
    <span class="c1"># filter out SNPs that has PIP &gt;= 0.95</span>
    <span class="n">significant</span> <span class="o">=</span> <span class="p">(</span><span class="n">outfile</span><span class="p">[</span><span class="n">outfile</span><span class="p">[</span><span class="s1">&#39;PIP&#39;</span><span class="p">]</span><span class="o">&gt;=</span><span class="mf">0.95</span><span class="p">])</span>
    <span class="n">snp_with_annot</span> <span class="o">=</span> <span class="n">snp_with_annot</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">significant</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">snp_with_annot</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># define a function to check duplicate</span>
<span class="k">def</span> <span class="nf">checkIfDuplicates</span><span class="p">(</span><span class="n">snp</span><span class="p">):</span>
    <span class="sd">&#39;&#39;&#39; Check if given list contains any duplicates &#39;&#39;&#39;</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">snp</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">snp</span><span class="p">)):</span>
        <span class="k">return</span> <span class="kc">False</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">True</span>

<span class="c1"># check if there is duplicates</span>
<span class="n">result</span> <span class="o">=</span> <span class="n">checkIfDuplicates</span><span class="p">(</span><span class="n">snp</span><span class="p">)</span>
<span class="k">if</span> <span class="n">result</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Yes, list contains duplicates&#39;</span><span class="p">)</span>
<span class="kn">else:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;No duplicates found in list&#39;</span><span class="p">)</span> 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># remove duplicated SNPs</span>
<span class="n">snp_with_annot_uniq</span> <span class="o">=</span> <span class="n">snp_with_annot</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="s1">&#39;SNP&#39;</span><span class="p">,</span> <span class="n">keep</span><span class="o">=</span><span class="s1">&#39;first&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">snp_with_annot_uniq</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">snp_with_annot_uniq</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">CS_with_annot</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>

<span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">filenames_with_annot</span><span class="p">:</span>
    <span class="c1"># read the data</span>
    <span class="n">outfile</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="n">delimiter</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    
    <span class="c1"># filter out SNPs that has CS</span>
    <span class="n">significant</span> <span class="o">=</span> <span class="p">(</span><span class="n">outfile</span><span class="p">[</span><span class="n">outfile</span><span class="p">[</span><span class="s1">&#39;CREDIBLE_SET&#39;</span><span class="p">]</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">CS_with_annot</span> <span class="o">=</span> <span class="n">CS_with_annot</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">significant</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">CS_with_annot</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">CS_with_annot</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># remove duplicated SNPs</span>
<span class="n">CS_with_annot_uniq</span> <span class="o">=</span> <span class="n">CS_with_annot</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="s1">&#39;SNP&#39;</span><span class="p">,</span> <span class="n">keep</span><span class="o">=</span><span class="s1">&#39;first&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">CS_with_annot_uniq</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">CS_with_annot_uniq</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">pd</span><span class="o">.</span><span class="n">options</span><span class="o">.</span><span class="n">mode</span><span class="o">.</span><span class="n">chained_assignment</span> <span class="o">=</span> <span class="kc">None</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># Read in the range file</span>
<span class="n">region_range</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;/mnt/mfs/statgen/tl3030/range.csv&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span>
<span class="c1">#chr1_160990767_161203192 = pd.read_csv(&quot;/mnt/mfs/statgen/tl3030/finemapping_result_97gene/finemap.1.160990767.161203192.gz&quot;, delimiter = &quot;\t&quot;)</span>
<span class="c1">#print(chr1_160990767_161203192.head())</span>

<span class="n">bpcol</span> <span class="o">=</span> <span class="n">CS_with_annot_uniq</span><span class="p">[[</span><span class="s1">&#39;CHR&#39;</span><span class="p">,</span> <span class="s1">&#39;BP&#39;</span><span class="p">]]</span>
<span class="c1">#print(bpcol.head())</span>

<span class="c1"># Assign SNPs to the gene region that it belong to</span>
<span class="n">j</span> <span class="o">=</span> <span class="mi">0</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">bp</span> <span class="ow">in</span> <span class="n">bpcol</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
    <span class="c1">#print(i, bp[&#39;CHR&#39;])</span>
    <span class="k">for</span> <span class="n">k</span><span class="p">,</span><span class="n">row</span> <span class="ow">in</span> <span class="n">region_range</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">bp</span><span class="p">[</span><span class="s1">&#39;CHR&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Chr&#39;</span><span class="p">])</span> <span class="ow">and</span> <span class="p">(</span><span class="n">bp</span><span class="p">[</span><span class="s1">&#39;BP&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">])</span> <span class="ow">and</span> <span class="p">(</span><span class="n">bp</span><span class="p">[</span><span class="s1">&#39;BP&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]):</span>
            <span class="c1">#print(row[&#39;Chr&#39;],row[&#39;Gene Name&#39;])</span>
            <span class="c1">#print(i, bp[&#39;CHR&#39;], row[&#39;Chr&#39;], row[&#39;Gene Name&#39;])</span>
            <span class="n">CS_with_annot_uniq</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">j</span><span class="p">,</span><span class="mi">15</span><span class="p">]</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Gene Name&#39;</span><span class="p">]</span>
            <span class="c1">#pass</span>
            <span class="c1">#CS_with_annot_uniq.loc[j,&#39;GENE&#39;]= row[&#39;Gene Name&#39;]</span>
    <span class="n">j</span> <span class="o">+=</span> <span class="mi">1</span>

<span class="nb">print</span><span class="p">(</span><span class="n">CS_with_annot_uniq</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">CS_with_annot_uniq</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">CS_with_annot_uniq</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;/mnt/mfs/statgen/tl3030/AD_2021_output/variants_with_CS_2021sumstat_97genes_with_annot.txt&#39;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;w&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">CS_with_annot_uniq</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CHR&#39;</span><span class="p">])</span> <span class="c1"># sort the file by chromosome</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">num_of_CS_with_annot</span> <span class="o">=</span> <span class="n">CS_with_annot_uniq</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CREDIBLE_SET&#39;</span><span class="p">,</span> <span class="s1">&#39;GENE&#39;</span><span class="p">],</span> <span class="n">keep</span> <span class="o">=</span> <span class="s1">&#39;last&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">num_of_CS_with_annot</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="n">num_of_CS_with_annot</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;GENE&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">num_of_gene_with_annot</span> <span class="o">=</span> <span class="n">CS_with_annot_uniq</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;GENE&#39;</span><span class="p">],</span> <span class="n">keep</span> <span class="o">=</span> <span class="s1">&#39;last&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">num_of_gene_with_annot</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">check_frequency_with_annot</span> <span class="o">=</span> <span class="n">CS_with_annot_uniq</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;CREDIBLE_SET&quot;</span><span class="p">,</span> <span class="s2">&quot;GENE&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="kp">name</span><span class="o">=</span><span class="s2">&quot;Time&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">check_frequency_with_annot</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">check_frequency_with_annot</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">check_frequency_with_annot</span><span class="p">[</span><span class="s1">&#39;Time&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">())</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">CS_with_1_variant_with_annot</span> <span class="o">=</span> <span class="n">check_frequency_with_annot</span><span class="p">[</span><span class="n">check_frequency_with_annot</span><span class="p">[</span><span class="s1">&#39;Time&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">CS_with_1_variant_with_annot</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">CS_with_1_variant_with_annot</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">gene_with_annot</span> <span class="o">=</span> <span class="n">CS_with_annot_uniq</span><span class="p">[</span><span class="s1">&#39;GENE&#39;</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">gene_with_annot</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>

<span class="n">gene_list_with_annot</span> <span class="o">=</span> <span class="n">gene_with_annot</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">gene_list_with_annot</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">gene_list_with_annot</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="summary-of-fine-mapping-result-without-functional-annotations">
<h3>Summary of Fine-mapping Result Without Functional Annotations<a class="headerlink" href="#summary-of-fine-mapping-result-without-functional-annotations" title="Permalink to this headline">¶</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os.path</span>
<span class="c1"># get the location of finemapping result files</span>
<span class="n">file_without_annot_location</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="s1">&#39;/mnt&#39;</span><span class="p">,</span> <span class="s1">&#39;mfs&#39;</span><span class="p">,</span> <span class="s1">&#39;statgen&#39;</span><span class="p">,</span><span class="s1">&#39;tl3030&#39;</span><span class="p">,</span><span class="s1">&#39;AD_2021_output&#39;</span><span class="p">,</span><span class="s1">&#39;without_annot&#39;</span><span class="p">,</span> <span class="s1">&#39;finemap.*.gz&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">file_without_annot_location</span><span class="p">)</span>

<span class="kn">import</span> <span class="nn">glob</span>
<span class="c1"># get a list of result file name</span>
<span class="n">filenames_without_annot</span> <span class="o">=</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="n">file_without_annot_location</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">filenames_without_annot</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">filenames_without_annot</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">snp_without_annot</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>

<span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">filenames_without_annot</span><span class="p">:</span>
    <span class="c1"># read the data</span>
    <span class="n">outfile</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="n">delimiter</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    
    <span class="c1"># filter out SNPs that has PIP &gt;= 0.95</span>
    <span class="n">significant</span> <span class="o">=</span> <span class="p">(</span><span class="n">outfile</span><span class="p">[</span><span class="n">outfile</span><span class="p">[</span><span class="s1">&#39;PIP&#39;</span><span class="p">]</span><span class="o">&gt;=</span><span class="mf">0.95</span><span class="p">])</span>
    <span class="n">snp_without_annot</span> <span class="o">=</span> <span class="n">snp_without_annot</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">significant</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">snp_without_annot</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">))</span>

<span class="c1"># remove duplicated SNPs</span>
<span class="n">snp_without_annot_uniq</span> <span class="o">=</span> <span class="n">snp_without_annot</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="s1">&#39;SNP&#39;</span><span class="p">,</span> <span class="n">keep</span><span class="o">=</span><span class="s1">&#39;first&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">snp_without_annot_uniq</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">snp_without_annot_uniq</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">CS_without_annot</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>

<span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">filenames_without_annot</span><span class="p">:</span>
    <span class="c1"># read the data</span>
    <span class="n">outfile</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="n">delimiter</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    
    <span class="c1"># filter out SNPs that has CS</span>
    <span class="n">significant</span> <span class="o">=</span> <span class="p">(</span><span class="n">outfile</span><span class="p">[</span><span class="n">outfile</span><span class="p">[</span><span class="s1">&#39;CREDIBLE_SET&#39;</span><span class="p">]</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">CS_without_annot</span> <span class="o">=</span> <span class="n">CS_without_annot</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">significant</span><span class="p">)</span>

<span class="c1"># remove duplicated SNPs</span>
<span class="n">CS_without_annot_uniq</span> <span class="o">=</span> <span class="n">CS_without_annot</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="s1">&#39;SNP&#39;</span><span class="p">,</span> <span class="n">keep</span><span class="o">=</span><span class="s1">&#39;first&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">CS_without_annot_uniq</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">CS_without_annot_uniq</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># Read in the range file</span>
<span class="n">region_range</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;/mnt/mfs/statgen/tl3030/range.csv&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span>

<span class="n">bpcol</span> <span class="o">=</span> <span class="n">CS_without_annot_uniq</span><span class="p">[[</span><span class="s1">&#39;CHR&#39;</span><span class="p">,</span> <span class="s1">&#39;BP&#39;</span><span class="p">]]</span>

<span class="c1"># Assign SNPs to the gene region that it belong to</span>
<span class="n">j</span> <span class="o">=</span> <span class="mi">0</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">bp</span> <span class="ow">in</span> <span class="n">bpcol</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
    <span class="c1">#print(i, bp[&#39;CHR&#39;])</span>
    <span class="k">for</span> <span class="n">k</span><span class="p">,</span><span class="n">row</span> <span class="ow">in</span> <span class="n">region_range</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">bp</span><span class="p">[</span><span class="s1">&#39;CHR&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Chr&#39;</span><span class="p">])</span> <span class="ow">and</span> <span class="p">(</span><span class="n">bp</span><span class="p">[</span><span class="s1">&#39;BP&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">])</span> <span class="ow">and</span> <span class="p">(</span><span class="n">bp</span><span class="p">[</span><span class="s1">&#39;BP&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]):</span>
            <span class="n">CS_without_annot_uniq</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">j</span><span class="p">,</span><span class="mi">15</span><span class="p">]</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Gene Name&#39;</span><span class="p">]</span>
            <span class="c1">#CS_without_annot_uniq.loc[j,&#39;GENE&#39;]= row[&#39;Gene Name&#39;]</span>
    <span class="n">j</span> <span class="o">+=</span> <span class="mi">1</span>

<span class="nb">print</span><span class="p">(</span><span class="n">CS_without_annot_uniq</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">CS_without_annot_uniq</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">num_of_CS_without_annot</span> <span class="o">=</span> <span class="n">CS_without_annot_uniq</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;CREDIBLE_SET&#39;</span><span class="p">,</span> <span class="s1">&#39;GENE&#39;</span><span class="p">],</span> <span class="n">keep</span> <span class="o">=</span> <span class="s1">&#39;last&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">num_of_CS_without_annot</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="n">num_of_CS_without_annot</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;GENE&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">num_of_gene_without_annot</span> <span class="o">=</span> <span class="n">CS_without_annot_uniq</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;GENE&#39;</span><span class="p">],</span> <span class="n">keep</span> <span class="o">=</span> <span class="s1">&#39;last&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">num_of_gene_without_annot</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">check_frequency_without_annot</span> <span class="o">=</span> <span class="n">CS_without_annot_uniq</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s2">&quot;CREDIBLE_SET&quot;</span><span class="p">,</span> <span class="s2">&quot;GENE&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="kp">name</span><span class="o">=</span><span class="s2">&quot;Time&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">check_frequency_without_annot</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">check_frequency_without_annot</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">check_frequency_without_annot</span><span class="p">[</span><span class="s1">&#39;Time&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">())</span>

<span class="n">CS_with_1_variant_without_annot</span> <span class="o">=</span> <span class="n">check_frequency_without_annot</span><span class="p">[</span><span class="n">check_frequency_without_annot</span><span class="p">[</span><span class="s1">&#39;Time&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">CS_with_1_variant_without_annot</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">CS_with_1_variant_without_annot</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">gene_without_annot</span> <span class="o">=</span> <span class="n">CS_without_annot_uniq</span><span class="p">[</span><span class="s1">&#39;GENE&#39;</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">gene_without_annot</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>

<span class="n">gene_list_without_annot</span> <span class="o">=</span> <span class="n">gene_without_annot</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">gene_list_without_annot</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">gene_list_without_annot</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            kernelName: "sos",
            path: "./pipeline/fine_mapping/SuSiE_Ann"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="../SuSiE/SuSiE_RSS.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Fine-mapping with SuSiE RSS model</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="../multivariate_fine_mapping.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Multivariate fine-mapping workflows</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics Consortium<br/>
        
            &copy; Copyright 2021, ADSP-FG xQTL methods and data integration subgroup.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>