
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17: http://docutils.sourceforge.net/" />

    <title>Meta-analysis with METAL &#8212; ADSP-FG Workflows</title>
    
  <link href="../../../_static/css/theme.css" rel="stylesheet">
  <link href="../../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/togglebutton.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="MASH analysis pipeline with data-driven prior matrices" href="../MASH/mashr_flashr_workflow.html" />
    <link rel="prev" title="Consolidating summary statistics from multiple studies" href="../Reformatting/Partitioned.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../../_static/xqtl_wf.svg" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">ADSP-FG Workflows</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../../README.html">
   xQTL analysis pipeline
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Complete Pipelines
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../master_control.html">
   xQTL master workflow template
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Genotypes and Covariates
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../Data_Processing/Factor_and_Covariate/Apex_factor_test_log.html">
   APEX factor analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../Data_Processing/Factor_and_Covariate/factor.html">
   PEER factor analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../Data_Processing/Data_Processing.html">
   Genotype and covariate data preprocessing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../Data_Processing/Genotype/GWAS_QC.html">
   GWAS data QC workflow
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../Data_Processing/Genotype/LOCO_GRM.html">
   Leave one chromosome out GRM
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../Data_Processing/Genotype/PCA.html">
   Principal component analysis
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../Phenotypes/apacalling.html">
   APA calling Pipeline
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../Data_Processing/Phenotype/Normailzation.html">
   Molecular phenotype normalization
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../Data_Processing/Phenotype/Residual_Y.html">
   Molecular phenotype residuals
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../Data_Processing/Phenotype/Region_extraction.html">
   Gene region extraction
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../QTL_association/TensorQTL/TensorQTL.html">
   TensorQTL QTL calling pipeline
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../QTL_association/QTL_association.html">
   cis/trans QTL analysis workflows
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../QTL_association/APEX/APEX.html">
   APEX QTL calling pipeline
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Statistical Fine-mapping
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../Fine_Mapping/SumStat/SuSiE_RSS.html">
   Fine-mapping with SuSiE RSS model
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Integration
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../Reformatting/Signal_Extraction.html">
   Top signal extraction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../Reformatting/Partitioned.html">
   Consolidating summary statistics from multiple studies
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Meta-analysis with METAL
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../MASH/mashr_flashr_workflow.html">
   MASH analysis pipeline with data-driven prior matrices
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../MASH/mixture_prior.html">
   A multivariate EBNM approach for mixture multivariate distribution estimate
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../Meta_Analysis.html">
   Meta analysis
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Utilities
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../Data_Processing/Genotype/plink2VCF.html">
   Genotype data reformatting
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../Reformatting/RDS_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../Reformatting/Merged.html">
   Summary statistics formatting: RDS files in R
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH/NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../../_sources/workflow/Meta_Analysis/METAL/METAL.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fworkflow/Meta_Analysis/METAL/METAL.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/workflow/Meta_Analysis/METAL/METAL.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#script-generation">
   Script generation
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="meta-analysis-with-metal">
<h1>Meta-analysis with METAL<a class="headerlink" href="#meta-analysis-with-metal" title="Permalink to this headline">¶</a></h1>
<p>METAT is a command line tool, takes in a script that documenting the input sumstat files, the field of each of those sumstat input file, and then the end of scipt.
METAL analysis is enssentially a weighted sum of Z score, therefore the input of snps from each chrm can be processed seperately. To make sure the same region was analysis, the input of this workflow shall be a list of path to the sumstat, join by ” ” to be analyzed at once.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">sumstat_list</span> <span class="o">=</span> <span class="nb">list</span>
<span class="kn">parameter:</span> <span class="kp">name</span> <span class="o">=</span> <span class="nb">str</span>
<span class="kn">parameter:</span> <span class="n">variant_id</span> <span class="o">=</span> <span class="s1">&#39;variant_id&#39;</span>
<span class="kn">parameter:</span> <span class="n">Sample_size</span> <span class="o">=</span> <span class="s1">&#39;N&#39;</span>
<span class="kn">parameter:</span> <span class="n">ALLELE</span> <span class="o">=</span> <span class="s1">&#39;alt ref&#39;</span>
<span class="kn">parameter:</span> <span class="n">FREQ</span>   <span class="o">=</span> <span class="s1">&#39;EFFECT_ALLELE_FREQ&#39;</span>
<span class="kn">parameter:</span> <span class="n">Beta</span> <span class="o">=</span> <span class="s1">&#39;beta&#39;</span>
<span class="kn">parameter:</span> <span class="n">SE</span> <span class="o">=</span> <span class="s1">&#39;se&#39;</span>
<span class="kn">parameter:</span> <span class="n">PVAL</span>   <span class="o">=</span> <span class="s1">&#39;pval&#39;</span>
<span class="kn">parameter:</span> <span class="n">wd</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">parameter:</span> <span class="n">SCHEME</span> <span class="o">=</span> <span class="s1">&#39;STDERR&#39;</span>
<span class="kn">parameter:</span> <span class="n">AVERAGEFREQ</span> <span class="o">=</span> <span class="s2">&quot;OFF&quot;</span>
<span class="kn">parameter:</span> <span class="n">MINMAXFREQ</span>  <span class="o">=</span> <span class="s2">&quot;OFF&quot;</span>
<span class="kn">parameter:</span> <span class="n">GENOMICCONTROL</span>  <span class="o">=</span> <span class="s2">&quot;OFF&quot;</span>
</pre></div>
</div>
</div>
</div>
<section id="script-generation">
<h2>Script generation<a class="headerlink" href="#script-generation" title="Permalink to this headline">¶</a></h2>
<p>The METAL script contains three sections, the first is the analysis parameters, the second is the inputs files, and the third is the ending of script.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">METAL_1</span><span class="p">,</span><span class="n">generate_METAL_script</span><span class="p">]</span>
<span class="kn">input:</span>  <span class="n">sumstat_list</span>
<span class="kn">output:</span> <span class="n">METAL_script</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">wd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">.METAL_script.txt&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
        <span class="n">echo</span> <span class="s1">&#39;</span>
        <span class="c1"># Meta-analysis weighted by:</span>
        <span class="n">SCHEME</span>   <span class="err">$</span><span class="p">[</span><span class="n">SCHEME</span><span class="p">]</span>
        
        <span class="c1"># Whether do genomics control</span>
        <span class="n">GENOMICCONTROL</span> <span class="err">$</span><span class="p">[</span><span class="n">GENOMICCONTROL</span><span class="p">]</span>
        
        <span class="c1"># Whether do AVERAGEFREQ or MINMAXFREQ control</span>
        <span class="n">AVERAGEFREQ</span> <span class="err">$</span><span class="p">[</span><span class="n">AVERAGEFREQ</span><span class="p">]</span>
        <span class="n">MINMAXFREQ</span> <span class="err">$</span><span class="p">[</span><span class="n">MINMAXFREQ</span><span class="p">]</span> 
        <span class="n">OUTFILE</span> <span class="err">$</span><span class="p">[</span><span class="n">wd</span><span class="p">]</span><span class="o">/</span><span class="err">$</span><span class="p">[</span><span class="kp">name</span><span class="p">]</span> <span class="o">.</span><span class="n">METAL</span><span class="o">.</span><span class="n">txt</span><span class="s1">&#39; &gt; $[_output]</span>

        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="err">$</span><span class="p">[</span><span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sumstat_list</span><span class="p">)]</span> <span class="p">;</span> <span class="n">do</span> 
        <span class="n">echo</span> <span class="s2">&quot;     </span>
        <span class="n">MARKER</span>   <span class="err">$</span><span class="p">[</span><span class="n">variant_id</span><span class="p">]</span>
        <span class="n">WEIGHT</span>   <span class="err">$</span><span class="p">[</span><span class="n">Sample_size</span><span class="p">]</span>
        <span class="n">ALLELE</span>   <span class="err">$</span><span class="p">[</span><span class="n">ALLELE</span><span class="p">]</span>
        <span class="c1">#FREQ     $[FREQ]</span>
        <span class="n">EFFECT</span>   <span class="err">$</span><span class="p">[</span><span class="n">Beta</span><span class="p">]</span>
        <span class="n">STDERR</span>   <span class="err">$</span><span class="p">[</span><span class="n">SE</span><span class="p">]</span>
        <span class="n">PVAL</span>     <span class="err">$</span><span class="p">[</span><span class="n">PVAL</span><span class="p">]</span>
        <span class="n">PROCESS</span>  <span class="err">$</span><span class="n">i</span> <span class="s2">&quot; &gt;&gt; $[_output];</span>
        <span class="n">done</span> 
        
        <span class="n">echo</span> <span class="s1">&#39;</span>
        <span class="n">ANALYZE</span><span class="s1">&#39; &gt;&gt; $[_output]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">METAL_2</span><span class="p">,</span><span class="n">Excute_METAL_script</span><span class="p">]</span>
<span class="kn">output:</span> <span class="n">METAL_result</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">wd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">1.METAL.txt&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
       <span class="n">metal</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">METAL_3</span><span class="p">,</span><span class="n">Output_reformatting</span><span class="p">]</span>
<span class="kn">output:</span> <span class="n">METAL_result</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">wd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">.METAL.txt&#39;</span>
<span class="kn">python:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
        <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">],</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">df</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;variant_id&quot;</span><span class="p">,</span><span class="s2">&quot;alt&quot;</span><span class="p">,</span><span class="s2">&quot;ref&quot;</span><span class="p">,</span><span class="s2">&quot;Weight&quot;</span><span class="p">,</span><span class="s2">&quot;beta&quot;</span><span class="p">,</span><span class="s2">&quot;pval&quot;</span><span class="p">,</span><span class="s2">&quot;Direction&quot;</span><span class="p">]</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">se</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;$[_output]&quot;</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">index</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            kernelName: "sos",
            path: "./workflow/Meta_Analysis/METAL"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="../Reformatting/Partitioned.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Consolidating summary statistics from multiple studies</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="../MASH/mashr_flashr_workflow.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">MASH analysis pipeline with data-driven prior matrices</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics Consortium<br/>
        
            &copy; Copyright 2021, ADSP-FG xQTL methods and data integration subgroup.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>