
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17: http://docutils.sourceforge.net/" />

    <title>Multivariate colocalization &#8212; ADSP-FG Project</title>
    
  <link href="../../../_static/css/theme.css" rel="stylesheet">
  <link href="../../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/togglebutton.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Functional annotation integration" href="../functional_annotation.html" />
    <link rel="prev" title="Colocalization with FastENLOC" href="fastenloc.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">ADSP-FG Project</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../../README.html">
   xQTL analysis pipeline
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Complete Pipelines
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../complete_analysis/eQTL_analysis.html">
   Bulk RNA-seq eQTL analysis
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../molecular_phenotypes/bulk_expression.html">
   RNA-seq Calling
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../molecular_phenotypes/scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../molecular_phenotypes/apa_calling.html">
   APA Calling
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/GWAS_QC.html">
     GWAS data QC workflow
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/PCA.html">
     Principal component analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/GRM.html">
     Genomic relationship matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/genotype_formatting.html">
     Genotype data formatting workflows
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/phenotype/normalization.html">
     Molecular phenotype normalization
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_formatting.html">
     Phenotype data reformat
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/covariate_preprocessing.html">
   Covariate data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/covariate/PEER_factor.html">
     Factor analysis using PEER
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/covariate/BiCV_factor.html">
     Factor analysis using Bi-Cross validation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/covariate/remove_covariates.html">
     Computing covariate residuals
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../association_scan/cisQTL_scan.html">
   cisQTL analysis workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL calling pipeline
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/APEX/APEX.html">
     APEX QTL calling pipeline
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../association_scan/transQTL_scan.html">
   transQTL analysis workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../multivariate/METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/METAL/METAL.html">
     Meta-analysis with METAL
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../multivariate/MASH_pipeline.html">
   Multivariate association with MASH
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/fastqtl_to_mash.html">
     eQTL summary statistics formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/mixture_prior.html">
     A multivariate EBNM approach for mixture multivariate distribution estimate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/mashr.html">
     MASH analysis pipeline with data-driven prior matrices
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Statistical Fine-mapping
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../fine_mapping/univariate_fine_mapping.html">
   Univariate fine-mapping workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE_RSS.html">
     Fine-mapping with SuSiE RSS model
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE.html">
     Fine-mapping with SuSiE model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../fine_mapping/multivariate_fine_mapping.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multi-omics integration
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="../colocalization.html">
   QTL and GWAS Colocalization
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2">
    <a class="reference internal" href="fastenloc.html">
     Colocalization with FastENLOC
    </a>
   </li>
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     Multivariate colocalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../functional_annotation.html">
   Functional annotation integration
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../enrichment/gwas_enrichment.html">
     Enrichment analysis: TORUS and GREGOR
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../SuSiE_Ann/polyfun.html">
     Fine-mapping with PolyFun
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Utilities
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../misc/summary_stats_merger.html">
   Summary statistics merger
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../misc/rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH/NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../../_sources/pipeline/integrative_analysis/COLOC/mvenloc.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fpipeline/integrative_analysis/COLOC/mvenloc.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/pipeline/integrative_analysis/COLOC/mvenloc.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#aim">
   Aim:
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#input">
   Input:
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#output">
   Output:
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#method">
   Method:
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#functions">
   Functions:
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#allele-flip">
     1) Allele Flip
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#construct-eqtl-data-frame">
     2) Construct eQTL data frame
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#remove-duplicate-variants">
     3) Remove Duplicate variants
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#workflow">
   Workflow:
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#step-1-extract-common-snps-between-the-gwas-summary-statistics-and-eqtl-data">
     Step 1: Extract common SNPS between the GWAS summary statistics and eQTL data
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#step-2-obtain-snp-pips-from-gwas-summary-statistics-using-susie-linear-regression">
     Step 2: Obtain SNP PIPs from GWAS summary statistics using SuSiE linear regression
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#step-3-construct-signal-clusters-using-eqtl-snp-pips">
     Step 3: Construct signal clusters using eQTL SNP PIPs
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#minimal-working-example">
   Minimal Working Example:
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <section id="multivariate-colocalization">
<h1>Multivariate colocalization<a class="headerlink" href="#multivariate-colocalization" title="Permalink to this headline">¶</a></h1>
<section id="aim">
<h2>Aim:<a class="headerlink" href="#aim" title="Permalink to this headline">¶</a></h2>
<p>This notebook aims to demonstrate a workflow of generating posterior inclusion probabilities (PIPs) from GWAS summary statistics using SuSiE linear regression and construsting SNP signal clusters from global eQTL analysis data obtained from multivariate SuSiE models.</p>
</section>
<section id="input">
<h2>Input:<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h2>
<ol class="simple">
<li><p>GWAS Summary Statistics with the following columns:</p></li>
</ol>
<ul class="simple">
<li><p>chr: chromosome number</p></li>
<li><p>bp: base pair position</p></li>
<li><p>a1: effect allele</p></li>
<li><p>a2: other allele</p></li>
<li><p>beta: effect size</p></li>
<li><p>se: standard error of beta</p></li>
<li><p>z: z score</p></li>
</ul>
<ol class="simple">
<li><p>eQTL data from multivariate SuSiE model with the following columns:</p></li>
</ol>
<ul class="simple">
<li><p>chr: chromosome number</p></li>
<li><p>bp: base pair position</p></li>
<li><p>a1: effect allele</p></li>
<li><p>a2: other allele</p></li>
<li><p>pip: posterior inclusion probability</p></li>
</ul>
<ol class="simple">
<li><p>LD correlation matrix</p></li>
</ol>
</section>
<section id="output">
<h2>Output:<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h2>
<p>Intermediate files:</p>
<ol class="simple">
<li><p>GWAS PIP file</p></li>
<li><p>eQTL annotation file</p></li>
</ol>
<p>Final Outputs:</p>
<ol class="simple">
<li><p>Enrichment analysis result prefix.enloc.enrich.rst: estimated enrichment parameters and standard errors.</p></li>
<li><p>Signal-level colocalization result prefix.enloc.sig.out: the main output from the colocalization analysis with the following format</p></li>
</ol>
<ul class="simple">
<li><p>column 1: signal cluster name (from eQTL analysis)</p></li>
<li><p>column 2: number of member SNPs</p></li>
<li><p>column 3: cluster PIP of eQTLs</p></li>
<li><p>column 4: cluster PIP of GWAS hits (without eQTL prior)</p></li>
<li><p>column 5: cluster PIP of GWAS hits (with eQTL prior)</p></li>
<li><p>column 6: regional colocalization probability (RCP)</p></li>
</ul>
<ol class="simple">
<li><p>SNP-level colocalization result prefix.enloc.snp.out: SNP-level colocalization output with the following form at</p></li>
</ol>
<ul class="simple">
<li><p>column 1: signal cluster name</p></li>
<li><p>column 2: SNP name</p></li>
<li><p>column 3: SNP-level PIP of eQTLs</p></li>
<li><p>column 4: SNP-level PIP of GWAS (without eQTL prior)</p></li>
<li><p>column 5: SNP-level PIP of GWAS (with eQTL prior)</p></li>
<li><p>column 6: SNP-level colocalization probability</p></li>
</ul>
<ol class="simple">
<li><p>Sorted list of colocalization signals with
<code class="docutils literal notranslate"><span class="pre">sort</span> <span class="pre">-grk6</span> <span class="pre">prefix.enloc.sig.out</span></code></p></li>
</ol>
</section>
<section id="method">
<h2>Method:<a class="headerlink" href="#method" title="Permalink to this headline">¶</a></h2>
</section>
<section id="functions">
<h2>Functions:<a class="headerlink" href="#functions" title="Permalink to this headline">¶</a></h2>
<section id="allele-flip">
<h3>1) Allele Flip<a class="headerlink" href="#allele-flip" title="Permalink to this headline">¶</a></h3>
<p>Takes into consideration 3 situations:</p>
<ol class="simple">
<li><p>“Major” and “minor” alleles flipped</p></li>
<li><p>Different strand but same variant</p></li>
<li><p>Remove variants with A/T and C/G alleles due to ambiguity</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">allele</span><span class="o">.</span><span class="n">qc</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">a1</span><span class="p">,</span><span class="n">a2</span><span class="p">,</span><span class="n">ref1</span><span class="p">,</span><span class="n">ref2</span><span class="p">)</span> <span class="p">{</span>
	<span class="c1"># a1 and a2 are the first data-set</span>
	<span class="c1"># ref1 and ref2 are the 2nd data-set</span>
	<span class="c1"># Make all the alleles into upper-case, as A,T,C,G:</span>
        <span class="n">a1</span> <span class="o">=</span> <span class="n">toupper</span><span class="p">(</span><span class="n">a1</span><span class="p">)</span>
        <span class="n">a2</span> <span class="o">=</span> <span class="n">toupper</span><span class="p">(</span><span class="n">a2</span><span class="p">)</span>
        <span class="n">ref1</span> <span class="o">=</span> <span class="n">toupper</span><span class="p">(</span><span class="n">ref1</span><span class="p">)</span>
        <span class="n">ref2</span> <span class="o">=</span> <span class="n">toupper</span><span class="p">(</span><span class="n">ref2</span><span class="p">)</span>
	<span class="c1"># Strand flip, to change the allele representation in the 2nd data-set</span>
	<span class="n">strand_flip</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">ref</span><span class="p">)</span> <span class="p">{</span>
		<span class="n">flip</span> <span class="o">=</span> <span class="n">ref</span>
		<span class="n">flip</span><span class="p">[</span><span class="n">ref</span> <span class="o">==</span> <span class="s2">&quot;A&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;T&quot;</span>
		<span class="n">flip</span><span class="p">[</span><span class="n">ref</span> <span class="o">==</span> <span class="s2">&quot;T&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;A&quot;</span>
		<span class="n">flip</span><span class="p">[</span><span class="n">ref</span> <span class="o">==</span> <span class="s2">&quot;G&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;C&quot;</span>
		<span class="n">flip</span><span class="p">[</span><span class="n">ref</span> <span class="o">==</span> <span class="s2">&quot;C&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;G&quot;</span>
		<span class="n">flip</span>
	<span class="p">}</span>
	<span class="n">flip1</span> <span class="o">=</span> <span class="n">strand_flip</span><span class="p">(</span><span class="n">ref1</span><span class="p">)</span>
	<span class="n">flip2</span> <span class="o">=</span> <span class="n">strand_flip</span><span class="p">(</span><span class="n">ref2</span><span class="p">)</span>
	<span class="n">snp</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
	<span class="c1"># Remove strand ambiguous SNPs (scenario 3)</span>
	<span class="n">snp</span><span class="p">[[</span><span class="s2">&quot;keep&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="err">!</span><span class="p">((</span><span class="n">a1</span><span class="o">==</span><span class="s2">&quot;A&quot;</span> <span class="o">&amp;</span> <span class="n">a2</span><span class="o">==</span><span class="s2">&quot;T&quot;</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">a1</span><span class="o">==</span><span class="s2">&quot;T&quot;</span> <span class="o">&amp;</span> <span class="n">a2</span><span class="o">==</span><span class="s2">&quot;A&quot;</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">a1</span><span class="o">==</span><span class="s2">&quot;C&quot;</span> <span class="o">&amp;</span> <span class="n">a2</span><span class="o">==</span><span class="s2">&quot;G&quot;</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">a1</span><span class="o">==</span><span class="s2">&quot;G&quot;</span> <span class="o">&amp;</span> <span class="n">a2</span><span class="o">==</span><span class="s2">&quot;C&quot;</span><span class="p">))</span>
	<span class="c1"># Remove non-ATCG coding</span>
	<span class="n">snp</span><span class="p">[[</span><span class="s2">&quot;keep&quot;</span><span class="p">]][</span> <span class="n">a1</span> <span class="o">!=</span> <span class="s2">&quot;A&quot;</span> <span class="o">&amp;</span> <span class="n">a1</span> <span class="o">!=</span> <span class="s2">&quot;T&quot;</span> <span class="o">&amp;</span> <span class="n">a1</span> <span class="o">!=</span> <span class="s2">&quot;G&quot;</span> <span class="o">&amp;</span> <span class="n">a1</span> <span class="o">!=</span> <span class="s2">&quot;C&quot;</span> <span class="p">]</span> <span class="o">=</span> <span class="n">F</span>
	<span class="n">snp</span><span class="p">[[</span><span class="s2">&quot;keep&quot;</span><span class="p">]][</span> <span class="n">a2</span> <span class="o">!=</span> <span class="s2">&quot;A&quot;</span> <span class="o">&amp;</span> <span class="n">a2</span> <span class="o">!=</span> <span class="s2">&quot;T&quot;</span> <span class="o">&amp;</span> <span class="n">a2</span> <span class="o">!=</span> <span class="s2">&quot;G&quot;</span> <span class="o">&amp;</span> <span class="n">a2</span> <span class="o">!=</span> <span class="s2">&quot;C&quot;</span> <span class="p">]</span> <span class="o">=</span> <span class="n">F</span>
	<span class="c1"># as long as scenario 1 is involved, sign_flip will return TRUE</span>
	<span class="n">snp</span><span class="p">[[</span><span class="s2">&quot;sign_flip&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="p">(</span><span class="n">a1</span> <span class="o">==</span> <span class="n">ref2</span> <span class="o">&amp;</span> <span class="n">a2</span> <span class="o">==</span> <span class="n">ref1</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">a1</span> <span class="o">==</span> <span class="n">flip2</span> <span class="o">&amp;</span> <span class="n">a2</span> <span class="o">==</span> <span class="n">flip1</span><span class="p">)</span>
	<span class="c1"># as long as scenario 2 is involved, strand_flip will return TRUE</span>
	<span class="n">snp</span><span class="p">[[</span><span class="s2">&quot;strand_flip&quot;</span><span class="p">]]</span> <span class="o">=</span> <span class="p">(</span><span class="n">a1</span> <span class="o">==</span> <span class="n">flip1</span> <span class="o">&amp;</span> <span class="n">a2</span> <span class="o">==</span> <span class="n">flip2</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">a1</span> <span class="o">==</span> <span class="n">flip2</span> <span class="o">&amp;</span> <span class="n">a2</span> <span class="o">==</span> <span class="n">flip1</span><span class="p">)</span>
	<span class="c1"># remove other cases, eg, tri-allelic, one dataset is A C, the other is A G, for example.</span>
	<span class="n">exact_match</span> <span class="o">=</span> <span class="p">(</span><span class="n">a1</span> <span class="o">==</span> <span class="n">ref1</span> <span class="o">&amp;</span> <span class="n">a2</span> <span class="o">==</span> <span class="n">ref2</span><span class="p">)</span> 
	<span class="n">snp</span><span class="p">[[</span><span class="s2">&quot;keep&quot;</span><span class="p">]][</span><span class="err">!</span><span class="p">(</span><span class="n">exact_match</span> <span class="o">|</span> <span class="n">snp</span><span class="p">[[</span><span class="s2">&quot;sign_flip&quot;</span><span class="p">]]</span> <span class="o">|</span> <span class="n">snp</span><span class="p">[[</span><span class="s2">&quot;strand_flip&quot;</span><span class="p">]])]</span> <span class="o">=</span> <span class="n">F</span>
	<span class="k">return</span><span class="p">(</span><span class="n">snp</span><span class="p">)</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="construct-eqtl-data-frame">
<h3>2) Construct eQTL data frame<a class="headerlink" href="#construct-eqtl-data-frame" title="Permalink to this headline">¶</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">eqtl</span><span class="o">.</span><span class="n">split</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">eqtl</span><span class="p">){</span>
  <span class="n">rows</span> <span class="o">=</span> <span class="n">length</span><span class="p">(</span><span class="n">eqtl</span><span class="p">)</span>
  <span class="nb">chr</span> <span class="o">=</span> <span class="n">vector</span><span class="p">(</span><span class="n">length</span> <span class="o">=</span> <span class="n">rows</span><span class="p">)</span>
  <span class="n">pos</span> <span class="o">=</span> <span class="n">vector</span><span class="p">(</span><span class="n">length</span> <span class="o">=</span> <span class="n">rows</span><span class="p">)</span>
  <span class="n">a1</span> <span class="o">=</span> <span class="n">vector</span><span class="p">(</span><span class="n">length</span> <span class="o">=</span> <span class="n">rows</span><span class="p">)</span>
  <span class="n">a2</span> <span class="o">=</span> <span class="n">vector</span><span class="p">(</span><span class="n">length</span> <span class="o">=</span> <span class="n">rows</span><span class="p">)</span>
  <span class="k">for</span> <span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">rows</span><span class="p">){</span>
    <span class="n">split1</span> <span class="o">=</span> <span class="n">str_split</span><span class="p">(</span><span class="n">eqtl</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="s2">&quot;:&quot;</span><span class="p">)</span>
    <span class="n">split2</span> <span class="o">=</span> <span class="n">str_split</span><span class="p">(</span><span class="n">split1</span><span class="p">[[</span><span class="mi">1</span><span class="p">]][</span><span class="mi">2</span><span class="p">],</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span>
    <span class="nb">chr</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">=</span> <span class="n">split1</span><span class="p">[[</span><span class="mi">1</span><span class="p">]][</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">pos</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">split2</span><span class="p">[[</span><span class="mi">1</span><span class="p">]][</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">a1</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">split2</span><span class="p">[[</span><span class="mi">1</span><span class="p">]][</span><span class="mi">2</span><span class="p">]</span>
    <span class="n">a2</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">split2</span><span class="p">[[</span><span class="mi">1</span><span class="p">]][</span><span class="mi">3</span><span class="p">]</span>

  <span class="p">}</span>
  <span class="n">eqtl</span><span class="o">.</span><span class="n">df</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">eqtl</span><span class="p">,</span><span class="nb">chr</span><span class="p">,</span><span class="n">pos</span><span class="p">,</span><span class="n">a1</span><span class="p">,</span><span class="n">a2</span><span class="p">)</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="remove-duplicate-variants">
<h3>3) Remove Duplicate variants<a class="headerlink" href="#remove-duplicate-variants" title="Permalink to this headline">¶</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">remove</span><span class="o">.</span><span class="n">dup</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">df</span><span class="p">){</span>
  <span class="n">df</span> <span class="o">=</span> <span class="n">df</span> <span class="o">%&gt;%</span> <span class="n">arrange</span><span class="p">(</span><span class="n">PosGRCh37</span><span class="p">,</span> <span class="o">-</span><span class="n">N</span><span class="p">)</span>
  <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="err">!</span><span class="n">duplicated</span><span class="p">(</span><span class="n">df</span><span class="err">$</span><span class="n">PosGRCh37</span><span class="p">),]</span>
  <span class="k">return</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">greedy</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">alpha</span><span class="p">){</span>
  <span class="nb">len</span> <span class="o">=</span> <span class="n">length</span><span class="p">(</span><span class="n">alpha</span><span class="p">)</span>
  <span class="n">r2</span> <span class="o">=</span> <span class="mf">0.5</span>
  <span class="n">c_r2</span> <span class="o">=</span> <span class="mi">1</span>
  <span class="n">ids</span> <span class="o">=</span> <span class="n">vector</span><span class="p">()</span>
  <span class="n">first</span> <span class="o">=</span> <span class="mi">1</span>
  <span class="n">less</span> <span class="o">=</span> <span class="mi">0</span>
  <span class="n">total</span> <span class="o">=</span> <span class="mi">0</span>
  <span class="k">for</span><span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="nb">len</span><span class="p">){</span>
    <span class="nb">max</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">alpha</span><span class="p">)</span>
    <span class="n">total</span> <span class="o">=</span> <span class="n">total</span> <span class="o">+</span> <span class="nb">max</span>
    <span class="n">idx</span> <span class="o">=</span> <span class="n">which</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">alpha</span><span class="p">)</span>
    <span class="k">if</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">ids</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">){</span>
      <span class="n">ids</span> <span class="o">=</span> <span class="n">append</span><span class="p">(</span><span class="n">ids</span><span class="p">,</span><span class="n">idx</span><span class="p">)</span>
    <span class="p">}</span>
    
    
    <span class="k">for</span><span class="p">(</span><span class="n">j</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="n">ids</span><span class="p">)){</span>
      <span class="n">new_r2</span> <span class="o">=</span> <span class="p">(</span><span class="n">ld_f2</span><span class="p">[</span><span class="n">ids</span><span class="p">[</span><span class="n">j</span><span class="p">],</span><span class="n">idx</span><span class="p">])</span><span class="o">^</span><span class="mi">2</span>
     <span class="c1"># print(new_r2)</span>
      <span class="k">if</span><span class="p">(</span><span class="n">new_r2</span> <span class="o">&lt;</span> <span class="n">r2</span><span class="p">){</span>
        <span class="n">less</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="k">break</span>
      <span class="p">}</span>
    <span class="p">}</span>
    
    <span class="k">if</span><span class="p">(</span><span class="n">less</span> <span class="o">==</span> <span class="mi">1</span><span class="p">){</span>
      <span class="k">break</span>
      <span class="p">}</span>
    
    <span class="k">if</span><span class="p">(</span><span class="n">first</span> <span class="o">==</span> <span class="mi">0</span><span class="p">){</span>
    <span class="n">ids</span> <span class="o">=</span> <span class="n">append</span><span class="p">(</span><span class="n">ids</span><span class="p">,</span><span class="n">idx</span><span class="p">)</span>
    <span class="p">}</span>

    <span class="n">alpha</span> <span class="o">=</span> <span class="n">alpha</span><span class="p">[</span><span class="o">-</span><span class="n">idx</span><span class="p">]</span>
    <span class="n">first</span> <span class="o">=</span> <span class="mi">0</span>
    
  <span class="p">}</span>
  <span class="k">return</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">ids</span><span class="p">,</span> <span class="n">total</span><span class="p">))</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="workflow">
<h2>Workflow:<a class="headerlink" href="#workflow" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span>
<span class="n">library</span><span class="p">(</span><span class="n">devtools</span><span class="p">)</span>
<span class="n">library</span><span class="p">(</span><span class="n">susieR</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<section id="step-1-extract-common-snps-between-the-gwas-summary-statistics-and-eqtl-data">
<h3>Step 1: Extract common SNPS between the GWAS summary statistics and eQTL data<a class="headerlink" href="#step-1-extract-common-snps-between-the-gwas-summary-statistics-and-eqtl-data" title="Permalink to this headline">¶</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># eqtl</span>
<span class="n">load</span><span class="p">(</span><span class="n">file</span><span class="o">.</span><span class="n">choose</span><span class="p">())</span>
<span class="n">eqtl</span> <span class="o">=</span> <span class="n">m_ENSG00000159640</span>

<span class="c1"># colnames</span>
<span class="n">load</span><span class="p">(</span><span class="s2">&quot;geneTpmResidualsAgeGenderAdj_rename.ENSG00000159640.transformed_XY.RData&quot;</span><span class="p">)</span>
<span class="n">var</span> <span class="o">=</span> <span class="n">colnames</span><span class="p">(</span><span class="n">scaled_ENSG00000159640</span><span class="p">[[</span><span class="mi">1</span><span class="p">]])</span>

<span class="c1">#ld</span>
<span class="n">ld</span> <span class="o">=</span> <span class="n">readRDS</span><span class="p">(</span><span class="n">file</span><span class="o">.</span><span class="n">choose</span><span class="p">())</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">gwas_filter</span> <span class="o">=</span> <span class="n">gwas</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">gwas</span><span class="err">$</span><span class="nb">id</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">var</span><span class="p">),]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">eqtl</span><span class="o">.</span><span class="n">df</span> <span class="o">=</span> <span class="n">eqtl</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">var</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># allele flip</span>
<span class="n">f_gwas</span> <span class="o">=</span> <span class="n">gwas</span> <span class="o">%&gt;%</span> <span class="nb">filter</span><span class="p">(</span><span class="nb">chr</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">eqtl</span><span class="o">.</span><span class="n">df</span><span class="err">$</span><span class="nb">chr</span> <span class="o">&amp;</span> <span class="n">PosGRCh37</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">eqtl</span><span class="o">.</span><span class="n">df</span><span class="err">$</span><span class="n">pos</span><span class="p">)</span>
<span class="n">eqtl</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">f</span> <span class="o">=</span> <span class="n">eqtl</span><span class="o">.</span><span class="n">df</span> <span class="o">%&gt;%</span> <span class="nb">filter</span><span class="p">(</span><span class="n">pos</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">f_gwas</span><span class="err">$</span><span class="n">PosGRCh37</span><span class="p">)</span>

<span class="c1"># check if there are duplicate pos</span>
<span class="n">length</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="n">f_gwas</span><span class="err">$</span><span class="n">PosGRCh37</span><span class="p">))</span>

<span class="c1"># multiple snps with same pos</span>
<span class="n">dup</span><span class="o">.</span><span class="n">pos</span> <span class="o">=</span> <span class="n">f_gwas</span> <span class="o">%&gt;%</span> <span class="n">group_by</span><span class="p">(</span><span class="n">PosGRCh37</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="nb">filter</span><span class="p">(</span><span class="n">n</span><span class="p">()</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">)</span> 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">f_gwas</span> <span class="o">=</span> <span class="n">remove</span><span class="o">.</span><span class="n">dup</span><span class="p">(</span><span class="n">f_gwas</span><span class="p">)</span>

<span class="n">qc</span> <span class="o">=</span> <span class="n">allele</span><span class="o">.</span><span class="n">qc</span><span class="p">(</span><span class="n">f_gwas</span><span class="err">$</span><span class="n">testedAllele</span><span class="p">,</span> <span class="n">f_gwas</span><span class="err">$</span><span class="n">otherAllele</span><span class="p">,</span> <span class="n">eqtl</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">f</span><span class="err">$</span><span class="n">a1</span><span class="p">,</span> <span class="n">eqtl</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">f</span><span class="err">$</span><span class="n">a2</span><span class="p">)</span>
<span class="n">keep</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">qc</span><span class="err">$</span><span class="n">keep</span><span class="p">)</span>
<span class="n">sign</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">qc</span><span class="err">$</span><span class="n">sign_flip</span><span class="p">)</span>
<span class="n">strand</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">qc</span><span class="err">$</span><span class="n">strand_flip</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># sign flip</span>
<span class="n">f_gwas</span><span class="err">$</span><span class="n">z</span><span class="p">[</span><span class="n">qc</span><span class="err">$</span><span class="n">sign_flip</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span> <span class="o">*</span> <span class="n">f_gwas</span><span class="err">$</span><span class="n">z</span><span class="p">[</span><span class="n">qc</span><span class="err">$</span><span class="n">sign_flip</span><span class="p">]</span>
<span class="n">f_gwas</span><span class="err">$</span><span class="n">testedAllele</span><span class="p">[</span><span class="n">qc</span><span class="err">$</span><span class="n">sign_flip</span><span class="p">]</span> <span class="o">=</span> <span class="n">eqtl</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">f</span><span class="err">$</span><span class="n">a1</span><span class="p">[</span><span class="n">qc</span><span class="err">$</span><span class="n">sign_flip</span><span class="p">]</span>
<span class="n">f_gwas</span><span class="err">$</span><span class="n">otherAllele</span><span class="p">[</span><span class="n">qc</span><span class="err">$</span><span class="n">sign_flip</span><span class="p">]</span> <span class="o">=</span> <span class="n">eqtl</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">f</span><span class="err">$</span><span class="n">a2</span><span class="p">[</span><span class="n">qc</span><span class="err">$</span><span class="n">sign_flip</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">f_gwas</span><span class="err">$</span><span class="n">testedAllele</span><span class="p">[</span><span class="n">qc</span><span class="err">$</span><span class="n">strand_flip</span><span class="p">]</span> <span class="o">=</span> <span class="n">eqtl</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">f</span><span class="err">$</span><span class="n">a1</span><span class="p">[</span><span class="n">qc</span><span class="err">$</span><span class="n">strand_flip</span><span class="p">]</span>
<span class="n">f_gwas</span><span class="err">$</span><span class="n">otherAllele</span><span class="p">[</span><span class="n">qc</span><span class="err">$</span><span class="n">strand_flip</span><span class="p">]</span> <span class="o">=</span> <span class="n">eqtl</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">f</span><span class="err">$</span><span class="n">a2</span><span class="p">[</span><span class="n">qc</span><span class="err">$</span><span class="n">strand_flip</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># remove ambigiuous </span>
<span class="k">if</span> <span class="p">(</span> <span class="nb">sum</span><span class="p">(</span><span class="err">!</span><span class="n">qc</span><span class="err">$</span><span class="n">keep</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="p">)</span> <span class="p">{</span>
  <span class="n">eqtl</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">f</span> <span class="o">=</span> <span class="n">eqtl</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">f</span><span class="p">[</span><span class="n">qc</span><span class="err">$</span><span class="n">keep</span><span class="p">,]</span>
  <span class="n">f_gwas</span> <span class="o">=</span> <span class="n">f_gwas</span><span class="p">[</span><span class="n">qc</span><span class="err">$</span><span class="n">keep</span><span class="p">,]</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># lds filtered</span>
<span class="n">eqtl_id</span> <span class="o">=</span> <span class="n">which</span><span class="p">(</span><span class="n">var</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">eqtl</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">f</span><span class="err">$</span><span class="n">eqtl</span><span class="p">)</span>
<span class="n">ld_f</span> <span class="o">=</span> <span class="n">ld</span><span class="p">[</span><span class="n">eqtl_id</span><span class="p">,</span> <span class="n">eqtl_id</span><span class="p">]</span>
<span class="n">dim</span><span class="p">(</span><span class="n">ld_f</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># ld missing</span>

<span class="n">miss</span> <span class="o">=</span> <span class="n">which</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">na</span><span class="p">(</span><span class="n">ld_f</span><span class="p">),</span> <span class="n">arr</span><span class="o">.</span><span class="n">ind</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>


<span class="n">miss_r</span> <span class="o">=</span> <span class="n">unique</span><span class="p">(</span><span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">miss</span><span class="p">)</span><span class="err">$</span><span class="n">row</span><span class="p">)</span>
<span class="n">miss_c</span> <span class="o">=</span> <span class="n">unique</span><span class="p">(</span><span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">miss</span><span class="p">)</span><span class="err">$</span><span class="n">col</span><span class="p">)</span>

<span class="n">total_miss</span> <span class="o">=</span> <span class="n">unique</span><span class="p">(</span><span class="n">union</span><span class="p">(</span><span class="n">miss_r</span><span class="p">,</span><span class="n">miss_c</span><span class="p">))</span>

<span class="k">if</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">total_miss</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">){</span> 
<span class="n">eqtl_id</span><span class="o">.</span><span class="n">f</span> <span class="o">=</span> <span class="n">eqtl</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">f</span><span class="p">[</span><span class="o">-</span><span class="n">total_miss</span><span class="p">,]</span>
<span class="p">}</span><span class="k">else</span><span class="p">{</span><span class="n">eqtl_id</span><span class="o">.</span><span class="n">f</span> <span class="o">=</span> <span class="n">eqtl</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">f</span><span class="p">}</span>

<span class="c1"># rename ids</span>

<span class="n">f_gwas</span> <span class="o">=</span> <span class="n">f_gwas</span> <span class="o">%&gt;%</span> <span class="n">mutate</span><span class="p">(</span><span class="nb">id</span> <span class="o">=</span> <span class="n">paste</span><span class="p">(</span><span class="n">f_gwas</span><span class="err">$</span><span class="nb">chr</span><span class="p">,</span> <span class="n">paste</span><span class="p">(</span><span class="n">f_gwas</span><span class="err">$</span><span class="n">PosGRCh37</span><span class="p">,</span> <span class="n">f_gwas</span><span class="err">$</span><span class="n">testedAllele</span><span class="p">,</span> <span class="n">f_gwas</span><span class="err">$</span><span class="n">otherAllele</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;_&quot;</span><span class="p">),</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;:&quot;</span><span class="p">))</span>

<span class="n">f_gwas</span><span class="o">.</span><span class="n">f</span> <span class="o">=</span> <span class="n">f_gwas</span> <span class="o">%&gt;%</span> <span class="nb">filter</span><span class="p">(</span><span class="nb">id</span> <span class="o">%</span><span class="ow">in</span><span class="o">%</span> <span class="n">eqtl_id</span><span class="o">.</span><span class="n">f</span><span class="err">$</span><span class="n">eqtl</span><span class="p">)</span>


<span class="k">if</span> <span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">total_miss</span><span class="p">)</span><span class="o">!=</span><span class="mi">0</span><span class="p">){</span>
<span class="n">ld_f2</span> <span class="o">=</span> <span class="n">ld_f</span><span class="p">[</span><span class="o">-</span><span class="n">total_miss</span><span class="p">,]</span>
<span class="n">ld_f2</span> <span class="o">=</span> <span class="n">ld_f2</span><span class="p">[,</span><span class="o">-</span><span class="n">total_miss</span><span class="p">]</span>
<span class="n">dim</span><span class="p">(</span><span class="n">ld_f2</span><span class="p">)</span>
<span class="p">}</span><span class="k">else</span><span class="p">{</span><span class="n">ld_f2</span> <span class="o">=</span> <span class="n">ld_f</span><span class="p">}</span>

<span class="n">miss</span> <span class="o">=</span> <span class="n">which</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">na</span><span class="p">(</span><span class="n">ld_f2</span><span class="p">),</span> <span class="n">arr</span><span class="o">.</span><span class="n">ind</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>

<span class="c1"># remove multi-allelic snps that aren&#39;t present in both eqtl and gwas </span>


<span class="n">idx</span> <span class="o">=</span> <span class="n">which</span><span class="p">(</span><span class="n">eqtl_id</span><span class="o">.</span><span class="n">f</span><span class="err">$</span><span class="n">eqtl</span> <span class="o">%</span><span class="n">notin</span><span class="o">%</span> <span class="n">f_gwas</span><span class="err">$</span><span class="nb">id</span><span class="p">)</span>
<span class="k">if</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">){</span>
<span class="n">eqtl_id</span><span class="o">.</span><span class="n">f</span> <span class="o">=</span> <span class="n">eqtl_id</span><span class="o">.</span><span class="n">f</span><span class="p">[</span><span class="o">-</span><span class="n">idx</span><span class="p">,]</span>
<span class="n">ld_f2</span> <span class="o">=</span> <span class="n">ld_f2</span><span class="p">[</span><span class="o">-</span><span class="n">idx</span><span class="p">,]</span>
<span class="n">ld_f2</span> <span class="o">=</span> <span class="n">ld_f2</span><span class="p">[,</span><span class="o">-</span><span class="n">idx</span><span class="p">]</span>
<span class="p">}</span>
<span class="n">dim</span><span class="p">(</span><span class="n">ld_f2</span><span class="p">)</span>
<span class="n">dim</span><span class="p">(</span><span class="n">eqtl_id</span><span class="o">.</span><span class="n">f</span><span class="p">)</span>
<span class="n">dim</span><span class="p">(</span><span class="n">f_gwas</span><span class="o">.</span><span class="n">f</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="step-2-obtain-snp-pips-from-gwas-summary-statistics-using-susie-linear-regression">
<h3>Step 2: Obtain SNP PIPs from GWAS summary statistics using SuSiE linear regression<a class="headerlink" href="#step-2-obtain-snp-pips-from-gwas-summary-statistics-using-susie-linear-regression" title="Permalink to this headline">¶</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">susie_results</span> <span class="o">=</span> <span class="n">susieR</span><span class="p">::</span><span class="n">susie_rss</span><span class="p">(</span><span class="n">z</span> <span class="o">=</span> <span class="n">f_gwas</span><span class="o">.</span><span class="n">f</span><span class="err">$</span><span class="n">z</span><span class="p">,</span><span class="n">R</span> <span class="o">=</span> <span class="n">ld_f2</span><span class="p">,</span> <span class="n">check_prior</span> <span class="o">=</span> <span class="n">F</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="kn">susieR:</span><span class="p">:</span><span class="n">susie_plot</span><span class="p">(</span><span class="n">susie_results</span><span class="p">,</span><span class="s2">&quot;PIP&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">susie_results</span><span class="err">$</span><span class="n">z</span> <span class="o">=</span> <span class="n">f_gwas</span><span class="o">.</span><span class="n">f</span><span class="err">$</span><span class="n">z</span>
<span class="kn">susieR:</span><span class="p">:</span><span class="n">susie_plot</span><span class="p">(</span><span class="n">susie_results</span><span class="p">,</span><span class="s2">&quot;z_original&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="step-3-construct-signal-clusters-using-eqtl-snp-pips">
<h3>Step 3: Construct signal clusters using eQTL SNP PIPs<a class="headerlink" href="#step-3-construct-signal-clusters-using-eqtl-snp-pips" title="Permalink to this headline">¶</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="k">for</span><span class="p">(</span> <span class="n">al</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="mi">10</span><span class="p">){</span>
  <span class="n">a</span> <span class="o">=</span> <span class="n">greedy</span><span class="p">(</span><span class="n">alpha</span><span class="p">[</span><span class="n">al</span><span class="p">,])</span>
  <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;cluster&quot;</span><span class="p">)</span>
  <span class="nb">print</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="minimal-working-example">
<h2>Minimal Working Example:<a class="headerlink" href="#minimal-working-example" title="Permalink to this headline">¶</a></h2>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            kernelName: "sos",
            path: "./pipeline/integrative_analysis/COLOC"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="fastenloc.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Colocalization with FastENLOC</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="../functional_annotation.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Functional annotation integration</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics Consortium<br/>
        
            &copy; Copyright 2021, ADSP-FG xQTL methods and data integration subgroup.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>