{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "attached-chancellor",
   "metadata": {
    "kernel": "SoS",
    "tags": []
   },
   "source": [
    "# Molecular phenotype normalization\n",
    "This is the normalization step for data processing pipeline for xqtl workflow, containing the generation of:\n",
    "1. whole Molecular_phenotype in bed format, normalized.\n",
    "\n",
    "\n",
    "### Input\n",
    "The input for this workflow is:\n",
    "1. 1 complete gene data table\n",
    "2. 1 complete gene tmp table\n",
    "3. 1 gtf table after the collapse_gene notation\n",
    "4. 1 index file cross referencing the sample name in the expression and genotype\n",
    "5. 1 vcf_chrom_list, provided by default, should be 1 column of chr1:chr22 chromosomes, without header \n",
    "\n",
    "\n",
    "**Violation of any of the following requirement will cause error and break the process unless noted otherwise**\n",
    "\n",
    "Requirement for input 1 and 2:\n",
    "1. Sep by \"\\t\"\n",
    "2. have 2 unneeded rows above colname\n",
    "3. file name end with gct\n",
    "4. can only have samples in input 4\n",
    "5. req 1,2,3 already satisfy by using output from bulk_expression_qc\n",
    "\n",
    "Requirement for 3 and 5:\n",
    "1. 3 must have the same gene name format as 1 & 2, i.e. ENSG00000000003 & ENSG00000000003.1 can not coexist\n",
    "\n",
    "    Violation will cause empty output\n",
    "    \n",
    "2. 3 must have the same chromosome name format as 5.\n",
    "    \n",
    "    Violation will cause empty output\n",
    "    \n",
    "3. 5 should have the same chromosome name format as the genotype data to be used.\n",
    "\n",
    "Requirement for input 4\n",
    "1. Sep by \"\\t\"\n",
    "2. must contains all the sample name in 1 & 2\n",
    "3. Only 2 columns\n",
    "    \n",
    "\n",
    "### Output\n",
    "For each collection, the output is 1:\n",
    "1. normalized Molecular_phenotype bed file\n",
    "\n",
    "It take care of the purple part of the following diagram\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "hired-wells",
   "metadata": {
    "kernel": "SoS",
    "tags": []
   },
   "source": [
    "### MWE:\n",
    "The mwe of input 1,2 can be generated by the mwe of bulk_expression_qc, that of input 5 can be generated by reference_data_processing\n",
    "The mwe of input 3,4 can be found here:https://drive.google.com/drive/u/0/folders/1Rv2bWHBbX_tastTh49ToYVDMV6rFP5Wk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "existing-nylon",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00201aa2-3ff7-4cab-84fc-3609ab5871c1",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "sos run normalization.ipynb Normalization \\\n",
    "--tpm_gct \"mwe.low_expression_filtered.outlier_removed.processed.tpm.gct\"      \\\n",
    "--counts_gct \"mwe.low_expression_filtered.outlier_removed.processed.geneCount.gct\"      \\\n",
    "--vcf_chr_list \"vcf_chrom_list\"   \\\n",
    "--sample_participant_lookup \"sampleSheetAfterQc.txt\" \\\n",
    "--container ./rna_quantification.sif --wd ./      \\\n",
    "--annotation_gtf Homo_sapiens.GRCh38.103.chr.reformated.ERCC.gene.gtf  &"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "instrumental-world",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[global]\n",
    "import os\n",
    "# Work directory & output directory\n",
    "parameter: wd = path\n",
    "# The filename namefor output data\n",
    "parameter: container = 'gaow/twas'\n",
    "# namefor the analysis output\n",
    "parameter: name= 'ROSMAP'\n",
    "# Output of bulk_expression_qc\n",
    "## An gene count table\n",
    "parameter: counts_gct = path\n",
    "## An gene TPM table\n",
    "parameter: tpm_gct = path\n",
    "# An gene gtf annotation table\n",
    "parameter: annotation_gtf = path\n",
    "# a file containing the number of chromosome in follow up analysis\n",
    "parameter: vcf_chr_list = path(\"./\")\n",
    "# A file to map sample ID from expression to genotype,must contain two columns, sample_id and participant_id, mapping IDs in the expression files to IDs in the genotype (these can be the same).\n",
    "parameter: sample_participant_lookup = path\n",
    "# For cluster jobs, number commands to run per job\n",
    "parameter: job_size = 1\n",
    "# Wall clock time expected\n",
    "parameter: walltime = \"5h\"\n",
    "# Memory expected\n",
    "parameter: mem = \"16G\"\n",
    "# Number of threads\n",
    "parameter: numThreads = 20\n",
    "parameter: container = \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "possible-wonder",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[Normalization]\n",
    "# Path to the input molecular phenotype data, should be a processd and indexed bed.gz file, with tabix index.\n",
    "input: counts_gct,tpm_gct, annotation_gtf\n",
    "output: f'{wd}/{_input[0]:bnnnnn}.expression.bed.gz',  # For factor\n",
    "task: trunk_workers = 1, trunk_size = 1, walltime = '4h',  mem = '20G', tags = f'{step_name}_{_output[0]:bn}'  \n",
    "bash: expand = \"${ }\", stderr = f'{_output[0]}.stderr', stdout = f'{_output[0]}.stdout',container = container\n",
    "    eqtl_prepare_expression.py ${tpm_gct} ${counts_gct} ${_input[2]} \\\n",
    "    ${sample_participant_lookup} ${ vcf_chr_list if vcf_chr_list is not path(\"./\") else \"\"} ${_output[0]:bnnn} \\\n",
    "    --tpm_threshold 0.1 \\\n",
    "    --count_threshold 1 \\\n",
    "    --sample_frac_threshold 0.2 \\\n",
    "    --normalization_method tmm"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "marine-excess",
   "metadata": {
    "kernel": "SoS"
   },
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "graphic-gospel",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca78514e-bfee-4c5c-830c-5ef6f153c46b",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "codemirror_mode": "sos",
   "file_extension": ".sos",
   "mimetype": "text/x-sos",
   "name": "sos",
   "nbconvert_exporter": "sos_notebook.converter.SoS_Exporter",
   "pygments_lexer": "sos"
  },
  "sos": {
   "kernels": [
    [
     "SoS",
     "sos",
     "",
     ""
    ]
   ],
   "version": "0.22.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
